{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1>Detection of Ampliconic gene families</h1>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from BCBio import GFF\n",
    "from BCBio.GFF import GFFExaminer\n",
    "examiner = GFFExaminer()\n",
    "\n",
    "from Bio import SeqIO\n",
    "from pprint import pprint\n",
    "import copy \n",
    "import gspread\n",
    "import os\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`data_dir` folder into which reference files should be loaded prior to analysis,\n",
    "expected structure\n",
    "\n",
    "```bash\n",
    "{data_dir}\n",
    "├── references\n",
    "│   ├── GorGor\n",
    "│   │   └── ...  # https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_029281585.2/\n",
    "│   ├── HomSap\n",
    "│   │   └── ...  # https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_009914755.1/\n",
    "│   ├── PanPan\n",
    "│   │   └── ...  # https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_029289425.2/\n",
    "│   ├── PanTro\n",
    "│   │   └── ...  # https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_028858775.2/\n",
    "│   ├── PonAbe\n",
    "│   │   └── ...  # https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_028885655.2/\n",
    "│   ├── PonPyg\n",
    "│   │   └── ...  # https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_028885625.2/\n",
    "│   └── SymSyn\n",
    "│       └── ...  # https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_028878055.3/\n",
    "│   └── MacFas\n",
    "│       └── ...  # https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_037993035.2/\n",
    "├── work_dir\n",
    "│       └── ...\n",
    "└── AdditionalFile2-SeqClasses.tsv # Download sequence classes file from additional data files\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the directory for the data\n",
    "data_dir = \"/home/emma/Amplicons/Workspaces/emma/downloaded_data\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "work_dir = f\"{data_dir}/work_dir/x_multicopy\" #Creates a new working directory if it doesn't exist already\n",
    "os.makedirs(work_dir, exist_ok=True) # ensures no error is raised if the directory already exists\n",
    "\n",
    "#Define the list of dictionaries (data) containing genome information for different species\n",
    "data = [\n",
    "    {'species':'PanTro',\n",
    "     'data': {'chr_y': \"NC_072422.2\",\n",
    "              'chr_x': \"NC_072421.2\",\n",
    "              'path_to_annotation_NCBI': f\"{data_dir}/references/PanTro/ncbi_dataset/data/GCF_028858775.2/genomic.gff\", #annotation file \n",
    "              'path_to_annotation_NCBI_chry': f\"{data_dir}/references/PanTro/ncbi_dataset/data/GCF_028858775.2/genomic_chrY.gff\", #Y specific annotation\n",
    "              'path_to_annotation_CAT': f\"{data_dir}/CAT/consensus_gene_set/GCF_028858775.2.gff3\", #path to an alternate annotation format (generated by CAT pipeline)\n",
    "              'ref':  f\"{data_dir}/references/PanTro/ncbi_dataset/data/GCF_028858775.2/GCF_028858775.2_NHGRI_mPanTro3-v2.0_pri_genomic.fna\", # primary reference genome file\n",
    "              'rna':  f\"{data_dir}/references/PanTro/ncbi_dataset/data/GCF_028858775.2/rna.fna\", #RNA sequences \n",
    "              'prot': f\"{data_dir}/references/PanTro/ncbi_dataset/data/GCF_028858775.2/protein.faa\", #protein sequences (FASTA format)\n",
    "              'cds': f\"{data_dir}/references/PanTro/ncbi_dataset/data/GCF_028858775.2/cds_from_genomic.fna\", #coding DNA seq derived from genome\n",
    " }},\n",
    "    {'species':'HomSap',\n",
    "     'data': {'chr_y': \"NC_060948.1\",\n",
    "              'chr_x': \"NC_060947.1\",\n",
    "              'path_to_annotation_NCBI': f\"{data_dir}/references/HomSap/ncbi_dataset/data/GCF_009914755.1/genomic.gff\",\n",
    "              'path_to_annotation_NCBI_chry': f\"{data_dir}/references/HomSap/ncbi_dataset/data/GCF_009914755.1/genomic_chrY.gff\",\n",
    "              'path_to_annotation_CAT': f\"{data_dir}/CAT/consensus_gene_set/hg38.gff3\",\n",
    "              'ref': f\"{data_dir}/references/HomSap/ncbi_dataset/data/GCF_009914755.1/GCF_009914755.1_T2T-CHM13v2.0_genomic.fna\",\n",
    "              'cds': f\"{data_dir}/references/HomSap/ncbi_dataset/data/GCF_009914755.1/cds_from_genomic.fna\",\n",
    "              'prot': f\"{data_dir}/references/HomSap/ncbi_dataset/data/GCF_009914755.1/protein.faa\",\n",
    "              }},\n",
    "     {'species':'PanPan',\n",
    "     'data': {'chr_y': \"NC_073273.2\",\n",
    "              'chr_x': \"NC_073272.2\",\n",
    "              'path_to_annotation_NCBI': f\"{data_dir}/references/PanPan/ncbi_dataset/data/GCF_029289425.2/genomic.gff\",\n",
    "              'path_to_annotation_NCBI_chry': f\"{data_dir}/references/PanPan/ncbi_dataset/data/GCF_029289425.2/genomic_chrY.gff\",\n",
    "              'path_to_annotation_CAT': f\"{data_dir}/CAT/consensus_gene_set/GCF_029289425.2.gff3\",\n",
    "              'ref': f\"{data_dir}/references/PanPan/ncbi_dataset/data/GCF_029289425.2/GCF_029289425.2_NHGRI_mPanPan1-v2.0_pri_genomic.fna\",\n",
    "              'cds': f'{data_dir}/references/PanPan/ncbi_dataset/data/GCF_029289425.2/cds_from_genomic.fna',\n",
    "              'prot': f'{data_dir}/references/PanPan/ncbi_dataset/data/GCF_029289425.2/protein.faa',\n",
    "              }},\n",
    "      {'species':'GorGor',\n",
    "     'data': {'chr_y': \"NC_073248.2\",\n",
    "              'chr_x': \"NC_073247.2\",\n",
    "              'path_to_annotation_NCBI': f\"{data_dir}/references/GorGor/ncbi_dataset/data/GCF_029281585.2/genomic.gff\",\n",
    "              'path_to_annotation_NCBI_chry': f\"{data_dir}/references/GorGor/ncbi_dataset/data/GCF_029281585.2/genomic_chrY.gff\",\n",
    "              'path_to_annotation_CAT': f\"{data_dir}/CAT/consensus_gene_set/GCF_029281585.2.gff3\",\n",
    "              'ref': f'{data_dir}/references/GorGor/ncbi_dataset/data/GCF_029281585.2/GCF_029281585.2_NHGRI_mGorGor1-v2.0_pri_genomic.fna',\n",
    "              'cds': f'{data_dir}/references/GorGor/ncbi_dataset/data/GCF_029281585.2/cds_from_genomic.fna',\n",
    "              'prot': f'{data_dir}/references/GorGor/ncbi_dataset/data/GCF_029281585.2/protein.faa',\n",
    "              }},\n",
    "    {'species':'PonPyg',\n",
    "     'data': {'chr_y': \"NC_072397.2\",\n",
    "              'chr_x': \"NC_072396.2\",\n",
    "              'path_to_annotation_NCBI': f\"{data_dir}/references/PonPyg/ncbi_dataset/data/GCF_028885625.2/genomic.gff\",\n",
    "              'path_to_annotation_NCBI_chry': f\"{data_dir}/references/PonPyg/ncbi_dataset/data/GCF_028885625.2/genomic_chrY.gff\",\n",
    "              'path_to_annotation_CAT': f\"{data_dir}/CAT/consensus_gene_set/GCF_028885625.2.gff3\",\n",
    "              'ref': f'{data_dir}/references/PonPyg/ncbi_dataset/data/GCF_028885625.2/GCF_028885625.2_NHGRI_mPonPyg2-v2.0_pri_genomic.fna',\n",
    "              'cds': f'{data_dir}/references/PonPyg/ncbi_dataset/data/GCF_028885625.2/cds_from_genomic.fna',\n",
    "              'prot': f'{data_dir}/references/PonPyg/ncbi_dataset/data/GCF_028885625.2/protein.faa',\n",
    "              }},\n",
    "    {'species':'PonAbe',\n",
    "     'data': {'chr_y': \"NC_072009.2\",\n",
    "              'chr_x': \"NC_072008.2\",\n",
    "              'path_to_annotation_NCBI': f\"{data_dir}/references/PonAbe/ncbi_dataset/data/GCF_028885655.2/genomic.gff\",\n",
    "              'path_to_annotation_NCBI_chry': f\"{data_dir}/references/PonAbe/ncbi_dataset/data/GCF_028885655.2/genomic_chrY.gff\",\n",
    "              'path_to_annotation_CAT': f\"{data_dir}/CAT/consensus_gene_set/GCF_028885655.2.gff3\",\n",
    "              'ref': f'{data_dir}/references/PonAbe/ncbi_dataset/data/GCF_028885655.2/GCF_028885655.2_NHGRI_mPonAbe1-v2.0_pri_genomic.fna',\n",
    "              'cds': f'{data_dir}/references/PonAbe/ncbi_dataset/data/GCF_028885655.2/cds_from_genomic.fna',\n",
    "              'prot': f'{data_dir}/references/PonAbe/ncbi_dataset/data/GCF_028885655.2/protein.faa',\n",
    "              }},    \n",
    "    {'species':'SymSyn',\n",
    "      'data': {'chr_y': \"NC_072448.2\",\n",
    "               'chr_x': \"NC_072447.2\",\n",
    "               'ref': f'{data_dir}/references/SymSyn/ncbi_dataset/data/GCF_028878055.3/GCF_028878055.3_NHGRI_mSymSyn1-v2.1_pri_genomic.fna',\n",
    "               'path_to_annotation_NCBI': f\"{data_dir}/references/SymSyn/ncbi_dataset/data/GCF_028878055.3/genomic.gff\",\n",
    "               'path_to_annotation_NCBI_chry': f\"{data_dir}/references/SymSyn/ncbi_dataset/data/GCF_028878055.3/genomic_chrY.gff\",\n",
    "               'path_to_annotation_CAT': f\"{data_dir}/CAT/consensus_gene_set/GCF_028878055.3.gff3\",\n",
    "               'cds': f'{data_dir}/references/SymSyn/ncbi_dataset/data/GCF_028878055.3/cds_from_genomic.fna',\n",
    "               'prot': f'{data_dir}/references/SymSyn/ncbi_dataset/data/GCF_028878055.3/protein.faa',\n",
    "               }},\n",
    "    {'species':'MacFas',\n",
    "      'data': {'chr_y': \"NC_132903.1\",\n",
    "               'chr_x': \"NC_088395.1\",\n",
    "               'ref': f'{data_dir}/references/MacFas/ncbi_dataset/data/GCF_037993035.2/GCF_037993035.2_T2T-MFA8v1.1_genomic.fna',\n",
    "               'path_to_annotation_NCBI': f\"{data_dir}/references/MacFas/ncbi_dataset/data/GCF_037993035.2/genomic.gff\",\n",
    "               'cds': f'{data_dir}/references/MacFas/ncbi_dataset/data/GCF_037993035.2/cds_from_genomic.fna',\n",
    "               'prot': f'{data_dir}/references/MacFas/ncbi_dataset/data/GCF_037993035.2/protein.faa',\n",
    "               }},\n",
    "]\n",
    "\n",
    "# Maps species identifiers to their common name\n",
    "species_to_sequence_spec = {\n",
    "    'PanTro': 'chimpanzee',\n",
    "    'HomSap': 'human',\n",
    "    'PanPan': 'bonobo',\n",
    "    'GorGor': 'gorilla',\n",
    "    'PonPyg': 'b-orang',\n",
    "    'PonAbe': 's-orang',\n",
    "    'SymSyn': 'siamang',\n",
    "    'MacFas': 'macaque'\n",
    "}\n",
    "\n",
    "# Extracting species names -> list of species identifiers by iterating t\n",
    "species_list = [d['species'] for d in data]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['PanTro',\n",
       " 'HomSap',\n",
       " 'PanPan',\n",
       " 'GorGor',\n",
       " 'PonPyg',\n",
       " 'PonAbe',\n",
       " 'SymSyn',\n",
       " 'MacFas']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PanTro /home/emma/Amplicons/Workspaces/emma/downloaded_data/references/PanTro/ncbi_dataset/data/GCF_028858775.2/genomic.gff NC_072421.2\n",
      "PanTro /home/emma/Amplicons/Workspaces/emma/downloaded_data/references/PanTro/ncbi_dataset/data/GCF_028858775.2/genomic.gff NC_072422.2\n",
      "HomSap /home/emma/Amplicons/Workspaces/emma/downloaded_data/references/HomSap/ncbi_dataset/data/GCF_009914755.1/genomic.gff NC_060947.1\n",
      "HomSap /home/emma/Amplicons/Workspaces/emma/downloaded_data/references/HomSap/ncbi_dataset/data/GCF_009914755.1/genomic.gff NC_060948.1\n",
      "PanPan /home/emma/Amplicons/Workspaces/emma/downloaded_data/references/PanPan/ncbi_dataset/data/GCF_029289425.2/genomic.gff NC_073272.2\n",
      "PanPan /home/emma/Amplicons/Workspaces/emma/downloaded_data/references/PanPan/ncbi_dataset/data/GCF_029289425.2/genomic.gff NC_073273.2\n",
      "GorGor /home/emma/Amplicons/Workspaces/emma/downloaded_data/references/GorGor/ncbi_dataset/data/GCF_029281585.2/genomic.gff NC_073247.2\n",
      "GorGor /home/emma/Amplicons/Workspaces/emma/downloaded_data/references/GorGor/ncbi_dataset/data/GCF_029281585.2/genomic.gff NC_073248.2\n",
      "PonPyg /home/emma/Amplicons/Workspaces/emma/downloaded_data/references/PonPyg/ncbi_dataset/data/GCF_028885625.2/genomic.gff NC_072396.2\n",
      "PonPyg /home/emma/Amplicons/Workspaces/emma/downloaded_data/references/PonPyg/ncbi_dataset/data/GCF_028885625.2/genomic.gff NC_072397.2\n",
      "PonAbe /home/emma/Amplicons/Workspaces/emma/downloaded_data/references/PonAbe/ncbi_dataset/data/GCF_028885655.2/genomic.gff NC_072008.2\n",
      "PonAbe /home/emma/Amplicons/Workspaces/emma/downloaded_data/references/PonAbe/ncbi_dataset/data/GCF_028885655.2/genomic.gff NC_072009.2\n",
      "SymSyn /home/emma/Amplicons/Workspaces/emma/downloaded_data/references/SymSyn/ncbi_dataset/data/GCF_028878055.3/genomic.gff NC_072447.2\n",
      "SymSyn /home/emma/Amplicons/Workspaces/emma/downloaded_data/references/SymSyn/ncbi_dataset/data/GCF_028878055.3/genomic.gff NC_072448.2\n",
      "MacFas /home/emma/Amplicons/Workspaces/emma/downloaded_data/references/MacFas/ncbi_dataset/data/GCF_037993035.2/genomic.gff NC_088395.1\n",
      "MacFas /home/emma/Amplicons/Workspaces/emma/downloaded_data/references/MacFas/ncbi_dataset/data/GCF_037993035.2/genomic.gff NC_132903.1\n"
     ]
    }
   ],
   "source": [
    "#subset annotation file to only include chromosome X & Y\n",
    "for species in species_list:\n",
    "\n",
    "    annotation_file =  [d for d in data if d['species'] == species][0]['data']['path_to_annotation_NCBI']\n",
    "    annotation_file_x = annotation_file.replace(\".gff\", \"_chrX.gff\")\n",
    "    chr = [d for d in data if d['species'] == species][0]['data']['chr_x']\n",
    "    print(species, annotation_file, chr)\n",
    "\n",
    "    ! cat $annotation_file | grep -v \"#\" | grep -w $chr | sort -k1,1V -k4,4n -k5,5rn -k3,3r - > $annotation_file_x\n",
    "\n",
    "    annotation_file_y = annotation_file.replace(\".gff\", \"_chrY.gff\")\n",
    "    chr = [d for d in data if d['species'] == species][0]['data']['chr_y']\n",
    "    print(species, annotation_file, chr)\n",
    "\n",
    "    ! cat $annotation_file | grep -v \"#\" | grep -w $chr | sort -k1,1V -k4,4n -k5,5rn -k3,3r - > $annotation_file_y"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Chromosome X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#runs around 20s without extracting the sequence\n",
    "#runs around 1m with extracting the sequence\n",
    "#only collect ID's will extract sequence from cds_from_genomic.fna\n",
    "record_dict = {}\n",
    "\n",
    "for species in species_list:\n",
    "\n",
    "\n",
    "    annotation_file =  [d for d in data if d['species'] == species][0]['data']['path_to_annotation_NCBI'].replace(\".gff\", \"_chrX.gff\")\n",
    "    handle = open(annotation_file)\n",
    "\n",
    "    # pprint.pprint(examiner.parent_child_map(handle))\n",
    "\n",
    "    # assert False\n",
    "\n",
    "    # records = GFF.parse(handle,base_dict=seq_dict)\n",
    "    records = GFF.parse(handle)\n",
    "    for rec in records:\n",
    "        protss = set()\n",
    "        for feature in rec.features:\n",
    "            if feature.type != \"gene\":\n",
    "                continue\n",
    "            if \"description\" in feature.qualifiers:\n",
    "                description = feature.qualifiers[\"description\"][0]\n",
    "\n",
    "                gene_id = feature.id.replace(\"gene-\", \"\")\n",
    "                gene_spec_id = gene_id + \"_\" + species\n",
    "                    \n",
    "                # print(feature.qualifiers)\n",
    "                for subf in feature.sub_features:\n",
    "\n",
    "                    if subf.type == \"mRNA\":\n",
    "                        mrna_id = subf.id.replace(\"rna-\", \"\")\n",
    "                        for subsubf in subf.sub_features:\n",
    "                            if subsubf.type != \"CDS\":\n",
    "                                continue\n",
    "                            # cds_id = subsubf.id.replace(\"cds-\", \"\")\n",
    "                            prot_id = subsubf.qualifiers[\"protein_id\"][0]\n",
    "                            if prot_id in protss:\n",
    "                                continue\n",
    "\n",
    "                            protss.add(prot_id)\n",
    "                            element = {}\n",
    "                            element[\"specie\"] = species\n",
    "                            element[\"name\"] = description\n",
    "                            element[\"id\"] = prot_id\n",
    "                            element[\"mrna_id\"] = mrna_id\n",
    "                            element[\"gene_id\"] = gene_id\n",
    "                            element[\"start\"] = int(subsubf.location.start)\n",
    "                            element[\"end\"] = int(subsubf.location.end)\n",
    "                            element[\"strand\"] = \"+\" if subsubf.location.strand == 1 else \"-\"  # Extracting strand (adapted from original script to later be able to look at the organization of the genes)\n",
    "                            element[\"gene_spec_id\"] = gene_spec_id\n",
    "                            \n",
    "                            # element[\"seq\"] = str(subf.extract(rec.seq))\n",
    "                            if gene_id in record_dict:\n",
    "                                record_dict[gene_id].append(element)\n",
    "                            else:\n",
    "                                record_dict[gene_id] = [element]\n",
    "\n",
    "    handle.close()\n",
    "\n",
    "#pprint(record_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#read sequence classes for all species\n",
    "seq_class_dicts = {}\n",
    "for species in species_list:\n",
    "\n",
    "    sequence_class_file = f\"{data_dir}/AdditionalFile2-SeqClasses.tsv\"\n",
    "    seq_class_dicts[species] = []\n",
    "    with open(sequence_class_file) as f:\n",
    "        for line in f:\n",
    "            line = line.strip().split(\"\\t\")\n",
    "            if line[0] == species_to_sequence_spec[species]:\n",
    "                if line[2] == \"chrX\":\n",
    "                    seq_class_dicts[species].append(line[2:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'PanTro': [['chrX', '0', '591255', 'SATELLITE'],\n",
       "  ['chrX', '591255', '695954', 'ANCESTRAL'],\n",
       "  ['chrX', '695954', '3166326', 'PAR'],\n",
       "  ['chrX', '3166326', '3424000', 'ANCESTRAL'],\n",
       "  ['chrX', '3424000', '3593000', 'AMPLICONIC'],\n",
       "  ['chrX', '3593000', '8612703', 'ANCESTRAL'],\n",
       "  ['chrX', '8612703', '8670650', 'AMPLICONIC'],\n",
       "  ['chrX', '8670650', '8987844', 'ANCESTRAL'],\n",
       "  ['chrX', '8987844', '8998644', 'AMPLICONIC'],\n",
       "  ['chrX', '8998644', '48626000', 'ANCESTRAL'],\n",
       "  ['chrX', '48626000', '48723464', 'AMPLICONIC'],\n",
       "  ['chrX', '48723464', '51703961', 'ANCESTRAL'],\n",
       "  ['chrX', '51703961', '51763787', 'AMPLICONIC'],\n",
       "  ['chrX', '51763787', '52083202', 'ANCESTRAL'],\n",
       "  ['chrX', '52083202', '52119541', 'AMPLICONIC'],\n",
       "  ['chrX', '52119541', '52219366', 'ANCESTRAL'],\n",
       "  ['chrX', '52219366', '52255733', 'AMPLICONIC'],\n",
       "  ['chrX', '52255733', '52498000', 'ANCESTRAL'],\n",
       "  ['chrX', '52498000', '53021090', 'AMPLICONIC'],\n",
       "  ['chrX', '53021090', '55492131', 'ANCESTRAL'],\n",
       "  ['chrX', '55492131', '55560177', 'AMPLICONIC'],\n",
       "  ['chrX', '55560177', '58108941', 'ANCESTRAL'],\n",
       "  ['chrX', '58108941', '60525984', 'SATELLITE'],\n",
       "  ['chrX', '60525984', '60829359', 'ANCESTRAL'],\n",
       "  ['chrX', '60829359', '61025000', 'AMPLICONIC'],\n",
       "  ['chrX', '61025000', '69454000', 'ANCESTRAL'],\n",
       "  ['chrX', '69454000', '69588637', 'AMPLICONIC'],\n",
       "  ['chrX', '69588637', '70539410', 'ANCESTRAL'],\n",
       "  ['chrX', '70539410', '70881264', 'AMPLICONIC'],\n",
       "  ['chrX', '70881264', '99768533', 'ANCESTRAL'],\n",
       "  ['chrX', '99768533', '100096366', 'AMPLICONIC'],\n",
       "  ['chrX', '100096366', '103774950', 'ANCESTRAL'],\n",
       "  ['chrX', '103774950', '103808022', 'AMPLICONIC'],\n",
       "  ['chrX', '103808022', '113212000', 'ANCESTRAL'],\n",
       "  ['chrX', '113212000', '113455000', 'AMPLICONIC'],\n",
       "  ['chrX', '113455000', '117572000', 'ANCESTRAL'],\n",
       "  ['chrX', '117572000', '117821000', 'AMPLICONIC'],\n",
       "  ['chrX', '117821000', '132578959', 'ANCESTRAL'],\n",
       "  ['chrX', '132578959', '132719345', 'AMPLICONIC'],\n",
       "  ['chrX', '132719345', '133176937', 'ANCESTRAL'],\n",
       "  ['chrX', '133176937', '133386039', 'AMPLICONIC'],\n",
       "  ['chrX', '133386039', '138526750', 'ANCESTRAL'],\n",
       "  ['chrX', '138526750', '138634855', 'AMPLICONIC'],\n",
       "  ['chrX', '138634855', '138967293', 'ANCESTRAL'],\n",
       "  ['chrX', '138967293', '139183000', 'AMPLICONIC'],\n",
       "  ['chrX', '139183000', '144173226', 'ANCESTRAL'],\n",
       "  ['chrX', '144173226', '144212161', 'AMPLICONIC'],\n",
       "  ['chrX', '144212161', '146918265', 'ANCESTRAL'],\n",
       "  ['chrX', '146918265', '147449963', 'AMPLICONIC'],\n",
       "  ['chrX', '147449963', '150159449', 'ANCESTRAL'],\n",
       "  ['chrX', '150159449', '150267474', 'AMPLICONIC'],\n",
       "  ['chrX', '150267474', '150548780', 'ANCESTRAL'],\n",
       "  ['chrX', '150548780', '150608383', 'AMPLICONIC'],\n",
       "  ['chrX', '150608383', '150716349', 'ANCESTRAL'],\n",
       "  ['chrX', '150716349', '150754873', 'AMPLICONIC'],\n",
       "  ['chrX', '150754873', '151723714', 'ANCESTRAL'],\n",
       "  ['chrX', '151723714', '151784038', 'AMPLICONIC'],\n",
       "  ['chrX', '151784038', '151943436', 'ANCESTRAL'],\n",
       "  ['chrX', '151943436', '152036661', 'AMPLICONIC'],\n",
       "  ['chrX', '152036661', '152717736', 'ANCESTRAL'],\n",
       "  ['chrX', '152717736', '152883896', 'AMPLICONIC'],\n",
       "  ['chrX', '152883896', '153892427', 'ANCESTRAL']],\n",
       " 'HomSap': [['chrX', '0', '2394410', 'PAR'],\n",
       "  ['chrX', '2394410', '47660000', 'ANCESTRAL'],\n",
       "  ['chrX', '47660000', '47837442', 'AMPLICONIC'],\n",
       "  ['chrX', '47837442', '50939485', 'ANCESTRAL'],\n",
       "  ['chrX', '50939485', '50996935', 'AMPLICONIC'],\n",
       "  ['chrX', '50996935', '51312463', 'ANCESTRAL'],\n",
       "  ['chrX', '51312463', '51348894', 'AMPLICONIC'],\n",
       "  ['chrX', '51348894', '51448683', 'ANCESTRAL'],\n",
       "  ['chrX', '51448683', '51485117', 'AMPLICONIC'],\n",
       "  ['chrX', '51485117', '51718000', 'ANCESTRAL'],\n",
       "  ['chrX', '51718000', '52258497', 'AMPLICONIC'],\n",
       "  ['chrX', '52258497', '54746693', 'ANCESTRAL'],\n",
       "  ['chrX', '54746693', '54812839', 'AMPLICONIC'],\n",
       "  ['chrX', '54812839', '57347714', 'ANCESTRAL'],\n",
       "  ['chrX', '57347714', '61248490', 'SATELLITE'],\n",
       "  ['chrX', '61248490', '61553393', 'ANCESTRAL'],\n",
       "  ['chrX', '61553393', '61675215', 'AMPLICONIC'],\n",
       "  ['chrX', '61675215', '70116403', 'ANCESTRAL'],\n",
       "  ['chrX', '70116403', '70232226', 'AMPLICONIC'],\n",
       "  ['chrX', '70232226', '71174416', 'ANCESTRAL'],\n",
       "  ['chrX', '71174416', '71520476', 'AMPLICONIC'],\n",
       "  ['chrX', '71520476', '87642550', 'ANCESTRAL'],\n",
       "  ['chrX', '87642550', '91570785', 'XTR'],\n",
       "  ['chrX', '91570785', '100642134', 'ANCESTRAL'],\n",
       "  ['chrX', '100642134', '100934056', 'AMPLICONIC'],\n",
       "  ['chrX', '100934056', '102410141', 'ANCESTRAL'],\n",
       "  ['chrX', '102410141', '102510844', 'AMPLICONIC'],\n",
       "  ['chrX', '102510844', '104700096', 'ANCESTRAL'],\n",
       "  ['chrX', '104700096', '104733867', 'AMPLICONIC'],\n",
       "  ['chrX', '104733867', '114134000', 'ANCESTRAL'],\n",
       "  ['chrX', '114134000', '114299000', 'AMPLICONIC'],\n",
       "  ['chrX', '114299000', '118413361', 'ANCESTRAL'],\n",
       "  ['chrX', '118413361', '118573219', 'AMPLICONIC'],\n",
       "  ['chrX', '118573219', '133441916', 'ANCESTRAL'],\n",
       "  ['chrX', '133441916', '133582232', 'AMPLICONIC'],\n",
       "  ['chrX', '133582232', '134042960', 'ANCESTRAL'],\n",
       "  ['chrX', '134042960', '134198068', 'AMPLICONIC'],\n",
       "  ['chrX', '134198068', '139312784', 'ANCESTRAL'],\n",
       "  ['chrX', '139312784', '139432092', 'AMPLICONIC'],\n",
       "  ['chrX', '139432092', '139552830', 'ANCESTRAL'],\n",
       "  ['chrX', '139552830', '139567026', 'AMPLICONIC'],\n",
       "  ['chrX', '139567026', '139781327', 'ANCESTRAL'],\n",
       "  ['chrX', '139781327', '140007734', 'AMPLICONIC'],\n",
       "  ['chrX', '140007734', '145060651', 'ANCESTRAL'],\n",
       "  ['chrX', '145060651', '145082934', 'AMPLICONIC'],\n",
       "  ['chrX', '145082934', '147806320', 'ANCESTRAL'],\n",
       "  ['chrX', '147806320', '148061999', 'AMPLICONIC'],\n",
       "  ['chrX', '148061999', '148184920', 'ANCESTRAL'],\n",
       "  ['chrX', '148184920', '148206523', 'AMPLICONIC'],\n",
       "  ['chrX', '148206523', '150945170', 'ANCESTRAL'],\n",
       "  ['chrX', '150945170', '151056456', 'AMPLICONIC'],\n",
       "  ['chrX', '151056456', '151333544', 'ANCESTRAL'],\n",
       "  ['chrX', '151333544', '151417247', 'AMPLICONIC'],\n",
       "  ['chrX', '151417247', '151517761', 'ANCESTRAL'],\n",
       "  ['chrX', '151517761', '151561456', 'AMPLICONIC'],\n",
       "  ['chrX', '151561456', '152572450', 'ANCESTRAL'],\n",
       "  ['chrX', '152572450', '152632742', 'AMPLICONIC'],\n",
       "  ['chrX', '152632742', '152791974', 'ANCESTRAL'],\n",
       "  ['chrX', '152791974', '152885316', 'AMPLICONIC'],\n",
       "  ['chrX', '152885316', '153574028', 'ANCESTRAL'],\n",
       "  ['chrX', '153574028', '153742802', 'AMPLICONIC'],\n",
       "  ['chrX', '153742802', '153925834', 'ANCESTRAL'],\n",
       "  ['chrX', '153925834', '154259566', 'PAR'],\n",
       "  ['chrX', '3756', '2396333', 'PAR'],\n",
       "  ['chrX', '2396334', '47676625', 'ANCESTRAL'],\n",
       "  ['chrX', '47676626', '47854033', 'AMPLICONIC'],\n",
       "  ['chrX', '47854034', '50976569', 'ANCESTRAL'],\n",
       "  ['chrX', '50976570', '51033938', 'AMPLICONIC'],\n",
       "  ['chrX', '51033939', '51349384', 'ANCESTRAL'],\n",
       "  ['chrX', '51349385', '51385843', 'AMPLICONIC'],\n",
       "  ['chrX', '51385844', '51485687', 'ANCESTRAL'],\n",
       "  ['chrX', '51485688', '51522143', 'AMPLICONIC'],\n",
       "  ['chrX', '51522144', '51755020', 'ANCESTRAL'],\n",
       "  ['chrX', '51755021', '52243350', 'AMPLICONIC'],\n",
       "  ['chrX', '52295804', '54783663', 'ANCESTRAL'],\n",
       "  ['chrX', '54783664', '54849821', 'AMPLICONIC'],\n",
       "  ['chrX', '54849822', '57399356', 'ANCESTRAL'],\n",
       "  ['chrX', '57399357', '61304749', 'SATELLITE'],\n",
       "  ['chrX', '61304750', '61609591', 'ANCESTRAL'],\n",
       "  ['chrX', '61609592', '61731451', 'AMPLICONIC'],\n",
       "  ['chrX', '61731452', '70173723', 'ANCESTRAL'],\n",
       "  ['chrX', '70173724', '70289192', 'AMPLICONIC'],\n",
       "  ['chrX', '70289193', '71230989', 'ANCESTRAL'],\n",
       "  ['chrX', '71230990', '71576992', 'AMPLICONIC'],\n",
       "  ['chrX', '71576993', '87717822', 'ANCESTRAL'],\n",
       "  ['chrX', '87723769', '91647350', 'XTR'],\n",
       "  ['chrX', '91647351', '100728911', 'ANCESTRAL'],\n",
       "  ['chrX', '100728912', '101020815', 'AMPLICONIC'],\n",
       "  ['chrX', '101020816', '102502636', 'ANCESTRAL'],\n",
       "  ['chrX', '102502637', '102603304', 'AMPLICONIC'],\n",
       "  ['chrX', '102603305', '104792494', 'ANCESTRAL'],\n",
       "  ['chrX', '104792495', '104826264', 'AMPLICONIC'],\n",
       "  ['chrX', '104826265', '114230186', 'ANCESTRAL'],\n",
       "  ['chrX', '114271781', '114437047', 'AMPLICONIC'],\n",
       "  ['chrX', '114445935', '118575508', 'ANCESTRAL'],\n",
       "  ['chrX', '118575509', '118735366', 'AMPLICONIC'],\n",
       "  ['chrX', '118735367', '133592973', 'ANCESTRAL'],\n",
       "  ['chrX', '133592974', '133733278', 'AMPLICONIC'],\n",
       "  ['chrX', '133733279', '134193999', 'ANCESTRAL'],\n",
       "  ['chrX', '134224840', '134331964', 'AMPLICONIC'],\n",
       "  ['chrX', '134331965', '139447204', 'ANCESTRAL'],\n",
       "  ['chrX', '139447205', '139566456', 'AMPLICONIC'],\n",
       "  ['chrX', '139566457', '139789413', 'ANCESTRAL'],\n",
       "  ['chrX', '139789414', '139803559', 'AMPLICONIC'],\n",
       "  ['chrX', '139803560', '139924303', 'ANCESTRAL'],\n",
       "  ['chrX', '139924304', '140150692', 'AMPLICONIC'],\n",
       "  ['chrX', '140150693', '145200847', 'ANCESTRAL'],\n",
       "  ['chrX', '145200848', '145222464', 'AMPLICONIC'],\n",
       "  ['chrX', '145222465', '147949132', 'ANCESTRAL'],\n",
       "  ['chrX', '147949133', '148204831', 'AMPLICONIC'],\n",
       "  ['chrX', '148204832', '148327646', 'ANCESTRAL'],\n",
       "  ['chrX', '148327647', '148349248', 'AMPLICONIC'],\n",
       "  ['chrX', '148349249', '151084848', 'ANCESTRAL'],\n",
       "  ['chrX', '151084849', '151195517', 'AMPLICONIC'],\n",
       "  ['chrX', '151195518', '151473035', 'ANCESTRAL'],\n",
       "  ['chrX', '151482652', '151556340', 'AMPLICONIC'],\n",
       "  ['chrX', '151556341', '151656865', 'ANCESTRAL'],\n",
       "  ['chrX', '151656866', '151700587', 'AMPLICONIC'],\n",
       "  ['chrX', '151700588', '152671217', 'ANCESTRAL'],\n",
       "  ['chrX', '152671218', '152731521', 'AMPLICONIC'],\n",
       "  ['chrX', '152731522', '152890757', 'ANCESTRAL'],\n",
       "  ['chrX', '152890758', '152972404', 'AMPLICONIC'],\n",
       "  ['chrX', '152972405', '153661156', 'ANCESTRAL'],\n",
       "  ['chrX', '153661157', '153829971', 'AMPLICONIC'],\n",
       "  ['chrX', '153829972', '154013003', 'ANCESTRAL'],\n",
       "  ['chrX', '154013004', '154346656', 'PAR']],\n",
       " 'PanPan': [['chrX', '0', '70282', 'ANCESTRAL'],\n",
       "  ['chrX', '70282', '2527836', 'PAR'],\n",
       "  ['chrX', '2527836', '2784000', 'ANCESTRAL'],\n",
       "  ['chrX', '2784000', '4699000', 'AMPLICONIC'],\n",
       "  ['chrX', '4699000', '9709644', 'ANCESTRAL'],\n",
       "  ['chrX', '9709644', '9837000', 'AMPLICONIC'],\n",
       "  ['chrX', '9837000', '10133825', 'ANCESTRAL'],\n",
       "  ['chrX', '10133825', '10144757', 'AMPLICONIC'],\n",
       "  ['chrX', '10144757', '38893529', 'ANCESTRAL'],\n",
       "  ['chrX', '38893529', '38912172', 'AMPLICONIC'],\n",
       "  ['chrX', '38912172', '39145881', 'ANCESTRAL'],\n",
       "  ['chrX', '39145881', '39164544', 'AMPLICONIC'],\n",
       "  ['chrX', '39164544', '49752000', 'ANCESTRAL'],\n",
       "  ['chrX', '49752000', '49920372', 'AMPLICONIC'],\n",
       "  ['chrX', '49920372', '53059393', 'ANCESTRAL'],\n",
       "  ['chrX', '53059393', '53117879', 'AMPLICONIC'],\n",
       "  ['chrX', '53117879', '53437475', 'ANCESTRAL'],\n",
       "  ['chrX', '53437475', '53473837', 'AMPLICONIC'],\n",
       "  ['chrX', '53473837', '53573734', 'ANCESTRAL'],\n",
       "  ['chrX', '53573734', '53610111', 'AMPLICONIC'],\n",
       "  ['chrX', '53610111', '53862897', 'ANCESTRAL'],\n",
       "  ['chrX', '53862897', '53924278', 'AMPLICONIC'],\n",
       "  ['chrX', '53924278', '54033152', 'ANCESTRAL'],\n",
       "  ['chrX', '54033152', '54375021', 'AMPLICONIC'],\n",
       "  ['chrX', '54375021', '56887352', 'ANCESTRAL'],\n",
       "  ['chrX', '56887352', '56955301', 'AMPLICONIC'],\n",
       "  ['chrX', '56955301', '59530707', 'ANCESTRAL'],\n",
       "  ['chrX', '59530707', '62384166', 'SATELLITE'],\n",
       "  ['chrX', '62384166', '62688918', 'ANCESTRAL'],\n",
       "  ['chrX', '62688918', '62807889', 'AMPLICONIC'],\n",
       "  ['chrX', '62807889', '71340000', 'ANCESTRAL'],\n",
       "  ['chrX', '71340000', '71476933', 'AMPLICONIC'],\n",
       "  ['chrX', '71476933', '72416259', 'ANCESTRAL'],\n",
       "  ['chrX', '72416259', '72758214', 'AMPLICONIC'],\n",
       "  ['chrX', '72758214', '101805633', 'ANCESTRAL'],\n",
       "  ['chrX', '101805633', '102103638', 'AMPLICONIC'],\n",
       "  ['chrX', '102103638', '105777052', 'ANCESTRAL'],\n",
       "  ['chrX', '105777052', '105810210', 'AMPLICONIC'],\n",
       "  ['chrX', '105810210', '115188000', 'ANCESTRAL'],\n",
       "  ['chrX', '115188000', '115309000', 'AMPLICONIC'],\n",
       "  ['chrX', '115309000', '119448658', 'ANCESTRAL'],\n",
       "  ['chrX', '119448658', '119765698', 'AMPLICONIC'],\n",
       "  ['chrX', '119765698', '134552645', 'ANCESTRAL'],\n",
       "  ['chrX', '134552645', '134693478', 'AMPLICONIC'],\n",
       "  ['chrX', '134693478', '135056000', 'ANCESTRAL'],\n",
       "  ['chrX', '135056000', '135366803', 'AMPLICONIC'],\n",
       "  ['chrX', '135366803', '140517619', 'ANCESTRAL'],\n",
       "  ['chrX', '140517619', '140609194', 'AMPLICONIC'],\n",
       "  ['chrX', '140609194', '140937816', 'ANCESTRAL'],\n",
       "  ['chrX', '140937816', '141146035', 'AMPLICONIC'],\n",
       "  ['chrX', '141146035', '146100867', 'ANCESTRAL'],\n",
       "  ['chrX', '146100867', '146148597', 'AMPLICONIC'],\n",
       "  ['chrX', '146148597', '148852000', 'ANCESTRAL'],\n",
       "  ['chrX', '148852000', '150075000', 'AMPLICONIC'],\n",
       "  ['chrX', '150075000', '152740000', 'ANCESTRAL'],\n",
       "  ['chrX', '152740000', '152847000', 'AMPLICONIC'],\n",
       "  ['chrX', '152847000', '153133324', 'ANCESTRAL'],\n",
       "  ['chrX', '153133324', '153192269', 'AMPLICONIC'],\n",
       "  ['chrX', '153192269', '153299354', 'ANCESTRAL'],\n",
       "  ['chrX', '153299354', '153337131', 'AMPLICONIC'],\n",
       "  ['chrX', '153337131', '154314082', 'ANCESTRAL'],\n",
       "  ['chrX', '154314082', '154374090', 'AMPLICONIC'],\n",
       "  ['chrX', '154374090', '154533262', 'ANCESTRAL'],\n",
       "  ['chrX', '154533262', '154626415', 'AMPLICONIC'],\n",
       "  ['chrX', '154626415', '155308398', 'ANCESTRAL'],\n",
       "  ['chrX', '155308398', '155475354', 'AMPLICONIC'],\n",
       "  ['chrX', '155475354', '155956000', 'ANCESTRAL'],\n",
       "  ['chrX', '155956000', '156091000', 'AMPLICONIC'],\n",
       "  ['chrX', '156091000', '156408085', 'ANCESTRAL'],\n",
       "  ['chrX', '156408085', '156502125', 'PAR'],\n",
       "  ['chrX', '156502125', '156785228', 'ANCESTRAL'],\n",
       "  ['chrX', '156785228', '157903709', 'SATELLITE'],\n",
       "  ['chrX', '157903709', '158072155', 'ANCESTRAL'],\n",
       "  ['chrX', '158072155', '158683551', 'SATELLITE'],\n",
       "  ['chrX', '158683551', '158981128', 'ANCESTRAL'],\n",
       "  ['chrX', '158981128', '160249396', 'SATELLITE']],\n",
       " 'GorGor': [['chrX', '0', '1814309', 'SATELLITE'],\n",
       "  ['chrX', '1814309', '2069345', 'ANCESTRAL'],\n",
       "  ['chrX', '2069345', '3143119', 'SATELLITE'],\n",
       "  ['chrX', '3143119', '3196000', 'ANCESTRAL'],\n",
       "  ['chrX', '3196000', '3335000', 'AMPLICONIC'],\n",
       "  ['chrX', '3335000', '3345362', 'ANCESTRAL'],\n",
       "  ['chrX', '3345362', '3982472', 'SATELLITE'],\n",
       "  ['chrX', '3982472', '4079849', 'ANCESTRAL'],\n",
       "  ['chrX', '4079849', '4708441', 'SATELLITE'],\n",
       "  ['chrX', '4708441', '5098024', 'ANCESTRAL'],\n",
       "  ['chrX', '5098024', '6392890', 'SATELLITE'],\n",
       "  ['chrX', '6392890', '6560823', 'ANCESTRAL'],\n",
       "  ['chrX', '6560823', '9553142', 'SATELLITE'],\n",
       "  ['chrX', '9553142', '12048775', 'PAR'],\n",
       "  ['chrX', '12048775', '12299075', 'ANCESTRAL'],\n",
       "  ['chrX', '12299075', '13337000', 'AMPLICONIC'],\n",
       "  ['chrX', '13337000', '13984000', 'ANCESTRAL'],\n",
       "  ['chrX', '13984000', '14437000', 'AMPLICONIC'],\n",
       "  ['chrX', '14437000', '58638000', 'ANCESTRAL'],\n",
       "  ['chrX', '58638000', '58807530', 'AMPLICONIC'],\n",
       "  ['chrX', '58807530', '61857791', 'ANCESTRAL'],\n",
       "  ['chrX', '61857791', '61919344', 'AMPLICONIC'],\n",
       "  ['chrX', '61919344', '62237443', 'ANCESTRAL'],\n",
       "  ['chrX', '62237443', '62398508', 'AMPLICONIC'],\n",
       "  ['chrX', '62398508', '62640000', 'ANCESTRAL'],\n",
       "  ['chrX', '62640000', '63100028', 'AMPLICONIC'],\n",
       "  ['chrX', '63100028', '65509642', 'ANCESTRAL'],\n",
       "  ['chrX', '65509642', '65577498', 'AMPLICONIC'],\n",
       "  ['chrX', '65577498', '68254530', 'ANCESTRAL'],\n",
       "  ['chrX', '68254530', '74519264', 'SATELLITE'],\n",
       "  ['chrX', '74519264', '74823345', 'ANCESTRAL'],\n",
       "  ['chrX', '74823345', '74955501', 'AMPLICONIC'],\n",
       "  ['chrX', '74955501', '84468600', 'ANCESTRAL'],\n",
       "  ['chrX', '84468600', '84827918', 'AMPLICONIC'],\n",
       "  ['chrX', '84827918', '102390000', 'ANCESTRAL'],\n",
       "  ['chrX', '102390000', '102543000', 'AMPLICONIC'],\n",
       "  ['chrX', '102543000', '113950000', 'ANCESTRAL'],\n",
       "  ['chrX', '113950000', '114260011', 'AMPLICONIC'],\n",
       "  ['chrX', '114260011', '118051468', 'ANCESTRAL'],\n",
       "  ['chrX', '118051468', '118084636', 'AMPLICONIC'],\n",
       "  ['chrX', '118084636', '127470000', 'ANCESTRAL'],\n",
       "  ['chrX', '127470000', '127675000', 'AMPLICONIC'],\n",
       "  ['chrX', '127675000', '131733649', 'ANCESTRAL'],\n",
       "  ['chrX', '131733649', '131895217', 'AMPLICONIC'],\n",
       "  ['chrX', '131895217', '146884874', 'ANCESTRAL'],\n",
       "  ['chrX', '146884874', '147025081', 'AMPLICONIC'],\n",
       "  ['chrX', '147025081', '147404243', 'ANCESTRAL'],\n",
       "  ['chrX', '147404243', '147587259', 'AMPLICONIC'],\n",
       "  ['chrX', '147587259', '152697442', 'ANCESTRAL'],\n",
       "  ['chrX', '152697442', '152819278', 'AMPLICONIC'],\n",
       "  ['chrX', '152819278', '153160033', 'ANCESTRAL'],\n",
       "  ['chrX', '153160033', '153382193', 'AMPLICONIC'],\n",
       "  ['chrX', '153382193', '153598022', 'ANCESTRAL'],\n",
       "  ['chrX', '153598022', '153606654', 'AMPLICONIC'],\n",
       "  ['chrX', '153606654', '153874094', 'ANCESTRAL'],\n",
       "  ['chrX', '153874094', '153882712', 'AMPLICONIC'],\n",
       "  ['chrX', '153882712', '154708000', 'ANCESTRAL'],\n",
       "  ['chrX', '154708000', '154811000', 'AMPLICONIC'],\n",
       "  ['chrX', '154811000', '158406701', 'ANCESTRAL'],\n",
       "  ['chrX', '158406701', '158460357', 'AMPLICONIC'],\n",
       "  ['chrX', '158460357', '161139705', 'ANCESTRAL'],\n",
       "  ['chrX', '161139705', '161614788', 'AMPLICONIC'],\n",
       "  ['chrX', '161614788', '164351005', 'ANCESTRAL'],\n",
       "  ['chrX', '164351005', '164457000', 'AMPLICONIC'],\n",
       "  ['chrX', '164457000', '164735256', 'ANCESTRAL'],\n",
       "  ['chrX', '164735256', '164821876', 'AMPLICONIC'],\n",
       "  ['chrX', '164821876', '164922559', 'ANCESTRAL'],\n",
       "  ['chrX', '164922559', '164969098', 'AMPLICONIC'],\n",
       "  ['chrX', '164969098', '165870000', 'ANCESTRAL'],\n",
       "  ['chrX', '165870000', '166006957', 'AMPLICONIC'],\n",
       "  ['chrX', '166006957', '166163654', 'ANCESTRAL'],\n",
       "  ['chrX', '166163654', '166250679', 'AMPLICONIC'],\n",
       "  ['chrX', '166250679', '166938133', 'ANCESTRAL'],\n",
       "  ['chrX', '166938133', '167098440', 'AMPLICONIC'],\n",
       "  ['chrX', '167098440', '168006544', 'ANCESTRAL'],\n",
       "  ['chrX', '168006544', '168261911', 'SATELLITE'],\n",
       "  ['chrX', '168261911', '168432720', 'ANCESTRAL'],\n",
       "  ['chrX', '168432720', '170048973', 'SATELLITE'],\n",
       "  ['chrX', '170048973', '170162781', 'ANCESTRAL'],\n",
       "  ['chrX', '170162781', '171024070', 'SATELLITE'],\n",
       "  ['chrX', '171024070', '171301162', 'ANCESTRAL'],\n",
       "  ['chrX', '171301162', '171990627', 'SATELLITE'],\n",
       "  ['chrX', '171990627', '172337054', 'ANCESTRAL'],\n",
       "  ['chrX', '172337054', '173231981', 'SATELLITE'],\n",
       "  ['chrX', '173231981', '173360224', 'ANCESTRAL'],\n",
       "  ['chrX', '173360224', '173845610', 'SATELLITE'],\n",
       "  ['chrX', '173845610', '174120509', 'ANCESTRAL'],\n",
       "  ['chrX', '174120509', '175592984', 'SATELLITE'],\n",
       "  ['chrX', '175592984', '175891520', 'ANCESTRAL'],\n",
       "  ['chrX', '175891520', '176746093', 'SATELLITE'],\n",
       "  ['chrX', '176746093', '177553137', 'ANCESTRAL']],\n",
       " 'PonPyg': [['chrX', '0', '48679', 'ANCESTRAL'],\n",
       "  ['chrX', '48679', '2356735', 'PAR'],\n",
       "  ['chrX', '2356735', '3204000', 'ANCESTRAL'],\n",
       "  ['chrX', '3204000', '4075000', 'AMPLICONIC'],\n",
       "  ['chrX', '4075000', '9790000', 'ANCESTRAL'],\n",
       "  ['chrX', '9790000', '9993000', 'AMPLICONIC'],\n",
       "  ['chrX', '9993000', '48907305', 'ANCESTRAL'],\n",
       "  ['chrX', '48907305', '49312475', 'AMPLICONIC'],\n",
       "  ['chrX', '49312475', '52285894', 'ANCESTRAL'],\n",
       "  ['chrX', '52285894', '52355699', 'AMPLICONIC'],\n",
       "  ['chrX', '52355699', '52674143', 'ANCESTRAL'],\n",
       "  ['chrX', '52674143', '52822474', 'AMPLICONIC'],\n",
       "  ['chrX', '52822474', '53050031', 'ANCESTRAL'],\n",
       "  ['chrX', '53050031', '53343364', 'AMPLICONIC'],\n",
       "  ['chrX', '53343364', '53436057', 'ANCESTRAL'],\n",
       "  ['chrX', '53436057', '53497662', 'AMPLICONIC'],\n",
       "  ['chrX', '53497662', '56116598', 'ANCESTRAL'],\n",
       "  ['chrX', '56116598', '56184629', 'AMPLICONIC'],\n",
       "  ['chrX', '56184629', '58605191', 'ANCESTRAL'],\n",
       "  ['chrX', '58605191', '66367560', 'SATELLITE'],\n",
       "  ['chrX', '66367560', '74978313', 'ANCESTRAL'],\n",
       "  ['chrX', '74978313', '75048113', 'AMPLICONIC'],\n",
       "  ['chrX', '75048113', '75974282', 'ANCESTRAL'],\n",
       "  ['chrX', '75974282', '76369354', 'AMPLICONIC'],\n",
       "  ['chrX', '76369354', '106403487', 'ANCESTRAL'],\n",
       "  ['chrX', '106403487', '106613478', 'AMPLICONIC'],\n",
       "  ['chrX', '106613478', '108202265', 'ANCESTRAL'],\n",
       "  ['chrX', '108202265', '108285968', 'AMPLICONIC'],\n",
       "  ['chrX', '108285968', '110538632', 'ANCESTRAL'],\n",
       "  ['chrX', '110538632', '110569197', 'AMPLICONIC'],\n",
       "  ['chrX', '110569197', '120056000', 'ANCESTRAL'],\n",
       "  ['chrX', '120056000', '120245000', 'AMPLICONIC'],\n",
       "  ['chrX', '120245000', '139731154', 'ANCESTRAL'],\n",
       "  ['chrX', '139731154', '139885288', 'AMPLICONIC'],\n",
       "  ['chrX', '139885288', '145817329', 'ANCESTRAL'],\n",
       "  ['chrX', '145817329', '145826208', 'AMPLICONIC'],\n",
       "  ['chrX', '145826208', '146258402', 'ANCESTRAL'],\n",
       "  ['chrX', '146258402', '146399000', 'AMPLICONIC'],\n",
       "  ['chrX', '146399000', '146942000', 'ANCESTRAL'],\n",
       "  ['chrX', '146942000', '147041000', 'AMPLICONIC'],\n",
       "  ['chrX', '147041000', '149898000', 'ANCESTRAL'],\n",
       "  ['chrX', '149898000', '150069000', 'AMPLICONIC'],\n",
       "  ['chrX', '150069000', '154432000', 'ANCESTRAL'],\n",
       "  ['chrX', '154432000', '154779927', 'AMPLICONIC'],\n",
       "  ['chrX', '154779927', '157716837', 'ANCESTRAL'],\n",
       "  ['chrX', '157716837', '157831245', 'AMPLICONIC'],\n",
       "  ['chrX', '157831245', '158126652', 'ANCESTRAL'],\n",
       "  ['chrX', '158126652', '158219819', 'AMPLICONIC'],\n",
       "  ['chrX', '158219819', '158325403', 'ANCESTRAL'],\n",
       "  ['chrX', '158325403', '158378525', 'AMPLICONIC'],\n",
       "  ['chrX', '158378525', '159357913', 'ANCESTRAL'],\n",
       "  ['chrX', '159357913', '159420441', 'AMPLICONIC'],\n",
       "  ['chrX', '159420441', '160406558', 'ANCESTRAL'],\n",
       "  ['chrX', '160406558', '160568409', 'AMPLICONIC'],\n",
       "  ['chrX', '160568409', '161100531', 'ANCESTRAL']],\n",
       " 'PonAbe': [['chrX', '0', '2382228', 'PAR'],\n",
       "  ['chrX', '2382228', '3232000', 'ANCESTRAL'],\n",
       "  ['chrX', '3232000', '3707000', 'AMPLICONIC'],\n",
       "  ['chrX', '3707000', '9384000', 'ANCESTRAL'],\n",
       "  ['chrX', '9384000', '9593000', 'AMPLICONIC'],\n",
       "  ['chrX', '9593000', '48522000', 'ANCESTRAL'],\n",
       "  ['chrX', '48522000', '48928044', 'AMPLICONIC'],\n",
       "  ['chrX', '48928044', '51907039', 'ANCESTRAL'],\n",
       "  ['chrX', '51907039', '51976814', 'AMPLICONIC'],\n",
       "  ['chrX', '51976814', '52294174', 'ANCESTRAL'],\n",
       "  ['chrX', '52294174', '52448670', 'AMPLICONIC'],\n",
       "  ['chrX', '52448670', '52680177', 'ANCESTRAL'],\n",
       "  ['chrX', '52680177', '53130538', 'AMPLICONIC'],\n",
       "  ['chrX', '53130538', '55758983', 'ANCESTRAL'],\n",
       "  ['chrX', '55758983', '55826998', 'AMPLICONIC'],\n",
       "  ['chrX', '55826998', '57106000', 'ANCESTRAL'],\n",
       "  ['chrX', '57106000', '57241000', 'AMPLICONIC'],\n",
       "  ['chrX', '57241000', '58372707', 'ANCESTRAL'],\n",
       "  ['chrX', '58372707', '67812153', 'SATELLITE'],\n",
       "  ['chrX', '67812153', '76522522', 'ANCESTRAL'],\n",
       "  ['chrX', '76522522', '76594015', 'AMPLICONIC'],\n",
       "  ['chrX', '76594015', '77524422', 'ANCESTRAL'],\n",
       "  ['chrX', '77524422', '77920893', 'AMPLICONIC'],\n",
       "  ['chrX', '77920893', '100504000', 'ANCESTRAL'],\n",
       "  ['chrX', '100504000', '100601000', 'AMPLICONIC'],\n",
       "  ['chrX', '100601000', '107933081', 'ANCESTRAL'],\n",
       "  ['chrX', '107933081', '108143033', 'AMPLICONIC'],\n",
       "  ['chrX', '108143033', '109732686', 'ANCESTRAL'],\n",
       "  ['chrX', '109732686', '109816410', 'AMPLICONIC'],\n",
       "  ['chrX', '109816410', '112076869', 'ANCESTRAL'],\n",
       "  ['chrX', '112076869', '112107464', 'AMPLICONIC'],\n",
       "  ['chrX', '112107464', '121564000', 'ANCESTRAL'],\n",
       "  ['chrX', '121564000', '121731000', 'AMPLICONIC'],\n",
       "  ['chrX', '121731000', '141177250', 'ANCESTRAL'],\n",
       "  ['chrX', '141177250', '141335458', 'AMPLICONIC'],\n",
       "  ['chrX', '141335458', '147727460', 'ANCESTRAL'],\n",
       "  ['chrX', '147727460', '147859000', 'AMPLICONIC'],\n",
       "  ['chrX', '147859000', '148134199', 'ANCESTRAL'],\n",
       "  ['chrX', '148134199', '148147211', 'AMPLICONIC'],\n",
       "  ['chrX', '148147211', '148403013', 'ANCESTRAL'],\n",
       "  ['chrX', '148403013', '148505000', 'AMPLICONIC'],\n",
       "  ['chrX', '148505000', '151376000', 'ANCESTRAL'],\n",
       "  ['chrX', '151376000', '151509000', 'AMPLICONIC'],\n",
       "  ['chrX', '151509000', '155912000', 'ANCESTRAL'],\n",
       "  ['chrX', '155912000', '156266970', 'AMPLICONIC'],\n",
       "  ['chrX', '156266970', '159199145', 'ANCESTRAL'],\n",
       "  ['chrX', '159199145', '159313722', 'AMPLICONIC'],\n",
       "  ['chrX', '159313722', '159604286', 'ANCESTRAL'],\n",
       "  ['chrX', '159604286', '159686360', 'AMPLICONIC'],\n",
       "  ['chrX', '159686360', '159816052', 'ANCESTRAL'],\n",
       "  ['chrX', '159816052', '159857290', 'AMPLICONIC'],\n",
       "  ['chrX', '159857290', '160837393', 'ANCESTRAL'],\n",
       "  ['chrX', '160837393', '160899969', 'AMPLICONIC'],\n",
       "  ['chrX', '160899969', '161887567', 'ANCESTRAL'],\n",
       "  ['chrX', '161887567', '162049411', 'AMPLICONIC'],\n",
       "  ['chrX', '162049411', '162586350', 'ANCESTRAL']],\n",
       " 'SymSyn': [['chrX', '0', '28000', 'ANCESTRAL'],\n",
       "  ['chrX', '28000', '289000', 'AMPLICONIC'],\n",
       "  ['chrX', '289000', '295261', 'ANCESTRAL'],\n",
       "  ['chrX', '295261', '1605026', 'SATELLITE'],\n",
       "  ['chrX', '1605026', '1606000', 'ANCESTRAL'],\n",
       "  ['chrX', '1606000', '1733000', 'AMPLICONIC'],\n",
       "  ['chrX', '1733000', '1740303', 'ANCESTRAL'],\n",
       "  ['chrX', '1740303', '2153317', 'SATELLITE'],\n",
       "  ['chrX', '2153317', '2249913', 'ANCESTRAL'],\n",
       "  ['chrX', '2249913', '3833478', 'SATELLITE'],\n",
       "  ['chrX', '3833478', '4185000', 'AMPLICONIC'],\n",
       "  ['chrX', '4185000', '4209204', 'ANCESTRAL'],\n",
       "  ['chrX', '4209204', '6596398', 'SATELLITE'],\n",
       "  ['chrX', '6596398', '7015000', 'AMPLICONIC'],\n",
       "  ['chrX', '7015000', '7021360', 'ANCESTRAL'],\n",
       "  ['chrX', '7021360', '8804147', 'SATELLITE'],\n",
       "  ['chrX', '8804147', '8957000', 'AMPLICONIC'],\n",
       "  ['chrX', '8957000', '8962608', 'ANCESTRAL'],\n",
       "  ['chrX', '8962608', '10711970', 'SATELLITE'],\n",
       "  ['chrX', '10711970', '11011326', 'ANCESTRAL'],\n",
       "  ['chrX', '11011326', '14128719', 'SATELLITE'],\n",
       "  ['chrX', '14128719', '14547000', 'AMPLICONIC'],\n",
       "  ['chrX', '14547000', '14549068', 'ANCESTRAL'],\n",
       "  ['chrX', '14549068', '16743352', 'PAR'],\n",
       "  ['chrX', '16743352', '16954836', 'ANCESTRAL'],\n",
       "  ['chrX', '16954836', '17040265', 'AMPLICONIC'],\n",
       "  ['chrX', '17040265', '17924000', 'ANCESTRAL'],\n",
       "  ['chrX', '17924000', '18035000', 'AMPLICONIC'],\n",
       "  ['chrX', '18035000', '61883661', 'ANCESTRAL'],\n",
       "  ['chrX', '61883661', '61945119', 'AMPLICONIC'],\n",
       "  ['chrX', '61945119', '65071925', 'ANCESTRAL'],\n",
       "  ['chrX', '65071925', '65123532', 'AMPLICONIC'],\n",
       "  ['chrX', '65123532', '65441039', 'ANCESTRAL'],\n",
       "  ['chrX', '65441039', '65477003', 'AMPLICONIC'],\n",
       "  ['chrX', '65477003', '65572127', 'ANCESTRAL'],\n",
       "  ['chrX', '65572127', '65607390', 'AMPLICONIC'],\n",
       "  ['chrX', '65607390', '65820314', 'ANCESTRAL'],\n",
       "  ['chrX', '65820314', '66034067', 'AMPLICONIC'],\n",
       "  ['chrX', '66034067', '68476354', 'ANCESTRAL'],\n",
       "  ['chrX', '68476354', '68635000', 'AMPLICONIC'],\n",
       "  ['chrX', '68635000', '68801801', 'ANCESTRAL'],\n",
       "  ['chrX', '68801801', '68860904', 'AMPLICONIC'],\n",
       "  ['chrX', '68860904', '70550338', 'ANCESTRAL'],\n",
       "  ['chrX', '70550338', '71318140', 'SATELLITE'],\n",
       "  ['chrX', '71318140', '79453052', 'ANCESTRAL'],\n",
       "  ['chrX', '79453052', '79514526', 'AMPLICONIC'],\n",
       "  ['chrX', '79514526', '80461477', 'ANCESTRAL'],\n",
       "  ['chrX', '80461477', '80780100', 'AMPLICONIC'],\n",
       "  ['chrX', '80780100', '85176790', 'ANCESTRAL'],\n",
       "  ['chrX', '85176790', '85238942', 'AMPLICONIC'],\n",
       "  ['chrX', '85238942', '106461087', 'ANCESTRAL'],\n",
       "  ['chrX', '106461087', '106655628', 'AMPLICONIC'],\n",
       "  ['chrX', '106655628', '110232465', 'ANCESTRAL'],\n",
       "  ['chrX', '110232465', '110268374', 'AMPLICONIC'],\n",
       "  ['chrX', '110268374', '119318000', 'ANCESTRAL'],\n",
       "  ['chrX', '119318000', '119587000', 'AMPLICONIC'],\n",
       "  ['chrX', '119587000', '138100000', 'ANCESTRAL'],\n",
       "  ['chrX', '138100000', '138259088', 'AMPLICONIC'],\n",
       "  ['chrX', '138259088', '138586809', 'ANCESTRAL'],\n",
       "  ['chrX', '138586809', '138787581', 'AMPLICONIC'],\n",
       "  ['chrX', '138787581', '152097226', 'ANCESTRAL'],\n",
       "  ['chrX', '152097226', '152502917', 'AMPLICONIC'],\n",
       "  ['chrX', '152502917', '155167114', 'ANCESTRAL'],\n",
       "  ['chrX', '155167114', '155244178', 'AMPLICONIC'],\n",
       "  ['chrX', '155244178', '155550319', 'ANCESTRAL'],\n",
       "  ['chrX', '155550319', '155692034', 'AMPLICONIC'],\n",
       "  ['chrX', '155692034', '156669084', 'ANCESTRAL'],\n",
       "  ['chrX', '156669084', '156726450', 'AMPLICONIC'],\n",
       "  ['chrX', '156726450', '157587253', 'ANCESTRAL'],\n",
       "  ['chrX', '157587253', '157737429', 'AMPLICONIC'],\n",
       "  ['chrX', '157737429', '158186000', 'ANCESTRAL'],\n",
       "  ['chrX', '158186000', '158289000', 'AMPLICONIC'],\n",
       "  ['chrX', '158289000', '158299652', 'ANCESTRAL'],\n",
       "  ['chrX', '158299652', '159014435', 'SATELLITE'],\n",
       "  ['chrX', '159014435', '159016000', 'ANCESTRAL'],\n",
       "  ['chrX', '159016000', '159305000', 'AMPLICONIC'],\n",
       "  ['chrX', '159305000', '159307480', 'ANCESTRAL'],\n",
       "  ['chrX', '159307480', '159658286', 'SATELLITE'],\n",
       "  ['chrX', '159658286', '159664000', 'ANCESTRAL'],\n",
       "  ['chrX', '159664000', '159790874', 'AMPLICONIC'],\n",
       "  ['chrX', '159790874', '160782074', 'SATELLITE'],\n",
       "  ['chrX', '160782074', '160788000', 'ANCESTRAL'],\n",
       "  ['chrX', '160788000', '160911000', 'AMPLICONIC'],\n",
       "  ['chrX', '160911000', '160912839', 'ANCESTRAL'],\n",
       "  ['chrX', '160912839', '162891368', 'SATELLITE'],\n",
       "  ['chrX', '162891368', '162898000', 'ANCESTRAL'],\n",
       "  ['chrX', '162898000', '163019000', 'AMPLICONIC'],\n",
       "  ['chrX', '163019000', '163022139', 'ANCESTRAL'],\n",
       "  ['chrX', '163022139', '165297312', 'SATELLITE'],\n",
       "  ['chrX', '165297312', '165304000', 'ANCESTRAL'],\n",
       "  ['chrX', '165304000', '165563000', 'AMPLICONIC'],\n",
       "  ['chrX', '165563000', '165588543', 'ANCESTRAL']],\n",
       " 'MacFas': []}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_class_dicts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extracts the protein sequences based on the protein file of the reference genome -> extract the protein ID name and its sequence\n",
    "prot_dicts = {}\n",
    "\n",
    "for species in species_list:\n",
    "\n",
    "    sequence_file = [d for d in data if d['species'] == species][0]['data']['prot']\n",
    "    sequence_handle = open(sequence_file)\n",
    "    seq_dict = SeqIO.to_dict(SeqIO.parse(sequence_handle, \"fasta\"))\n",
    "\n",
    "    prot_dicts[species] = seq_dict\n",
    "\n",
    "    sequence_handle.close()\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#collect protein sequence for each gene -> match the protein ID in the record_dict dictionary (contains information about genes) to the protein ID and its sequence. It then appends the sequence\n",
    "for record in record_dict:\n",
    "    \n",
    "    for element in record_dict[record]:\n",
    "        species = element[\"specie\"]\n",
    "        id = element[\"id\"]\n",
    "        name = element[\"name\"]\n",
    "        # for prot in prot_dicts[specie]:\n",
    "        #     print(prot)\n",
    "        #     print(prot_dicts[specie][prot])\n",
    "        #     assert False\n",
    "        prot_sequences = [prot_dicts[species][x] for x in prot_dicts[species] if (id in x )]\n",
    "        if len(prot_sequences) == 0:\n",
    "            print(\"No sequence found\")\n",
    "            print(id)\n",
    "            assert False\n",
    "        seq = str(prot_sequences[0].seq)\n",
    "        if len(seq) > 1:\n",
    "            for prot in prot_sequences:\n",
    "                if len(prot.seq) > len(seq):\n",
    "                    seq = str(prot.seq)\n",
    "\n",
    "        element[\"seq\"] = seq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#reduce list of prot sequences to longest sequence per gene\n",
    "gene_record_dict = {}\n",
    "for record in record_dict:\n",
    "    gene_spec_ids = set()\n",
    "    gene_spec_elements = []\n",
    "    \n",
    "    for element in record_dict[record]:\n",
    "        # gene_id = element[\"gene_id\"]\n",
    "        gene_spec_id = element[\"gene_spec_id\"]\n",
    "        \n",
    "        if gene_spec_id not in gene_spec_ids:\n",
    "            gene_spec_elements.append(element)\n",
    "        else:\n",
    "            if len(element[\"seq\"]) > len([x for x in gene_spec_elements if x[\"gene_spec_id\"] == gene_spec_id][0][\"seq\"]):\n",
    "                for el in gene_spec_elements:\n",
    "                    if el[\"gene_spec_id\"] == gene_spec_id:\n",
    "                        el[\"seq\"] = element[\"seq\"]\n",
    "                        break\n",
    "\n",
    "        gene_spec_ids.add(element[\"gene_spec_id\"])\n",
    "    \n",
    "    gene_record_dict[record] = gene_spec_elements\n",
    "\n",
    "#pprint(gene_record_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print one protein file per specie\n",
    "cmd = f\"mkdir -p {work_dir}/protein_extracted_longest\"\n",
    "subprocess.run(cmd, shell=True)\n",
    "for species in species_list:\n",
    "    file = f\"{work_dir}/protein_extracted_longest/{species}.faa\"\n",
    "\n",
    "\n",
    "    species_records = {}\n",
    "    for gene in gene_record_dict:\n",
    "        species_records[gene] = []\n",
    "        for element in gene_record_dict[gene]:\n",
    "            if element[\"specie\"] == species:\n",
    "                species_records[gene].append(element)\n",
    "\n",
    "\n",
    "    with open(file, \"w\") as outfile:\n",
    "        for entry in species_records:\n",
    "            element = species_records[entry]\n",
    "            if len(element) > 1:\n",
    "                print(species)\n",
    "                assert False\n",
    "            if len(element) == 0:\n",
    "                continue\n",
    "            element = element[0]\n",
    "            outfile.write(f\">{element['id']} gene:{element['gene_id']} mrna:{element['mrna_id']} {element['specie']} {element['name']}\\n\")\n",
    "            outfile.write(f\"{element['seq']}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='cat /home/emma/Amplicons/Workspaces/emma/downloaded_data/work_dir/x_multicopy/protein_extracted_longest/*.faa     > /home/emma/Amplicons/Workspaces/emma/downloaded_data/work_dir/x_multicopy/protein_extracted_longest/blastdb/all_proteins.faa', returncode=0)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Create a BLAST database from the protein files generate above. \n",
    "cmd = f\"mkdir -p {work_dir}/protein_extracted_longest/blastdb\" #create the directory\n",
    "subprocess.run(cmd, shell=True, check=True) #execute the command\n",
    "\n",
    "#concatanate all fasta files into one\n",
    "cmd = f\"cat {work_dir}/protein_extracted_longest/*.faa \\\n",
    "    > {work_dir}/protein_extracted_longest/blastdb/all_proteins.faa\" #matches all the .faa files (protein files for all species in the protein_extracted_longest foder\n",
    "# run command \n",
    "subprocess.run(cmd, shell=True, check=True) #execute the command. A concanated FASTA file is created -> contains all protein sequences from the .faa files created earlier, with their respective headers. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 05/26/2025 15:47:44\n",
      "New DB name:   /home/emma/Amplicons/Workspaces/emma/downloaded_data/work_dir/x_multicopy/protein_extracted_longest/blastdb/all_proteins\n",
      "New DB title:  /home/emma/Amplicons/Workspaces/emma/downloaded_data/work_dir/x_multicopy/protein_extracted_longest/blastdb/all_proteins.faa\n",
      "Sequence type: Protein\n",
      "Deleted existing Protein BLAST database named /home/emma/Amplicons/Workspaces/emma/downloaded_data/work_dir/x_multicopy/protein_extracted_longest/blastdb/all_proteins\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 7620 sequences in 0.14383 seconds.\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='makeblastdb -in /home/emma/Amplicons/Workspaces/emma/downloaded_data/work_dir/x_multicopy/protein_extracted_longest/blastdb/all_proteins.faa      -dbtype prot     -out /home/emma/Amplicons/Workspaces/emma/downloaded_data/work_dir/x_multicopy/protein_extracted_longest/blastdb/all_proteins', returncode=0)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#use the makeblastbd command to create a BLAST database from the concatenated FASTA file of protein sequences. \n",
    "cmd = f\"makeblastdb -in {work_dir}/protein_extracted_longest/blastdb/all_proteins.faa \\\n",
    "     -dbtype prot\\\n",
    "     -out {work_dir}/protein_extracted_longest/blastdb/all_proteins\"\n",
    "\n",
    "subprocess.run(cmd, shell=True, check=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Run a BLAST search for each species' protein sequences against the BLAST database created. \n",
    "# blast all against DB\n",
    "for s in species_list:\n",
    "    cmd = f\"blastp -query {work_dir}/protein_extracted_longest/{s}.faa \\\n",
    "    -db {work_dir}/protein_extracted_longest/blastdb/all_proteins \\\n",
    "    -out {work_dir}/protein_extracted_longest/{s}.blastp.tsv \\\n",
    "    -outfmt \\\"6 qseqid sseqid pident mismatch gapopen gaps qcovs qcovshsp evalue\\\" \"\n",
    "    subprocess.run(cmd, shell=True, check=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#collect edges for each protein\n",
    "edges = {}\n",
    "identity = 50\n",
    "coverage = 35\n",
    "score = 0.001 \n",
    "\n",
    "for s in species_list:\n",
    "    file = f\"{work_dir}/protein_extracted_longest/{s}.blastp.tsv\"\n",
    "    with open(file, \"r\") as infile:\n",
    "        for line in infile:\n",
    "            line = line.strip().split(\"\\t\")\n",
    "            if float(line[2]) >= identity and int(line[6]) >= coverage and float(line[7]) < score:\n",
    "                if line[0] in edges:\n",
    "                    edges[line[0]].append(line[1])\n",
    "                else:\n",
    "                    edges[line[0]] = [line[1]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#only keep two way edges\n",
    "edges_2way = []\n",
    "vertices = set()\n",
    "for node_A in edges:\n",
    "    for node_B in edges[node_A]:\n",
    "        if node_B in edges and node_A in edges[node_B]:\n",
    "            tuple = (node_A, node_B)\n",
    "            tuple = sorted(tuple)\n",
    "\n",
    "            if tuple in edges_2way:\n",
    "                continue\n",
    "\n",
    "            edges_2way.append(tuple)\n",
    "            vertices.add(node_A)\n",
    "            vertices.add(node_B)     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create clusters with transitive clusering\n",
    "clusters = []\n",
    "for edge in edges_2way:\n",
    "    found = False\n",
    "    for cluster in clusters:\n",
    "        if edge[0] in cluster or edge[1] in cluster:\n",
    "            cluster.add(edge[0])\n",
    "            cluster.add(edge[1])\n",
    "            found = True\n",
    "            break\n",
    "    if not found:\n",
    "        clusters.append(set(edge))\n",
    "\n",
    "\n",
    "#merge clusters with common elements\n",
    "merged_clusters = []\n",
    "for cluster in clusters:\n",
    "    found = False\n",
    "    for merged_cluster in merged_clusters:\n",
    "        if len(cluster.intersection(merged_cluster)) > 0:\n",
    "            merged_cluster.update(cluster)\n",
    "            found = True\n",
    "            break\n",
    "    if not found:\n",
    "        merged_clusters.append(cluster)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Link every protein to a their original species\n",
    "prot_to_species = {}\n",
    "\n",
    "for cluster in merged_clusters:\n",
    "\n",
    "    for prot in cluster:\n",
    "        # print(prot)\n",
    "        species = \"\"\n",
    "        for gene in gene_record_dict:\n",
    "            for element in gene_record_dict[gene]:\n",
    "                if element[\"id\"] == prot:\n",
    "                    species = element[\"specie\"]\n",
    "                    prot_to_species[prot] = species\n",
    "                        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#read blast results into pandas dataframe\n",
    "# columns :species, gene_A, gene_B, identity, coverage, evalue\n",
    "\n",
    "data_frames = {}\n",
    "for species in species_list:\n",
    "    blast_results = pd.DataFrame(columns=[\"gene_A\", \"gene_B\", \"identity\", \"coverage\", \"evalue\"]) \n",
    "    blast_results_list = []\n",
    "    file = f\"{work_dir}/protein_extracted_longest/{species}.blastp.tsv\"\n",
    "    with open(file, \"r\") as infile:\n",
    "        for line in infile:\n",
    "            line = line.strip().split(\"\\t\")\n",
    "            # if float(line[2]) >= identity and int(line[6]) >= coverage and float(line[7]) < score:\n",
    "            blast_results_list.append({\"species\": species, \"gene_A\": line[0], \"gene_B\": line[1], \"identity\": line[2], \"coverage\": line[6], \"evalue\": line[7]})\n",
    "\n",
    "    blast_results = pd.DataFrame(blast_results_list)\n",
    "    data_frames[species] = blast_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "palindrome_locations_file_path = \"./PalindromeLocations.tsv\" #this file is distributed along the code\n",
    "species_trans = {\n",
    "    \"HomSap_x\": \"chm13\"    ,\n",
    "    \"HomSap_y\": \"hg002\"    ,\n",
    "    \"GorGor\": \"mGorGor1\" ,\n",
    "    \"PanTro\": \"mPanTro3\" ,\n",
    "    \"PanPan\": \"mPanPan1\" ,\n",
    "    \"PonAbe\": \"mPonAbe1\" ,\n",
    "    \"PonPyg\": \"mPonPyg2\" ,\n",
    "    \"SymSyn\": \"mSymSyn1\" \n",
    "}\n",
    "palindrome_locations_file = open(palindrome_locations_file_path, \"r\")\n",
    "palindrome_locations = palindrome_locations_file.readlines()\n",
    "palindrome_locations_file.close()\n",
    "chr = \"x\"\n",
    "\n",
    "palindrome_lines_per_species = {}\n",
    "\n",
    "for species in species_list:\n",
    "    # Build the list of translation keys we actually have (HomSap is divided into X and Y so this should be taken into account):\n",
    "    if species == \"HomSap\":\n",
    "        raw_keys = [f\"{species}_x\", f\"{species}_y\"]\n",
    "    else:\n",
    "        raw_keys = [species]\n",
    "    # Only keep keys that exist in species_trans. Check if species exists in species_trans -> BECAUSE MacFas does not have palindrome information so this should be empty\n",
    "\n",
    "    keys = [k for k in raw_keys if k in species_trans]\n",
    "\n",
    "    # If none of those keys are in species_trans, just give an empty list and skip\n",
    "    if not keys:\n",
    "        palindrome_lines_per_species[species] = []\n",
    "        continue\n",
    "        \n",
    "    palindrome_lines = []\n",
    "        \n",
    "    for row in palindrome_locations:\n",
    "        row = row.strip().split(\"\\t\")\n",
    "        if len(row) < 4:\n",
    "            continue\n",
    "        if row[1] != '' and row[1] != 'start':\n",
    "                    \n",
    "            if species == \"HomSap\":\n",
    "                sp = f\"{species}_{chr}\"\n",
    "                species_chr = f\"{species_trans[sp]}.chr{chr.upper()}\"\n",
    "            else:\n",
    "                species_chr = f\"{species_trans[species]}.chr{chr.upper()}\"\n",
    "\n",
    "            if row[0] == species_chr:\n",
    "                palindrome_lines.append([x.replace(',','') for x in row])\n",
    "    palindrome_lines_per_species[species] = palindrome_lines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#palindrome_lines_per_species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a directory as a shell command\n",
    "cmd = f\"mkdir -p {work_dir}/protein_extracted_longest/clusters_merged\"\n",
    "subprocess.run(cmd, shell=True, check=True)\n",
    "\n",
    "file = open(f\"{work_dir}/protein_extracted_longest/clusters_merged/clusters.tsv\", \"w\")\n",
    "file_identities = open(f\"{work_dir}/protein_extracted_longest/clusters_merged/clusters_identities.tsv\", \"w\")\n",
    "file_classes = open(f\"{work_dir}/protein_extracted_longest/clusters_merged/clusters_classes.tsv\", \"w\")\n",
    " \n",
    "header = \"id\\tgene_symbol\\tDescription(s)\"\n",
    "for species in species_list:\n",
    "    header += f\"\\t{species}\"\n",
    "\n",
    "print(header, file=file)\n",
    "print(header, file=file_identities)\n",
    "print(header, file=file_classes)\n",
    "\n",
    "counter = 0\n",
    "prot_to_cluster = {}\n",
    "list_of_all_identities = []\n",
    "\n",
    "for cluster in merged_clusters:\n",
    "\n",
    "    cluster_copy = copy.deepcopy(cluster)\n",
    "    counter += 1\n",
    "    species_count = {}\n",
    "    gene_names = set()\n",
    "    gene_symbol = \"\"\n",
    "    report = False\n",
    "    gene_identity_counter_per_species = {}\n",
    "\n",
    "    #class counter intialization -> so initialize counters for sequences classes (PAR, ampliconic, ancestral, palindrome), per species\n",
    "    classes_per_species = {}\n",
    "    for species in species_list:\n",
    "        classes_per_species[species] = {\n",
    "                \"par_counter\" : 0,\n",
    "                \"ampl_counter\" : 0,\n",
    "                \"ancestral_counter\" : 0,\n",
    "                \"palindrome_counter\" : 0\n",
    "        }\n",
    "    \n",
    "    # create cluster FASTA file \n",
    "    # write sequences for each cluster into a seperate FASTA file for further analysis\n",
    "    # then match proteins to their repsective genes using gene_record_dict and extract relevant information (species, description, sequence) \n",
    "    # HERE THERE SHOULD BE START AND END coordinates??\n",
    "    cluster_fasta_file = f\"{work_dir}/protein_extracted_longest/clusters_merged/cluster_{counter}.faa\"\n",
    "    with open(cluster_fasta_file, \"w\") as cluster_outfile:\n",
    "        for prot in cluster:\n",
    "            # print(prot)\n",
    "            species = \"\"\n",
    "            desc = \"\"\n",
    "            for gene in gene_record_dict:\n",
    "                for element in gene_record_dict[gene]:\n",
    "                    if element[\"id\"] == prot:\n",
    "                        #print(\"Element dictionary:\", element)\n",
    "                        gene_names.add(element[\"name\"])\n",
    "                        species = element[\"specie\"]\n",
    "                        # print(element)\n",
    "                        desc = element[\"name\"]\n",
    "                        report_gene = gene\n",
    "\n",
    "                        prot_to_cluster[prot] = counter\n",
    "                        \n",
    "                        if species in species_count:\n",
    "                            report = True\n",
    "                            # species = element\n",
    "                            species_count[species] += 1\n",
    "                        else:\n",
    "                            species_count[species] = 1\n",
    "                        sequence = element[\"seq\"]\n",
    "\n",
    "                        el_start =  element[\"start\"]\n",
    "                        el_end =  element[\"end\"]\n",
    "                        \n",
    "                        #Determine which class (PAR, ampliconic, ancestral) the sequence belongs to, based on genomic coordinates. SO HERE IT USES COORDINATES?!\n",
    "                        for seq_class in seq_class_dicts[species]:\n",
    "                            class_name = seq_class[3]\n",
    "                            if el_start >= int(seq_class[1]) and el_end <= int(seq_class[2]):\n",
    "\n",
    "                                if seq_class[3] == \"PAR\":\n",
    "                                    classes_per_species[species][\"par_counter\"] += 1\n",
    "                                elif seq_class[3] == \"AMPLICONIC\":\n",
    "                                    classes_per_species[species][\"ampl_counter\"] += 1\n",
    "                                elif seq_class[3] == \"ANCESTRAL\" :\n",
    "                                    classes_per_species[species][\"ancestral_counter\"] += 1\n",
    "                                break\n",
    "                            else:\n",
    "                                #TODO if more than 50% of the sequence is in the class, count it as in the class\n",
    "                                \n",
    "                                if el_start < int(seq_class[1]) and el_end > int(seq_class[1]):\n",
    "                                    if int(seq_class[1]) - el_start > 0.5 * len(sequence):\n",
    "                                        if seq_class[3] == \"PAR\":\n",
    "                                            classes_per_species[species][\"par_counter\"] += 1\n",
    "                                        elif seq_class[3] == \"AMPLICONIC\":\n",
    "                                            classes_per_species[species][\"ampl_counter\"] += 1\n",
    "                                        elif seq_class[3] == \"ANCESTRAL\":\n",
    "                                            classes_per_species[species][\"ancestral_counter\"] += 1\n",
    "                                        break\n",
    "                                if el_start < int(seq_class[2]) and el_end > int(seq_class[2]):\n",
    "                                    if el_end - int(seq_class[2]) > 0.5 * len(sequence):\n",
    "                                        if seq_class[3] == \"PAR\":\n",
    "                                            classes_per_species[species][\"par_counter\"] += 1\n",
    "                                        elif seq_class[3] == \"AMPLICONIC\":\n",
    "                                            classes_per_species[species][\"ampl_counter\"] += 1\n",
    "                                        elif seq_class[3] == \"ANCESTRAL\":\n",
    "                                            classes_per_species[species][\"ancestral_counter\"] += 1\n",
    "                                        break\n",
    "\n",
    "                        # check if the sequence is within palindrome regions and count accordingly \n",
    "                        for palindrome in palindrome_lines_per_species[species]:\n",
    "                            if el_start >= int(palindrome[1]) and el_end <= int(palindrome[2]):\n",
    "                                if counter == 21:\n",
    "                                    print(palindrome)\n",
    "                                classes_per_species[species][\"palindrome_counter\"] += 1\n",
    "                                break\n",
    "                            else:\n",
    "                                if el_start < int(palindrome[1]) and el_end > int(palindrome[2]):\n",
    "                                    \n",
    "                                    print(f\"start: {el_start} end: {el_end} palindrome: {palindrome[0]} start: {palindrome[1]} end: {palindrome[2]}\")\n",
    "                                    assert False\n",
    "                                if el_start < int(palindrome[1]) and el_end > int(palindrome[2]):\n",
    "                                    print(f\"start: {el_start} end: {el_end} palindrome: {palindrome[0]} start: {palindrome[1]} end: {palindrome[2]}\")\n",
    "                                    assert False\n",
    "\n",
    "                        if element[\"specie\"] == \"HomSap\":\n",
    "                            if gene.startswith(\"LOC\"):\n",
    "                                continue\n",
    "                            gene_symbol = gene\n",
    "                        break\n",
    "            \n",
    "            print(f\">{prot}_gene:{report_gene}_species:{species} description:{desc}\\n{sequence}\", file=cluster_outfile)   \n",
    "            for prot_B in cluster_copy:\n",
    "                if prot == prot_B:\n",
    "                    continue\n",
    "                species = prot_to_species[prot]\n",
    "                if species != prot_to_species[prot_B]:\n",
    "                    continue\n",
    "                # blast results initiated in later cells, TODO: move to top\n",
    "                blast_results = data_frames[species]\n",
    "                blast_result = blast_results[(blast_results[\"gene_A\"] == prot) & (blast_results[\"gene_B\"] == prot_B)]\n",
    "                if blast_result.empty:\n",
    "                    continue\n",
    "\n",
    "                identity1 = blast_result.iloc[0][\"identity\"]\n",
    "\n",
    "                blast_result = blast_results[(blast_results[\"gene_A\"] == prot_B) & (blast_results[\"gene_B\"] == prot)]\n",
    "                if blast_result.empty:\n",
    "                    continue\n",
    "                identity2 = blast_result.iloc[0][\"identity\"]\n",
    "\n",
    "                if identity2 < identity1:\n",
    "                    identity1 = identity2\n",
    "                if species not in gene_identity_counter_per_species:\n",
    "                    gene_identity_counter_per_species[species] = f\"{identity1}\"\n",
    "                else:\n",
    "                    gene_identity_counter_per_species[species] += f\";{identity1}\"\n",
    "                list_of_all_identities.append(float(identity1))\n",
    "            cluster_copy.remove(prot)\n",
    "\n",
    "# once processing is complete, write the collected information (clusters, identities, classes) into respective file           \n",
    "    if report:\n",
    "        \n",
    "        if gene_symbol == \"\":\n",
    "            gene_symbol = list(gene_names)[0]\n",
    "        line = f\"{counter}\\t{gene_symbol}\\t{';'.join(gene_names)}\"\n",
    "        for species in species_list:\n",
    "            if species not in species_count:\n",
    "                species_count[species] = 0\n",
    "            line += f\"\\t{species_count[species]}\"\n",
    "\n",
    "        print(line, file=file)\n",
    "\n",
    "        identity_line = f\"{counter}\\t{gene_symbol}\\t{';'.join(gene_names)}\"\n",
    "        for species in species_list:\n",
    "            if species not in gene_identity_counter_per_species:\n",
    "                gene_identity_counter_per_species[species] = 0\n",
    "            identity_line += f\"\\t{gene_identity_counter_per_species[species]}\"\n",
    "        print(identity_line, file=file_identities)\n",
    "\n",
    "        class_line = f\"{counter}\\t{gene_symbol}\\t{';'.join(gene_names)}\"\n",
    "        \n",
    "        per_line_class_counter = {\n",
    "            \"par_counter\" : 0 ,\n",
    "            \"ampl_counter\" : 0,\n",
    "            \"ancestral_counter\" : 0,\n",
    "            \"palindrome_counter\" : 0\n",
    "        }\n",
    "        for species in species_list:\n",
    "            par_counter = classes_per_species[species][\"par_counter\"]\n",
    "            per_line_class_counter[\"par_counter\"] += par_counter\n",
    "            ampl_counter = classes_per_species[species][\"ampl_counter\"]\n",
    "            per_line_class_counter[\"ampl_counter\"] += ampl_counter\n",
    "            ancestral_counter = classes_per_species[species][\"ancestral_counter\"]\n",
    "            per_line_class_counter[\"ancestral_counter\"] += ancestral_counter\n",
    "            palindrome_counter = classes_per_species[species][\"palindrome_counter\"]\n",
    "            per_line_class_counter[\"palindrome_counter\"] += palindrome_counter\n",
    "            class_line += f\"\\t{ampl_counter} ({palindrome_counter}) + {ancestral_counter} + {par_counter}\"\n",
    "\n",
    "\n",
    "        class_line += f\"\\t{per_line_class_counter['ampl_counter']} ({per_line_class_counter['palindrome_counter']}) + {per_line_class_counter['ancestral_counter']} + {per_line_class_counter['par_counter']}\"\n",
    "        print(class_line, file=file_classes)\n",
    "\n",
    "#close files\n",
    "file.close()\n",
    "file_identities.close()\n",
    "file_classes.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene details have been saved to /home/emma/Amplicons/Workspaces/emma/downloaded_data/work_dir/x_multicopy/protein_extracted_longest/clusters_merged/gene_details.tsv\n"
     ]
    }
   ],
   "source": [
    "## Extract the genes coordinates and which classification they belong to. \n",
    "output_dir = f\"{work_dir}/protein_extracted_longest/clusters_merged\"\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "gene_details_file_path = os.path.join(output_dir, \"gene_details.tsv\")\n",
    "\n",
    "with open(gene_details_file_path, \"w\") as gene_details_file:\n",
    "    # Write header to the file\n",
    "    print(\"Species\\tGene\\tGene_symbol\\tStart\\tEnd\\tStrand\\tClass\", file=gene_details_file)\n",
    "\n",
    "    # Iterate over each species and their corresponding genes\n",
    "    for species in species_list:\n",
    "        for gene in gene_record_dict:\n",
    "            for element in gene_record_dict[gene]:\n",
    "                if element[\"specie\"] == species:\n",
    "                    gene_name = element[\"name\"]\n",
    "                    gene_symbol = gene\n",
    "                    start = element[\"start\"]\n",
    "                    end = element[\"end\"]\n",
    "                    sequence_class = \"Unknown\"\n",
    "                    strand = element[\"strand\"]\n",
    "\n",
    "                    # Determine the sequence class\n",
    "                    for seq_class in seq_class_dicts[species]:\n",
    "                        if start >= int(seq_class[1]) and end <= int(seq_class[2]):\n",
    "                            sequence_class = seq_class[3]\n",
    "                            break\n",
    "\n",
    "                    # Print gene details to the file\n",
    "                    print(f\"{species}\\t{gene_name}\\t{gene_symbol}\\t{start}\\t{end}\\t{strand}\\t{sequence_class}\", file=gene_details_file)\n",
    "\n",
    "print(f\"Gene details have been saved to {gene_details_file_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract only the ampliconic gene families \n",
    "# Defined as sequence identity for at least two copies have ≥97% sequence identity within at least one species.\n",
    "\n",
    "# Load the Cluster identities TSV file\n",
    "df = pd.read_csv(f\"{work_dir}/protein_extracted_longest/clusters_merged/clusters_identities.tsv\", sep='\\t')\n",
    "\n",
    "# Check if at least 2 copies have >= 97% identity\n",
    "def is_ampliconic(sequence_identities):\n",
    "    # Split the string by semicolons and convert to floats\n",
    "    identities = [float(identity) for identity in sequence_identities.split(';') if identity]\n",
    "    \n",
    "    # Check if at least one species has 2 or more identities >= 97%\n",
    "    return sum([identity >= 97.0 for identity in identities]) >= 1\n",
    "\n",
    "# Filter the rows based on the condition for PanTro, HomSap, and PanPan\n",
    "filtered_rows = []\n",
    "\n",
    "for idx, row in df.iterrows():\n",
    "    if is_ampliconic(row['PanTro']) or is_ampliconic(row['HomSap']) or is_ampliconic(row['PanPan']) or is_ampliconic(row['GorGor']) or is_ampliconic(row['PonPyg']) or is_ampliconic(row['PonAbe']) or is_ampliconic(row['SymSyn']):\n",
    "        filtered_rows.append(row)\n",
    "\n",
    "# Create a new dataframe with only the ampliconic families\n",
    "ampliconic_df = pd.DataFrame(filtered_rows)\n",
    "\n",
    "# Save the filtered dataframe to a new TSV file\n",
    "ampliconic_df.to_csv(f\"{work_dir}/protein_extracted_longest/clusters_merged/ampliconic_families.tsv\", sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updated gene_details.tsv with gene_family_symbol column.\n"
     ]
    }
   ],
   "source": [
    "# append the gene family to the genes of the gene_details.tsv file \n",
    "# Load the gene_details and clusters files\n",
    "gene_details_file = f\"{work_dir}/protein_extracted_longest/clusters_merged/gene_details.tsv\"\n",
    "clusters_file = f\"{work_dir}/protein_extracted_longest/clusters_merged/ampliconic_families.tsv\"\n",
    "\n",
    "gene_details_df = pd.read_csv(gene_details_file, sep='\\t')\n",
    "clusters_df = pd.read_csv(clusters_file, sep='\\t')\n",
    "\n",
    "gene_details_df[\"gene_family_symbol\"] = \"\"\n",
    "\n",
    "# Iterate through each row in gene_details_df\n",
    "for index, row in gene_details_df.iterrows():\n",
    "    gene_name = row[\"Gene\"]\n",
    "    \n",
    "    # Search for the gene_name in the \"Description(s)\" column of clusters_df\n",
    "    matching_row = clusters_df[clusters_df[\"Description(s)\"].str.contains(gene_name, na=False, regex=False)]\n",
    "    \n",
    "    if not matching_row.empty:\n",
    "        # Get the gene_symbol from the matching row\n",
    "        gene_symbol = matching_row.iloc[0][\"gene_symbol\"]\n",
    "        # Append the gene_symbol to the \"gene_family_symbol\" column in gene_details_df\n",
    "        gene_details_df.at[index, \"gene_family_symbol\"] = gene_symbol\n",
    "\n",
    "# Save the updated DataFrame back to a new file\n",
    "gene_details_df.to_csv(f\"{work_dir}/protein_extracted_longest/clusters_merged/gene_details_updated.tsv\", sep='\\t', index=False)\n",
    "\n",
    "print(\"Updated gene_details.tsv with gene_family_symbol column.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAsEAAAIlCAYAAAA0ZCXgAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjAsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvlHJYcgAAAAlwSFlzAAAPYQAAD2EBqD+naQAAUQBJREFUeJzt3XuczHX///Hn7Gn2kF2stYdaLKGyIodECikqUvmWQi46oEgWpaR+uyqHdCVdTkV9IxJdlxQqp3LIJZFSTlFZi+y2S+xg1+zp8/vDd6fGnmbZnZn1edxvt7lds5/Pe2Zen3d73Tznve/P+20xDMMQAAAAYCI+ni4AAAAAcDdCMAAAAEyHEAwAAADTIQQDAADAdAjBAAAAMB1CMAAAAEyHEAwAAADTIQQDAADAdAjBAAAAMB1CMACPGzBggCwWiw4ePOjpUlCKN954Q1dffbWCgoJksVg0d+7cUttbLBZ17NjR5fdPSkqSxWLR+vXrL6rOirR+/XpZLBYlJSW5/JqDBw/KYrFowIABlVYXgItHCAZQIQr/4b/99ttLbLNly5YKDwcEDvdYuHChRo4cqeDgYI0YMUKJiYlq3ry5p8uqEOUN65JUr1491atXr1LqAeAefp4uAAAmTpyo5557TpdffrmnS0EJPv/8c0nSZ599pqioKJdes3fvXgUHB1dmWZXu+uuv1969e1WrVi2XX3P55Zdr7969CgsLq8TKAFwsQjAAj4uOjlZ0dLSny0Apjh49KkkuB2BJuuqqqyqrHLcJDg4u93X4+/tfEtcOXOqYDgHA40qaE7xkyRJ16NBBtWvXVmBgoGJjY3X77bfrk08+kSTNnTtXcXFxkqR58+bJYrE4Hn+fV5qVlaWkpCRdddVVCgwMVM2aNdWtWzdt3ry52HqOHTumQYMGqXbt2goODlbr1q21dOlSzZ07t8hc2L9Px/j555/Vs2dP1apVy+l6li5dqt69e+vKK69UcHCwwsLCdNNNN2nJkiVFPvvv77d37151795d1atXV40aNdS7d28dO3ZMkvTtt9/qtttuU2hoqGrUqKGBAwfqzJkz5er3FStWqFOnTgoLC1NQUJCaN2+uqVOnKj8/39Gm8JrXrVsnSY7+dWUqQEnTDA4fPqzevXurZs2auuyyy9ShQwdt3Lix1PfauHGj7rrrLtWqVUtWq1UNGzbUCy+8oKysLKd2f5/D+/3336tr166qVq2awsLCdO+99zr9jhW2laQNGzY4/f4U/jc+f05w4X+flJQUpaSkOL3m/DbFTdE5deqUEhMT1aRJEwUFBal69eq6/fbbtWnTpiJtU1NTNXz4cDVs2FBBQUGqWbOmmjZtqiFDhshms5XaXwDKxkgwAK80a9YsDRkyRNHR0br33nsVHh6u1NRUbd26VZ988onuueceNW/eXMOHD9ebb76pZs2a6Z577nG8vjCk2e12de7cWVu2bFGLFi2UkJCg9PR0LV68WKtXr9bixYvVs2dPx+tOnz6tDh06aM+ePWrfvr3at2+v33//Xb1791aXLl1KrPfXX3/VDTfcoCZNmqh///76888/FRAQIEkaM2aMAgIC1L59e0VHRysjI0PLli3Tfffdp3/9618aNmxYkfdLTk5Wu3bt1KpVKz322GP67rvvtGjRIh0+fFivvvqqbrvtNt12220aNGiQ1q9fr3feeUeSNGfOHJf6980331RCQoJq1qypPn36KCQkRMuXL9eIESP09ddf6z//+Y8sFouaN2+uxMREzZ07VykpKUpMTJQkVa9e3aXPOV9qaqratm2r33//XV27dlWLFi20d+9e3XbbberUqVOxr3nrrbc0ZMgQ1ahRQ3fddZciIiK0bds2jR8/XuvWrdO6descfV3ou+++02uvvaaOHTtq8ODB+uGHH/TJJ59o586d2rVrlwIDA1WvXj0lJiZq3Lhxqlu3rlNoLWm+c/Xq1ZWYmKipU6dKkhISEhznyppX/Oeff+rmm2/W7t27ddNNN6lr167KzMzUp59+qk6dOunf//6343c4KytLN954ow4ePKguXbro3nvvVU5Ojg4cOKC5c+dq9OjRCg0NLfXzAJTBAIAKkJycbEgyGjRoYCQmJhb7ePTRRw1JRv/+/Z1e279/f0OSkZyc7DjWokULIyAgwEhPTy/yWceOHSvyuee/Z6GXXnrJkGT07dvXKCgocBz/8ccfDavVatSoUcOw2WyO4y+88IIhyRg6dKjT+6xbt86QZEgy3nvvvSKfL8l48cUXi63ht99+K3Ls1KlTRtOmTY2wsDDjzJkzxb7f1KlTHccLCgqMO++805BkVK9e3fjkk08c53Jycoxrr73W8Pf3N9LS0oqt4fx6/Pz8jNq1axuHDh1yHLfb7UaHDh0MScb8+fOdXlN4vDwkGR06dHA6Vvjf+pVXXnE6/vbbbzuue926dY7ju3fvNvz8/IzrrrvOOH78uNNrJk6caEgy/vnPfzqO/f2/06JFi5za9+vXz5BkfPjhh2XWef77JSYmOh2vW7euUbdu3WJfU9LvZJ8+fQxJxv/+7/86HU9LSzNiY2ONiIgIIzs72zAMw1i2bJkhyRgxYkSR97fZbIbdbi/2swG4jhAMoEL8PbyV9XA1BIeEhBgnTpxw6XNLCsH169c3/P39jcOHDxc5N3jw4CKBr169eobVai02fHft2rXEEBwVFVXuYPL6668bkoz169cXeb/69esb+fn5Tu3ff/99Q5LRqVOnIu9VGPb/HiBLUtj21VdfLXLum2++MSQZnTt3djpeESHYbrcbgYGBRu3atR1hr1B+fr7RqFGjItfw1FNPGZKMr7/+usj75+fnGxEREUbLli0dxwpD680331ykfeG5kSNHllpnca+52BCckZFh+Pr6FunXQv/6178MScby5csNw/grBD///PPFtgdw8ZgOAaBCde3aVStXriz23JYtW9S2bVuX3qdXr1567rnnFB8frwcffFAdO3ZU+/bty/VneJvNpgMHDujqq6/WFVdcUeR8x44d9fbbb2vHjh166KGHZLPZdPDgQTVp0kQRERFF2rdr106rVq0q9rOaNWtW5E/yhdLT0zVp0iR98cUXSklJUXZ2ttP5wpvOzn8/Hx/n2zYKbx4s7k/1hed+//33Ymv4ux9++EFS8X++v+GGGxQUFKQdO3aU+T7ltW/fPp09e1a33HKLAgMDnc75+PioXbt22r9/v9PxLVu2SJJWrlyptWvXFnlPf39//fzzz0WOt2jRosixwt+BkydPXuglXLBt27YpPz9fZ8+eLXbN4V9++UWS9PPPP6t79+66+eabFRUVpYkTJ2rHjh3q1q2b2rdvr6ZNmzrmMQO4OIRgAF5p9OjRCg8P11tvvaUpU6bo9ddfl5+fn+68805NnTrVcUNcaQpvHoqMjCz2fOFKB5mZmU7tiwvApb1Paef+/PNPtW7dWocOHdKNN96oW2+9VdWrV5evr6927NihTz/9VHa7vcjripvv6efnV+a53NzcEmssVFa/1K5d26UwXV6F/Vy7du1izxdXz59//ilJGj9+fLk+q7jlyQr76O83/rlL4XX897//1X//+98S2xXe3BgWFqZvvvlGiYmJWr58uWOJuiuuuEJjxozRkCFDKr9o4BJHCAbglSwWix577DE99thjOn78uL7++mt9+OGH+uijj/TLL79o586d8vX1LfU9CsPiH3/8Uez5wuOF7Qr/NyMjo9T2JdVbnHfffVeHDh3SK6+8orFjxzqdmzRpkj799NNSrqBy/L1f6tatW+R8enp6pdx0VRhM09PTiz1fXP8W1mGz2VStWrUKr8ldCq9j1KhR+uc//+nSa+rVq6d58+YpPz9fO3fu1OrVq/Wvf/1LQ4cOdawWAuDCsUQaAK8XHh6ue+65R4sXL9Ytt9yivXv36tdff5UkRxAubnQvNDRU9evX16+//lrsyOaGDRsk/TW9IDQ0VPXq1dOvv/5abBAuaUm10vz222+SpB49ehQ59/XXX5f7/SrCddddJ0nFbk+8detWZWdnV8pucI0bN1ZgYKC+++47nT171ulcQUFBsf3bpk0bSX9Ni6gMPj4+5R4d9vX1LddrWrduLYvFom+++aa85cnX11fNmzfX6NGj9eGHH0qSli1bVu73AeCMEAzAK61atUp5eXlOx3Jzcx1/Vg4KCpIk1ahRQxaLRUeOHCn2ffr376/c3FyNGTNGhmE4ju/atUvvvfeewsLCnJZW69u3r+x2u8aNG+f0PuvXry9xPnBpCkdaz18HduHChY4/cbtbnz595OfnpylTpjjNR87NzdVzzz0nSZWyDXVAQIB69eql9PR0vf76607n3nnnnSLzgSVpyJAh8vPz07Bhw3T48OEi50+ePOmY43yhatasWeLvT2mvOXbsWJEwX5KoqCj16tVLmzdv1muvveb0u1jo22+/dax7vGvXLqWkpBRpUzhaXvj7D+DCMR0CgFd64IEHFBwcrPbt26tu3brKzc3VmjVrtGfPHj3wwAOqU6eOJOmyyy5T69attXHjRj388MNq2LChfHx81KdPH9WpU0ejR4/WZ599pvnz52vv3r3q3LmzMjIytHjxYuXm5ur99993+jP7s88+qyVLlmjGjBn66aef1L59ex05ckQfffSR7rrrLi1fvrzIDWul6devn1599VUNGzZM69atU926dfXTTz9p7dq16tmzpz7++OMK77uyNGjQQK+++qpGjRqla6+9Vr169VJISIhWrFihn3/+WXfffbceeuihSvnsSZMm6csvv9QLL7ygTZs26brrrtPevXv1+eefq0uXLlq9erVT+/j4eM2cOVNPPPGEGjdurDvvvFMNGjRw3PS4YcMGDRgwQG+99dYF13TLLbfoo48+0n333afrrrtOvr6+6tatm5o2bVrqa7777jvddddduummmxzrQLdv377E18ycOVP79u3T6NGjNX/+fLVt21ZhYWE6fPiwtm/frl9++UWpqakKDg7W2rVrNWrUKN1444266qqrFB4ergMHDmjZsmUKCgrSk08+ecHXC+AcQjAArzRx4kStXLlSW7du1fLlyxUSEqIrr7xSb7/9th555BGntvPnz9eIESP0ySefKDMzU4Zh6IYbblCdOnUUGBior776Sq+++qoWL16sN954Q8HBwbr55pv1/PPPFwkt1apV08aNGzVmzBh9+umn2rZtm5o0aaIPP/xQBw4c0PLly8s1X/aKK67Qhg0bNHr0aK1du1Z5eXlq0aKFVq9ercOHD3skBEvSyJEjdeWVV2rKlClasGCBcnJy1KhRI73++ut66qmnKm0FgujoaG3evFmjR4/WqlWrtHHjRrVs2VJr1qzRV199VSQES9LAgQPVvHlzTZkyRRs3btSyZcsUFhamOnXqaMSIEerfv/9F1fTmm29Kkr766istXbpUBQUFioqKKjUEv/jiizpx4oRWrFihr776SgUFBUpMTCw1BNesWVObN2/W9OnTtXjxYn3wwQeOz2rWrJlefPFF1apVS9K5VVYOHjyojRs36uOPP9bp06d1+eWX68EHH9To0aN19dVXX9Q1A5AsRnF/kwEAFPHQQw/pgw8+0J49ewghAFDFMScYAM6Tmppa5NiGDRu0aNEiNW7cmAAMAJcApkMAwHnuvPNOBQUFqXnz5goJCdGePXu0cuVK+fr6atq0aZ4uDwBQAZgOAQDnmTp1qj744AP99ttvOnXqlKpXr64bb7xRY8aMcSzZBQCo2gjBAAAAMB3mBAMAAMB0CMEAAAAwHUJwORiGIZvNVuxOPwAAAKg6CMHlcOrUKYWFhenUqVOeLgUAAAAXgSXSAAAATCw3N1ezZ8+WJA0aNEj+/v4ersg9WB2iHGw2m8LCwpSZmVmubVMBAAC81ZkzZ3TZZZdJkk6fPq2QkBAPV+QeTIcAAACA6RCCAQAAYDqEYAAAAJgOIRgAAACmQwgGAACA6RCCAQAAYDqsEwwAAGBiVqtVK1ascDw3C9YJLgfWCQYAALg0MB0CAAAApsN0CAAAABPLzc3VBx98IEnq27cv2yajKKZDAACASw3bJgMAAAAmQQgGAACA6RCCAQAAYDqEYAAAAJgOIRgAAACmQwgGAACA6bBOMAAAgIlZrVZ99NFHjudmwTrB5cA6wQAAAFJGRoZsNpvL7UNDQxUREVGJFZUfI8EAAABwWUZGhp4YOEj2rCyXX2MNDtasObO9KggTggEAAEwsLy9PS5culSTde++98vMrPR7abDbZs7I06sF+io2KLvP9D6el6vVF82Wz2QjBAAAA8A52u129evWSdG7b5LJCcKHYqGg1iK1TmaVVKlaHAAAAgOkQggEAAGA6hGAAAACYDiEYAAAApkMIBgAAgOkQggEAAGA6LJEGAABgYgEBAXrvvfccz82CEAwAAGBi/v7+GjBggKfLcDumQwAAAMB0GAkGAAAwsby8PK1atUqS1LVrV5d3jKvqzHGVAAAAKJbdblf37t0llW/b5KqO6RAAAAAwHUIwAAAATIcQDAAAANMhBAMAAMB0CMEAAAAwHUIwAAAATMcca2AAAACgWAEBAZo+fbrjuVkQggEAAEzM399fQ4cO9XQZbsd0CAAAAJgOI8EAAAAmlp+fr6+//lqSdNNNN8nX19fDFbkHIRgAAMDEzp49q06dOkk6t21ySEiIhytyD6ZDAAAAwHQIwQAAADAdQjAAAABMhxAMAAAA0yEEAwAAwHQIwQAAADAdlkgDAAAwMX9/f02ePNnx3CwIwQAAACYWEBCgZ555xtNluB3TIQAAAGA6jAQDAACYWH5+vr7//ntJUosWLdg2GQAAAJe+s2fP6vrrr5fEtskAAADAJY0QDAAAANMhBAMAAMB0CMEAAAAwHUIwAAAATIcQDAAAANNhiTQAAAAT8/f3V2JiouO5WRCCAQAATCwgIEBJSUmeLsPtmA4BAAAA0/F4CN64caPuuusuxcTEyGKx6JNPPnE6bxiGkpKSFBMTo6CgIHXs2FG7d+92amO32zVs2DDVqlVLISEh6tGjh44cOeLU5sSJE+rXr5/CwsIUFhamfv366eTJk5V8dQAAAN6toKBAu3fv1u7du1VQUODpctzG4yH4zJkzatasmaZPn17s+cmTJ2vKlCmaPn26tm3bpqioKN122206deqUo01CQoKWLl2qRYsWadOmTTp9+rS6d++u/Px8R5s+ffpox44dWrlypVauXKkdO3aoX79+lX59AAAA3iw7O1vx8fGKj49Xdna2p8txG4/PCb7jjjt0xx13FHvOMAxNnTpVY8eOVc+ePSVJ8+bNU2RkpBYuXKjBgwcrMzNT7777rubPn69bb71VkrRgwQLFxsZq7dq16tq1q/bu3auVK1dqy5YtatOmjSRpzpw5atu2rfbt26fGjRu752IBAADgFTw+Elya5ORkpaWlqUuXLo5jVqtVHTp00ObNmyVJ27dvV25urlObmJgYxcfHO9p88803CgsLcwRgSbrhhhsUFhbmaFMcu90um83m9AAAAEDV59UhOC0tTZIUGRnpdDwyMtJxLi0tTQEBAapRo0apbWrXrl3k/WvXru1oU5yJEyc65hCHhYUpNjb2oq4HAAAA3sGrQ3Ahi8Xi9LNhGEWOne/8NsW1L+t9xowZo8zMTMfj8OHD5awcAAAA3sirQ3BUVJQkFRmtTU9Pd4wOR0VFKScnRydOnCi1zR9//FHk/TMyMoqMMv+d1WpVaGio0wMAAABVn1eH4Li4OEVFRWnNmjWOYzk5OdqwYYPatWsnSWrZsqX8/f2d2qSmpmrXrl2ONm3btlVmZqa2bt3qaPPtt98qMzPT0QYAAADm4fHVIU6fPq1ff/3V8XNycrJ27NihmjVrqk6dOkpISNCECRPUsGFDNWzYUBMmTFBwcLD69OkjSQoLC9Ojjz6qUaNGKTw8XDVr1tTTTz+tpk2bOlaLuPrqq3X77bdr4MCBevvttyVJgwYNUvfu3VkZAgAAmJq/v7+efvppx3Oz8HgI/u6779SpUyfHzyNHjpQk9e/fX3PnztXo0aOVnZ2tIUOG6MSJE2rTpo1Wr16tatWqOV7zxhtvyM/PT7169VJ2drY6d+6suXPnytfX19Hmgw8+0FNPPeVYRaJHjx4lrk0MAABgFgEBAXrttdc8XYbbWQzDMDxdRFVhs9kUFhamzMxM5gcDAABT+u2335TwxBBNTRitBrF1ym5/+JASpk7W1Fkz1aBBAzdU6BqPjwQDAADAcwoKCnTo0CFJUp06deTj49W3jFUYQjAAAICJZWdnKy4uTtK5e7VCQkI8XJF7mCPqAwAAAH9DCAYAAIDpEIIBAABgOoRgAAAAmA4hGAAAAKZDCAYAAIDpsEQaAACAifn5+WnIkCGO52ZhnisFAABAEVarVTNmzPB0GW7HdAgAAACYDiPBAAAAJmYYho4dOyZJqlWrliwWi4crcg9CMAAAgIllZWWpdu3aktg2GQAAALikEYIBAABgOoRgAAAAmA4hGAAAAKZDCAYAAIDpEIIBAABgOiyRBgAAYGJ+fn7q37+/47lZmOdKAQAAUITVatXcuXM9XYbbMR0CAAAApsNIMAAAgIkZhqGsrCxJUnBwsGm2TWYkGAAAwMSysrJ02WWX6bLLLnOEYTMgBAMAAMB0CMEAAAAwHUIwAAAATIcQDAAAANMhBAMAAMB0CMEAAAAwHdYJBgAAMDFfX1/dd999judmQQgGAAAwscDAQP373//2dBlux3QIAAAAmA4hGAAAAKZDCAYAADCxM2fOyGKxyGKx6MyZM54ux20IwQAAADAdQjAAAABMhxAMAAAA0yEEAwAAwHQIwQAAADAdQjAAAABMhx3jAAAATMzX11d33nmn47lZEIIBAABMLDAwUJ999pmny3A7pkMAAADAdAjBAAAAMB1CMAAAgImdOXNGISEhCgkJMdW2ycwJBgAAMLmsrCxPl+B2jAQDAADAdAjBAAAAMB1CMAAAAEyHEAwAAADTIQQDAADAdFgdAgAAwMR8fHzUoUMHx3OzIAQDAACYWFBQkNavX+/pMtzOPHEfAAAA+D+EYAAAAJgOIRgAAMDEzpw5o4iICEVERLBtMgAAAMzj2LFjni7B7RgJBgAAgOkQggEAAGA6hGAAAACYDiEYAAAApkMIBgAAgOmwOgQAAICJ+fj4qFWrVo7nZkEIBgAAMLGgoCBt27bN02W4nXniPgAAAPB/CMEAAAAwHa8PwXl5eXrhhRcUFxenoKAg1a9fXy+99JIKCgocbQzDUFJSkmJiYhQUFKSOHTtq9+7dTu9jt9s1bNgw1apVSyEhIerRo4eOHDni7ssBAADwKllZWapXr57q1aunrKwsT5fjNl4fgl999VW99dZbmj59uvbu3avJkyfrtdde07Rp0xxtJk+erClTpmj69Onatm2boqKidNttt+nUqVOONgkJCVq6dKkWLVqkTZs26fTp0+revbvy8/M9cVkAAABewTAMpaSkKCUlRYZheLoct/H6G+O++eYb3X333erWrZskqV69evrwww/13XffSTr3H27q1KkaO3asevbsKUmaN2+eIiMjtXDhQg0ePFiZmZl69913NX/+fN16662SpAULFig2NlZr165V165dPXNxAAAA8AivHwlu3769vvzyS+3fv1+S9OOPP2rTpk268847JUnJyclKS0tTly5dHK+xWq3q0KGDNm/eLEnavn27cnNzndrExMQoPj7e0aY4drtdNpvN6QEAAICqz+tHgp999lllZmbqqquukq+vr/Lz8zV+/Hj17t1bkpSWliZJioyMdHpdZGSkUlJSHG0CAgJUo0aNIm0KX1+ciRMnaty4cRV5OQAAAPACXj8SvHjxYi1YsEALFy7U999/r3nz5umf//yn5s2b59TOYrE4/WwYRpFj5yurzZgxY5SZmel4HD58+MIvBAAAAF7D60eCn3nmGT333HN68MEHJUlNmzZVSkqKJk6cqP79+ysqKkrSudHe6Ohox+vS09Mdo8NRUVHKycnRiRMnnEaD09PT1a5duxI/22q1ymq1VsZlAQAAwIO8fiQ4KyuryBZ+vr6+jiXS4uLiFBUVpTVr1jjO5+TkaMOGDY6A27JlS/n7+zu1SU1N1a5du0oNwQAAAJc6i8Wia665Rtdcc02Zf0W/lHj9SPBdd92l8ePHq06dOmrSpIl++OEHTZkyRY888oikc//hEhISNGHCBDVs2FANGzbUhAkTFBwcrD59+kiSwsLC9Oijj2rUqFEKDw9XzZo19fTTT6tp06aO1SIAAADMKDg4uMj+Cmbg9SF42rRpevHFFzVkyBClp6crJiZGgwcP1v/7f//P0Wb06NHKzs7WkCFDdOLECbVp00arV69WtWrVHG3eeOMN+fn5qVevXsrOzlbnzp01d+5c+fr6euKyAAAA4EEWw0yrIl8km82msLAwZWZmKjQ01NPlAAAAuN1vv/2mhCeGaGrCaDWIrVN2+8OHlDB1sqbOmqkGDRq4oULXeP2cYAAAAFSerKwsNWnSRE2aNDHVtslePx0CAAAAlccwDO3Zs8fx3CwYCQYAAIDpEIIBAABgOoRgAAAAmA4hGAAAAKZDCAYAAIDpsDoEAACAiVksFtWtW9fx3CwIwQAAACYWHBysgwcPeroMt2M6BAAAAEyHEAwAAADTIQQDAACYWHZ2tlq3bq3WrVsrOzvb0+W4DXOCAQAATKygoEDfffed47lZMBIMAAAA0yEEAwAAwHQIwQAAADAdQjAAAABMhxAMAAAA02F1CAAAAJOrVauWp0twO0IwAACAiYWEhCgjI8PTZbgd0yEAAABgOoRgAAAAmA4hGAAAwMSys7PVsWNHdezYkW2TAQAAYA4FBQXasGGD47lZMBIMAAAA0yEEAwAAwHQIwQAAADAdQjAAAABMhxAMAAAA02F1CAAAAJMLDg72dAluRwgGAAAwsZCQEJ05c8bTZbgd0yEAAABgOoRgAAAAmA4hGAAAwMTOnj2rbt26qVu3bjp79qyny3Eb5gQDAACYWH5+vj7//HPHc7NgJBgAAACmQwgGAACA6RCCAQAAYDqEYAAAAJgOIRgAAACmQwgGAACA6bBEGgAAgImFhITIMAxPl+F2jAQDAADAdAjBAAAAMB1CMAAAgImdPXtW999/v+6//35TbZtMCAYAADCx/Px8/ec//9F//vMftk0GAAAALmWsDgGUU0ZGhmw2m8vtQ0NDFRERUYkVAQCA8iIEA+WQkZGhJwYOkj0ry+XXWIODNWvObIIwAABehBAMlIPNZpM9K0ujHuyn2KjoMtsfTkvV64vmy2azEYIBAPAihGDgAsRGRatBbB1PlwEAAC4QN8YBAADAdBgJBgAAMLHg4GCdPn3a8dwsLmgk2NfXV1u3bi323Pbt2+Xr63tRRQEAAMA9LBaLQkJCFBISIovF4uly3OaCQrBhGCWeKygoMFUHAgAAoOq54DnBJQXd7du3Kyws7IILAgAAgPvY7XYNGDBAAwYMkN1u93Q5buPynOA333xTb775pqRzAfiee+6R1Wp1apOdna309HTdd999FVslAAAAKkVeXp7mzZsnSZoxY0aRfHepcjkE165dW02aNJEkHTx4UPXr11f16tWd2litVjVt2lTDhw+v0CIBAACAiuRyCO7du7d69+4tSerUqZNmzZqlq666qtIKAwAAACrLBS2Rtm7duoquAwAAAHCbC14n2DAMbdu2TSkpKcrOzi5y/h//+MdFFQYAAABUlgsKwfv371ePHj30yy+/FLtcmsViIQQDAADAa11QCB46dKjOnj2rxYsX69prrzXNXYQAAAC4NFxQCN66davmzJnDUmgAAABVXHBwsNLT0x3PzeKCQvBll12m0NDQiq4FAAAAbmaxWBQREeHpMtzugnaMe/jhh7Vw4cKKrgUAAABwiwsaCY6Pj9eHH36oHj166K677lJ4eHiRNj179rzo4gAAAFC57Ha7Ro4cKUmaMmWKae71uqCR4D59+ig5OVkrVqzQ4MGDdd999zk97r///got8vfff9dDDz2k8PBwBQcHq3nz5tq+fbvjvGEYSkpKUkxMjIKCgtSxY0ft3r3b6T3sdruGDRumWrVqKSQkRD169NCRI0cqtE4AAICqJi8vTzNnztTMmTOVl5fn6XLcxus3yzhx4oRuvPFGderUSV988YVq166t3377zWnL5smTJ2vKlCmaO3euGjVqpFdeeUW33Xab9u3bp2rVqkmSEhIStHz5ci1atEjh4eEaNWqUunfvru3bt8vX19dt1wMAAADPu6AQ3KFDh4quo0SvvvqqYmNj9d577zmO1atXz/HcMAxNnTpVY8eOdUzBmDdvniIjI7Vw4UINHjxYmZmZevfddzV//nzdeuutkqQFCxYoNjZWa9euVdeuXd12PTAfe06OUlJSXGobGhpqypsTAABwtwveMc5dli1bpq5du+r+++/Xhg0bdPnll2vIkCEaOHCgJCk5OVlpaWnq0qWL4zVWq1UdOnTQ5s2bNXjwYG3fvl25ublObWJiYhQfH6/NmzeXGILtdrvsdrvjZ5vNVklXiUvV8cyTOnDggCaNe1lWa0CZ7a3BwZo1ZzZBGACASnZBIfiWW24p9bzFYtGXX355QQWd78CBA5o1a5ZGjhyp559/Xlu3btVTTz0lq9Wqf/zjH0pLS5MkRUZGOr0uMjLSMfqWlpamgIAA1ahRo0ibwtcXZ+LEiRo3blyFXAfM6XRWlgL8/DTigb5qVC+u1LaH01L1+qL5stlshGAAACrZBYXggoICWSwWp2PHjh3Tvn37VLt2bTVq1KhCiiv8rFatWmnChAmSpOuuu067d+/WrFmznLZmPr8ewzCKHDtfWW3GjBnjuFtSOjcSHBsbeyGXAZO7IjJKDWLreLoMAADwfy4oBK9fv77Y4/v379fdd9+txMTEi6nJSXR0tK655hqnY1dffbWWLFkiSYqKipJ0brQ3Ojra0SY9Pd0xOhwVFaWcnBydOHHCaTQ4PT1d7dq1K/GzrVaraZYJAQAAMJMLWiKtJI0aNdIzzzyj0aNHV9h73njjjdq3b5/Tsf3796tu3bqSpLi4OEVFRWnNmjWO8zk5OdqwYYMj4LZs2VL+/v5ObVJTU7Vr165SQzAAAMClLigoSMnJyUpOTlZQUJCny3GbCr8xrl69etq1a1eFvd+IESPUrl07TZgwQb169dLWrVs1e/ZszZ49W9K5aRAJCQmaMGGCGjZsqIYNG2rChAkKDg5Wnz59JElhYWF69NFHNWrUKIWHh6tmzZp6+umn1bRpU8dqEQAAAGbk4+PjtPKWWVR4CF6yZIliYmIq7P1at26tpUuXasyYMXrppZcUFxenqVOnqm/fvo42o0ePVnZ2toYMGaITJ06oTZs2Wr16tWONYEl644035Ofnp169eik7O1udO3fW3LlzWSMYAADAhC4oBD/yyCNFjtntdv3000/as2ePJk+efNGF/V337t3VvXv3Es9bLBYlJSUpKSmpxDaBgYGaNm2apk2bVqG1AQAAVGU5OTkaO3asJGn8+PEKCCh7Sc9LwQWF4K+++qrIqgqBgYGqV6+exowZ45iGAAAAAO+Wm5urf/7zn5KkpKQkQnBpDh48WMFlAAAAAO5ToatDAAAAAFXBBd8Y9+eff+qNN97Ql19+qePHj6tWrVq69dZblZCQUGRnNgAAAMCbXNBI8O+//64WLVpo/PjxyszMVJ06dXTy5Em9/PLLatGihY4ePVrRdQIAAAAV5oJC8PPPP6/s7Gx9++232r17t9asWaPdu3fr22+/VXZ2tp5//vmKrhMAAACoMBcUgleuXKlXXnlFrVu3djreunVrvfTSS/riiy8qpDgAAACgMlzQnODMzMwSdxaJi4tTZmbmxdQEAAAANwkKCnLs9mumbZMvaCQ4Li5On332WbHnvvjiC8XFxV1UUQAAAHAPHx8fNWnSRE2aNJGPj3kWDrugkeCHH35Yzz33nAoKCtS/f39FR0crNTVVCxYs0LRp0zRp0qSKrhMAAACoMBcUgp955hn99ttvmj59umbMmOE4bhiGBg0apKeffrrCCgQAAEDlycnJ0YQJEySdW/yAHeNKYbFY9Pbbb2vkyJFat26djh8/rvDwcN1yyy1q1KhRRdcIAACASpKbm6tx48ZJOjfQaZYQ7PLEjxMnTuh//ud/tGLFCsexxo0b6/HHH9fYsWP1+OOPa//+/fqf//kfHT9+vFKKBQAAACqCyyH4nXfe0Y8//qjbb7+9xDa33367du7c6TRFAgAAAPA2LofgRYsWaeDAgfLzK3kGhZ+fnwYOHKhly5ZVSHEAAABAZXA5BO/fv1+tWrUqs12LFi20f//+iyoKAAAAqEwuh+C8vDz5+/uX2c7f31+5ubkXVRQAAABQmVwOwdHR0dqzZ0+Z7Xbv3q2oqKiLKgoAAACoTC6H4A4dOmjmzJmljvLm5uZq1qxZ6tSpU4UUBwAAgMoVGBiorVu3auvWrQoMDPR0OW7jcggeMWKEfv75Z9177706evRokfNHjx7VPffco3379mnEiBEVWiQAAAAqh6+vr1q3bq3WrVvL19fX0+W4jcubZVx77bWaMWOGhgwZori4OLVs2VJxcXGSpOTkZG3fvl0FBQWaNWuWmjZtWmkFAwAAABerXDvGDRw4UPHx8ZowYYLWrVunLVu2SJKCg4N1++23a8yYMbrhhhsqpVAAAABUvJycHL355puSpOHDh5tmx7hyb5vctm1bLV++XAUFBTp27JgkqVatWvLxcXlmBQAAALxEbm6uRo8eLUkaMmQIIbgsPj4+ql27dkXWAgAAALgFw7cAAAAwHUIwAAAATIcQDAAAANMhBAMAAMB0CMEAAAAwnQteHQIAAABVX2BgoNatW+d4bhaEYAAAABPz9fVVx44dPV2G2zEdAgAAAKbDSDAAAICJ5ebmavbs2ZKkQYMGyd/f38MVuQchGAAAwMRycnL05JNPSpIGDBhgmhDMdAgAAACYDiEYAAAApkMIBgAAgOkwJxioojIyMmSz2Vxun5OTo4CAAJfahoaGKiIi4kJLAwDA6xGCgSooIyNDTwwcJHtWlkvt7Tk5Sk5J0ZVxcfLzK/v/9tbgYM2aM5sgDAC4ZBGCgSrIZrPJnpWlUQ/2U2xUdJntt/y0Q+Nnz9RT9/VWo3pxpbY9nJaq1xfNl81mIwQDAC5ZhGCgCouNilaD2DpltktJPSpJuiIyyqX2AADzsFqtWrFiheO5WRCCAS9iz8lRSkpKme1SUlKUl5fnhooAAJc6Pz8/devWzdNluB0hGPASxzNP6sCBA5o07mVZraXfwHYmK0t/pKXJnpvjpuoAALi0EIIBL3E6K0sBfn4a8UDfMuftFs7xzc/Ld1N1AIBLVW5urj744ANJUt++fU2zYxwhGPAyrszbLZzjCwDAxcrJydHDDz8sSbr//vtNE4LZLAMAAACmQwgGAACA6RCCAQAAYDqEYAAAAJgOIRgAAACmQwgGAACA6bBEGgAAgIlZrVZ99NFHjudmQQgGUISr2zdLUmhoqCIiIiq5IgBAZfHz89P999/v6TLcjhAMwEl5tm+WJGtwsGbNmU0QBgBUKYRgAE7Ks33z4bRUvb5ovmw2GyEYAKqovLw8LV26VJJ07733ys/PHPHQHFcJoNxc2b4ZAFD12e129erVS5J0+vRp04RgVocAAACA6RCCAQAAYDqEYAAAAJgOIRgAAACmQwgGAACA6RCCAQAAYDrmWAMDAAAAxQoICNB7773neG4WhGAAAAAT8/f314ABAzxdhtsxHQIAAACmw0gwAACAieXl5WnVqlWSpK5du5pmxzhzXCUAAACKZbfb1b17d0lsm+zVJk6cKIvFooSEBMcxwzCUlJSkmJgYBQUFqWPHjtq9e7fT6+x2u4YNG6ZatWopJCREPXr00JEjR9xcPQAAALxBlQrB27Zt0+zZs3Xttdc6HZ88ebKmTJmi6dOna9u2bYqKitJtt92mU6dOOdokJCRo6dKlWrRokTZt2qTTp0+re/fuys/Pd/dlAAAAwMOqTAg+ffq0+vbtqzlz5qhGjRqO44ZhaOrUqRo7dqx69uyp+Ph4zZs3T1lZWVq4cKEkKTMzU++++65ef/113Xrrrbruuuu0YMEC7dy5U2vXrvXUJQEAAMBDqkwIHjp0qLp166Zbb73V6XhycrLS0tLUpUsXxzGr1aoOHTpo8+bNkqTt27crNzfXqU1MTIzi4+MdbYpjt9tls9mcHgAAAKj6qsTM50WLFun777/Xtm3bipxLS0uTJEVGRjodj4yMVEpKiqNNQECA0whyYZvC1xdn4sSJGjdu3MWWDwAAAC/j9SPBhw8f1vDhw7VgwQIFBgaW2M5isTj9bBhGkWPnK6vNmDFjlJmZ6XgcPny4fMUDAADAK3n9SPD27duVnp6uli1bOo7l5+dr48aNmj59uvbt2yfp3GhvdHS0o016erpjdDgqKko5OTk6ceKE02hwenq62rVrV+JnW61WWa3Wir4kAAAArxEQEKDp06c7npuF148Ed+7cWTt37tSOHTscj1atWqlv377asWOH6tevr6ioKK1Zs8bxmpycHG3YsMERcFu2bCl/f3+nNqmpqdq1a1epIRgAAOBS5+/vr6FDh2ro0KHy9/f3dDlu4/UjwdWqVVN8fLzTsZCQEIWHhzuOJyQkaMKECWrYsKEaNmyoCRMmKDg4WH369JEkhYWF6dFHH9WoUaMUHh6umjVr6umnn1bTpk2L3GgHAACAS5/Xh2BXjB49WtnZ2RoyZIhOnDihNm3aaPXq1apWrZqjzRtvvCE/Pz/16tVL2dnZ6ty5s+bOnStfX18PVg4AAOBZ+fn5+vrrryVJN910k2myUZUMwevXr3f62WKxKCkpSUlJSSW+JjAwUNOmTdO0adMqtzgAAIAq5OzZs+rUqZOkc/syhISEeLgi9/D6OcEAAABARSMEAwAAwHQIwQAAADAdQjAAAABMhxAMAAAA0yEEAwAAwHSq5BJpAAAAqBj+/v6aPHmy47lZEIIBAABMLCAgQM8884yny3A7pkMAAADAdBgJBgAAMLH8/Hx9//33kqQWLVqwbTIAAAAufWfPntX1118viW2TAQAAgEsaIRgAAACmQwgGAACA6RCCAQAAYDqEYAAAAJgOq0MAcJuMjAzZbDaX24eGhioiIqISKwIAmBUhGIBbZGRk6ImBg2TPynL5NdbgYM2aM5sgDACVyN/fX4mJiY7nZkEIBnBR7Dk5SklJKbNdSkqKzthsevahhxUbFV1m+8NpqXp90XzZbDZCMABUooCAACUlJXm6DLcjBAO4YMczT+rAgQOaNO5lWa0BpbY9k5WlP9LSVDs8XA1i67ipQgAAikcIBnDBTmdlKcDPTyMe6KtG9eJKbbvlpx0aP3um8vPy3VQdAMAVBQUF2rt3ryTp6quvlo+POdZNIAQDuGhXREaVObqbknrUTdUAAMojOztb8fHxktg2GQAAALikEYIBAABgOoRgAAAAmA4hGAAAAKZDCAYAAIDpEIIBAABgOiyRBgAAYGL+/v56+umnHc/NghAMAABgYgEBAXrttdc8XYbbEYIBSRkZGbLZbGW2S0lJUV5enhsqAgAAlYkQDNPLyMjQEwMHyZ6VVWbbM1lZ+iMtTfbcHDdUBgBA5SsoKNChQ4ckSXXq1GHbZMAsbDab7FlZGvVgP8VGRZfadstPOzR+9kzl5+W7qToAACpXdna24uLiJJlr22RCMPB/YqOi1SC2TqltUlKPuqkaAABQmQjBALyWPSdHKSkpLrUNDQ1VREREJVcEALhUEIIBeKXjmSd14MABTRr3sqzWgDLbW4ODNWvObIIwAMAlhGAAXul0VpYC/Pw04oG+alQvrtS2h9NS9fqi+bLZbIRgAIBLCMEAvNoVkVFlztUGAKC8zLEGBgAAAPA3jAQDAACYmJ+fn4YMGeJ4bhbmuVIAAAAUYbVaNWPGDGVkZOjIkSNltr9Udk8lBAMAAJicGXdPJQQDAACYmGEYSk5Olu3kSY3p94jqRMeU2v5S2T2VEAwAAGBiWVlZatOmjSTp9adGmWb3VFaHAAAAgOkQggEAAGA6hGAAAACYDiEYAAAApkMIBgAAgOkQggEAAGA6hGAAAAAT8/PzU8+ePXVFdIx8fX09XY7bEIIBAABMzGq1avLkyWoeHy+rf4Cny3EbQjAAAABMhxAMAABgYoZhKCsrS3n5eTIMw9PluA0hGAAAwMSysrJ07bXXauVXXynbftbT5bgNIRgAAACmQwgGAACA6RCCAQAAYDqEYAAAAJgOIRgAAACmQwgGAACA6fh5ugAAQNkyMjJks9lcbh8aGqqIiIhKrAjApcLX11e33367fvz+B/n6mGfbZEIwAHi5jIwMPTFwkOxZWS6/xhocrFlzZhOEAZQpMDBQ06dPV8ITQ2QNMM+2yYRgAPByNptN9qwsjXqwn2KjostsfzgtVa8vmi+bzUYIBoASEIIBoIqIjYpWg9g6ni4DAC4J3BgHAABgYmfOnNGVV16pFWtWK+tstqfLcRtGggFcEuw5OUpJSXGpLTeNAQAIwQCqvOOZJ3XgwAFNGveyrNayb+rgpjEAgNeH4IkTJ+rjjz/Wzz//rKCgILVr106vvvqqGjdu7GhjGIbGjRun2bNn68SJE2rTpo1mzJihJk2aONrY7XY9/fTT+vDDD5Wdna3OnTtr5syZuuKKKzxxWQAq0OmsLAX4+WnEA33VqF5cqW25aQwAIFWBOcEbNmzQ0KFDtWXLFq1Zs0Z5eXnq0qWLzpw542gzefJkTZkyRdOnT9e2bdsUFRWl2267TadOnXK0SUhI0NKlS7Vo0SJt2rRJp0+fVvfu3ZWfn++JywJQCa6IjFKD2DqlPlxZXQEAcOnz+pHglStXOv383nvvqXbt2tq+fbtuvvlmGYahqVOnauzYserZs6ckad68eYqMjNTChQs1ePBgZWZm6t1339X8+fN16623SpIWLFig2NhYrV27Vl27dnX7dQHwHOYPAwC8PgSfLzMzU5JUs2ZNSVJycrLS0tLUpUsXRxur1aoOHTpo8+bNGjx4sLZv367c3FynNjExMYqPj9fmzZtLDMF2u112u93xc3l2a4LnubrDVkpKivLy8txQEbwB84cBAFIVC8GGYWjkyJFq37694uPjJUlpaWmSpMjISKe2kZGRjpGetLQ0BQQEqEaNGkXaFL6+OBMnTtS4ceMq8hLgJuXZYetMVpb+SEuTPTfHDZXB05g/DADOfH191bFjR+3ZtYttk73Vk08+qZ9++kmbNm0qcs5isTj9bBhGkWPnK6vNmDFjNHLkSMfPNptNsbGx5awanlCeHba2/LRD42fPVH4e88PNpHD+MACYXWBgoN555x22TfZWw4YN07Jly7Rx40anFR2ioqIknRvtjY7+K+ykp6c7RoejoqKUk5OjEydOOI0Gp6enq127diV+ptVqldVqrehLgRu5ssNWSupRN1UDAAC8hdevDmEYhp588kl9/PHH+uqrrxQX5/zny7i4OEVFRWnNmjWOYzk5OdqwYYMj4LZs2VL+/v5ObVJTU7Vr165SQzAAAAAuTV4/Ejx06FAtXLhQn376qapVq+aYwxsWFqagoCBZLBYlJCRowoQJatiwoRo2bKgJEyYoODhYffr0cbR99NFHNWrUKIWHh6tmzZp6+umn1bRpU8dqEQAAAGZ05swZNW3aVDl2uyY8PszT5biN14fgWbNmSZI6duzodPy9997TgAEDJEmjR49Wdna2hgwZ4tgsY/Xq1apWrZqj/RtvvCE/Pz/16tXLsVnG3Llz5etrngngALwLK5gA8BbZ2dmeLsHtvD4EG4ZRZhuLxaKkpCQlJSWV2CYwMFDTpk3TtGnTKrA6ALgwrGACVD2ufnGVWGO8KvD6EAwAlyJvWsGEf9iBspXni6vEGuNVASEYADzI0yuY8A874JryfHFljfGqgRAMACbGP+xA+bjyxRVVAyEYVQo3EgGVg3/YAZgNIRhVBjcSAQBQ8Xx8fHT99dfr119+kY/F67eQqDCEYFQZ3nQjEQAAl4qgoCAtXLhQCU8MUaCJdsolBKPK8fSNRAAAoOozz5g3AAAA8H8IwQAAACZ25swZtW7dWqvXr1PWWfPsHMd0CAAAAJM7ceKEp0twO0aCAQAAYDqEYAAAAJgOIRgAAACmQwgGAACA6XBjHABcguw5OUpJSSmzHVuMAzArQjAAlMLVMFkoNDRUERERlVhR2Y5nntSBAwc0adzLsloDSm3LFuMAfHx81LRpUx1KSWHbZABA+cJkIWtwsGbNme3RIHw6K0sBfn4a8UBfNaoXV2pbthgHEBQUpKVLl7JtMgDgnPKESUk6nJaq1xfNl81m8/hosCRdERnFFuMAUAJCMACUwZUwCQCoWswz8QMAAABFZGVlqUOHDvry643KPnvW0+W4DSPBAACXledGQW+4SRBA2QzD0O+//37uuQwPV+M+hGAAgEvKe6OgN9wkCFyKMjIyZLPZXGrLl9GSEYIBAC4pz42CF3KTIP+wA2XLyMjQEwMHyZ6V5VJ7voyWjBAMABXIDJtUVMaNgvzDDrjGZrPJnpWlUQ/2U2xUdKltvW3FGm9DCAaACsImFc7KM384JSVFZ2w2PfvQw/zDDrggNiqaVWsuEiEYACoIm1T8pbzzhwu/FNQOD3fpH3Zu0ANwsQjBAFDB2KSi/BuNlOdLQWXeoFeeeckSARvez5UvjNnZ2apbt66OZxyTRRY3VeZ5hGAAQKVxdf5web4UVNYNeuWdlywxNxnerTxfGOtfEasQP3/5+JpnCwlCMACgSqroG/TKc8ORxNxkeD+maJWOEAwAwN+U54Yj5iajKmCKVvEIwQAAXAA2D8GlIis7W4/8vzFKzUhXtp1tkwEAQCkqe/MQwF0Mw1DK0f/bNtk8uyYTggEAuBiVsXkIgMpHCAYAAMViK2tcygjBAACgCLayxqWOEAwAAIooz5JxzHlGVUQIBgCgiqvMaQvlWTIOqEoIwQAAVGFMW8DFslgsigyvpT8zT8pinl2TCcEAAFRl5Z22MOH9d7Vz507VrVu31LYpKSnKy8tzuY7ybBwicSOdNwkOCtLCV6foyfFJCrIGeroctyEEAwBwCXBl2kJ5Nvg4k5WlP9LSZM/NKfOzy7txiMSINDyPEAwAgEmUZ4OPLT/t0PjZM5Wfl1+h7ytxIx28AyEYAHBJc/XP9OX9839V5soGHympRyvlfS8E6xVXruyzZzXklUSlHD2qs3a7p8txG0IwAOCSVVl//of7cONf5SsoKNC+g8nnnpto32RCMADgklVZf/6H+5hhveLyjHSb6S8WlY0QDAC45FXWn//hPq6uV1yeVSq8YepEeUe6+YtFxSEEw6P49gvALKpaOKuKyrtKRWVOnSjPXPQzNpuefejhMke6Jf5iUZEIwfAYvv0CMAtvCmfeojJuWCzP9JfyrJlc3jouZC567fBwl0a6+YtFxSEEw2PKM89L4tsvgKqrvOGsKs5rLY/KvmHRlekv5f1iUp46mIteNRCC4XGuzvPi2y+Aqq6ylhCrarwhJJZ3beMLqaMqzUUPu6yaTmed8XQZbkUIBgAAHuENIdHVLybeElYrQ0hwsD6eOkNPjk9ScKB5tk328XQBAAAAgLsxEgwAgJcpz0oSrJwDXBhCMAAAXqQyb9gCipN99qxGTp6glFS2TQYAAB7ijhu2gL8rKCjQj/t/PvecbZMBAIAnccMWULm4MQ4AAACmQwgGAACA6RCCAQAAYDrMCUaFy8jIkM1mK7Mdy/oAAABPIQSjQmVkZOiJgYNkz8oqsy3L+gAA4B0CAwJkz831dBluRQhGhbLZbLJnZWnUg/0UGxVdaluW9QEAwPNCgoP12cx3TLdtMiEYlSI2Ktrj+8EDAACUhBvjAAAAYDqMBMMl3OwGAMCl6azdrufffF2H0lJlzzHPfTqEYJSJm90AALh05efn69udP557XlDg4WrchxCMMnGzGwAAuNQQguEybnYDAACXCtOF4JkzZ+q1115TamqqmjRpoqlTp+qmm27ydFklcnUuriSFhoYqIiKikisCAACo+kwVghcvXqyEhATNnDlTN954o95++23dcccd2rNnj+rUKX2E0xPKMxdXkqzBwZo1Z7bLQZib3QAAgFmZKgRPmTJFjz76qB577DFJ0tSpU7Vq1SrNmjVLEydO9HB1RZVnLu7htFRNeP9d7dy5U3Xr1i3zvY8fP65XksbJcCHccrMbAAC41JgmBOfk5Gj79u167rnnnI536dJFmzdvLvY1drtddrvd8XNmZqYkuTw94WKdOnVKuXl5OpOdrVNZZ0ptezTjD/3666965cVEBVgDynzvrKwspaen66ne/RQdUbvUtnt++0XvL/9Ee3/9VTllbKn42+FDyi8o0P6Dyco3Sr/DtDxtK/O9L/WavaUOajZXHdRsrjqouWrXkX32rOP5r4cOylpGjihvzb//8Ydy8/J06tQpt2UoSapWrZosFkvJDQyT+P333w1Jxn//+1+n4+PHjzcaNWpU7GsSExMNSTx48ODBgwcPHjyq2CMzM7PUbGiakeBC538jMAyjxG8JY8aM0ciRIx0/FxQU6M8//1R4eHjp3yxMwGazKTY2VocPH1ZoaKiny/Fq9JXr6CvX0Veuo69cR1+5jr5ynaf6qlq1aqWeN00IrlWrlnx9fZWWluZ0PD09XZGRkcW+xmq1ymq1Oh2rXr16ZZVYJYWGhvJ/fhfRV66jr1xHX7mOvnIdfeU6+sp13tZXPp4uwF0CAgLUsmVLrVmzxun4mjVr1K5dOw9VBQAAAE8wzUiwJI0cOVL9+vVTq1at1LZtW82ePVuHDh3S448/7unSAAAA4EamCsEPPPCAjh8/rpdeekmpqamKj4/X559/7tKSYnBmtVqVmJhYZLoIiqKvXEdfuY6+ch195Tr6ynX0leu8ta8shmEYni4CAAAAcCfTzAkGAAAAChGCAQAAYDqEYAAAAJgOIRgAAACmQwhGiWbNmqVrr73Wsbh127Zt9cUXXzjOG4ahpKQkxcTEKCgoSB07dtTu3bs9WLH3mDhxoiwWixISEhzH6K9zkpKSZLFYnB5RUVGO8/STs99//10PPfSQwsPDFRwcrObNm2v79u2O8/TXOfXq1Svye2WxWDR06FBJ9NPf5eXl6YUXXlBcXJyCgoJUv359vfTSSyooKHC0ob/+curUKSUkJKhu3boKCgpSu3bttG3bNsd5s/bVxo0bdddddykmJkYWi0WffPKJ03lX+sVut2vYsGGqVauWQkJC1KNHDx05csR9F1HqpsowtWXLlhmfffaZsW/fPmPfvn3G888/b/j7+xu7du0yDMMwJk2aZFSrVs1YsmSJsXPnTuOBBx4woqOjDZvN5uHKPWvr1q1GvXr1jGuvvdYYPny44zj9dU5iYqLRpEkTIzU11fFIT093nKef/vLnn38adevWNQYMGGB8++23RnJysrF27Vrj119/dbShv85JT093+p1as2aNIclYt26dYRj009+98sorRnh4uLFixQojOTnZ+Pe//21cdtllxtSpUx1t6K+/9OrVy7jmmmuMDRs2GL/88ouRmJhohIaGGkeOHDEMw7x99fnnnxtjx441lixZYkgyli5d6nTelX55/PHHjcsvv9xYs2aN8f333xudOnUymjVrZuTl5bnlGgjBKJcaNWoY77zzjlFQUGBERUUZkyZNcpw7e/asERYWZrz11lserNCzTp06ZTRs2NBYs2aN0aFDB0cIpr/+kpiYaDRr1qzYc/STs2effdZo3759iefpr5INHz7caNCggVFQUEA/nadbt27GI4884nSsZ8+exkMPPWQYBr9Xf5eVlWX4+voaK1ascDrerFkzY+zYsfTV/zk/BLvSLydPnjT8/f2NRYsWOdr8/vvvho+Pj7Fy5Uq31M10CLgkPz9fixYt0pkzZ9S2bVslJycrLS1NXbp0cbSxWq3q0KGDNm/e7MFKPWvo0KHq1q2bbr31Vqfj9JezX375RTExMYqLi9ODDz6oAwcOSKKfzrds2TK1atVK999/v2rXrq3rrrtOc+bMcZynv4qXk5OjBQsW6JFHHpHFYqGfztO+fXt9+eWX2r9/vyTpxx9/1KZNm3TnnXdK4vfq7/Ly8pSfn6/AwECn40FBQdq0aRN9VQJX+mX79u3Kzc11ahMTE6P4+Hi39R0hGKXauXOnLrvsMlmtVj3++ONaunSprrnmGqWlpUmSIiMjndpHRkY6zpnNokWL9P3332vixIlFztFff2nTpo3ef/99rVq1SnPmzFFaWpratWun48eP00/nOXDggGbNmqWGDRtq1apVevzxx/XUU0/p/fffl8TvVUk++eQTnTx5UgMGDJBEP53v2WefVe/evXXVVVfJ399f1113nRISEtS7d29J9NffVatWTW3bttXLL7+so0ePKj8/XwsWLNC3336r1NRU+qoErvRLWlqaAgICVKNGjRLbVDZTbZuM8mvcuLF27NihkydPasmSJerfv782bNjgOG+xWJzaG4ZR5JgZHD58WMOHD9fq1auLjBj8Hf0l3XHHHY7nTZs2Vdu2bdWgQQPNmzdPN9xwgyT6qVBBQYFatWqlCRMmSJKuu+467d69W7NmzdI//vEPRzv6y9m7776rO+64QzExMU7H6adzFi9erAULFmjhwoVq0qSJduzYoYSEBMXExKh///6OdvTXOfPnz9cjjzyiyy+/XL6+vmrRooX69Omj77//3tGGvirehfSLO/uOkWCUKiAgQFdeeaVatWqliRMnqlmzZnrzzTcdd/Of/20tPT29yDc/M9i+fbvS09PVsmVL+fn5yc/PTxs2bNC//vUv+fn5OfqE/ioqJCRETZs21S+//MLv1Xmio6N1zTXXOB27+uqrdejQIUmiv4qRkpKitWvX6rHHHnMco5+cPfPMM3ruuef04IMPqmnTpurXr59GjBjh+CsW/eWsQYMG2rBhg06fPq3Dhw9r69atys3NVVxcHH1VAlf6JSoqSjk5OTpx4kSJbSobIRjlYhiG7Ha74//8a9ascZzLycnRhg0b1K5dOw9W6BmdO3fWzp07tWPHDsejVatW6tu3r3bs2KH69evTXyWw2+3au3evoqOj+b06z4033qh9+/Y5Hdu/f7/q1q0rSfRXMd577z3Vrl1b3bp1cxyjn5xlZWXJx8f5n39fX1/HEmn0V/FCQkIUHR2tEydOaNWqVbr77rvpqxK40i8tW7aUv7+/U5vU1FTt2rXLfX3nltvvUCWNGTPG2Lhxo5GcnGz89NNPxvPPP2/4+PgYq1evNgzj3PInYWFhxscff2zs3LnT6N27tymWhXHV31eHMAz6q9CoUaOM9evXGwcOHDC2bNlidO/e3ahWrZpx8OBBwzDop7/bunWr4efnZ4wfP9745ZdfjA8++MAIDg42FixY4GhDf/0lPz/fqFOnjvHss88WOUc//aV///7G5Zdf7lgi7eOPPzZq1apljB492tGG/vrLypUrjS+++MI4cOCAsXr1aqNZs2bG9ddfb+Tk5BiGYd6+OnXqlPHDDz8YP/zwgyHJmDJlivHDDz8YKSkphmG41i+PP/64ccUVVxhr1641vv/+e+OWW25hiTR4h0ceecSoW7euERAQYERERBidO3d2BGDDOLcESmJiohEVFWVYrVbj5ptvNnbu3OnBir3L+SGY/jqncK1If39/IyYmxujZs6exe/dux3n6ydny5cuN+Ph4w2q1GldddZUxe/Zsp/P0119WrVplSDL27dtX5Bz99BebzWYMHz7cqFOnjhEYGGjUr1/fGDt2rGG32x1t6K+/LF682Khfv74REBBgREVFGUOHDjVOnjzpOG/Wvlq3bp0hqcijf//+hmG41i/Z2dnGk08+adSsWdMICgoyunfvbhw6dMht12AxDMNwz5gzAAAA4B2YEwwAAADTIQQDAADAdAjBAAAAMB1CMAAAAEyHEAwAAADTIQQDAADAdAjBAAAAMB1CMAAAAEyHEAwAXmDu3LmyWCz67rvvSm03YMAA1atXr1Jr2bNnj5KSknTw4EGXPn/ChAn65JNPKrUmAKhohGAAgJM9e/Zo3LhxxYbgF198UUuXLnU6RggGUBX5eboAAEDV0aBBA0+XAAAVgpFgAPBSc+fOVePGjWW1WnX11Vfr/fffL7ZdTk6OXnnlFV111VWyWq2KiIjQww8/rIyMDKd29erVU/fu3bVy5Uq1aNFCQUFBuuqqq/S///u/Tp95//33S5I6deoki8Uii8WiuXPnSio6HcJisejMmTOaN2+eo23Hjh118OBB+fn5aeLEiUXq3bhxoywWi/79739fZA8BwIUjBAOAF5o7d64efvhhXX311VqyZIleeOEFvfzyy/rqq6+c2hUUFOjuu+/WpEmT1KdPH3322WeaNGmS1qxZo44dOyo7O9up/Y8//qhRo0ZpxIgR+vTTT3Xttdfq0Ucf1caNGyVJ3bp104QJEyRJM2bM0DfffKNvvvlG3bp1K7bOb775RkFBQbrzzjsdbWfOnKl69eqpR48eeuutt5Sfn+/0munTpysmJkb33ntvRXUXAJQb0yEAwMsUFBRo7NixatGihZYuXSqLxSJJat++vRo2bKiYmBhH248++kgrV67UkiVL1LNnT8fxZs2aqXXr1po7d66eeOIJx/Fjx47pv//9r+rUqSNJuvnmm/Xll19q4cKFuvnmmxUREaGGDRtKkq655hrdcMMNpdZ6ww03yMfHRxEREUXaPvXUU+rUqZOWL1+ue+65R5J09OhRLV26VC+++KL8/PgnCIDnMBIMAF5m3759Onr0qPr06eMIwJJUt25dtWvXzqntihUrVL16dd11113Ky8tzPJo3b66oqCitX7/eqX3z5s0dAViSAgMD1ahRI6WkpFT4dXTs2FHNmjXTjBkzHMfeeustWSwWDRo0qMI/DwDKgxAMAF7m+PHjkqSoqKgi584/9scff+jkyZMKCAiQv7+/0yMtLU3Hjh1zah8eHl7kPa1Wa5FpExXlqaee0pdffql9+/YpNzdXc+bM0X333VfstQGAO/G3KADwMoVBNS0trci584/VqlVL4eHhWrlyZbHvVa1atYovsBz69OmjZ599VjNmzNANN9ygtLQ0DR061KM1AYBECAYAr9O4cWNFR0frww8/1MiRIx1TIlJSUrR582anOcHdu3fXokWLlJ+frzZt2lTI51utVklyeXS4tJHkwMBADRo0SNOnT9fmzZvVvHlz3XjjjRVSJwBcDEIwAHgZHx8fvfzyy3rsscd07733auDAgTp58qSSkpKKTCN48MEH9cEHH+jOO+/U8OHDdf3118vf319HjhzRunXrdPfdd5d7FYb4+HhJ0uzZs1WtWjUFBgYqLi6u2KkUktS0aVOtX79ey5cvV3R0tKpVq6bGjRs7zg8ZMkSTJ0/W9u3b9c4775SzNwCgcjAnGAC80KOPPqp33nlHe/bsUc+ePfXSSy/p+eef1y233OLUztfXV8uWLdPzzz+vjz/+WPfee6/uueceTZo0SYGBgWratGm5PzsuLk5Tp07Vjz/+qI4dO6p169Zavnx5ie3ffPNNNWzYUA8++KBat26twYMHO52//PLL1b59e9WsWVN9+vQpdz0AUBkshmEYni4CAHDpSk9PV926dTVs2DBNnjzZ0+UAgCSmQwAAKsmRI0d04MABvfbaa/Lx8dHw4cM9XRIAODAdAgBQKd555x117NhRu3fv1gcffKDLL7/c0yUBgAPTIQAAAGA6jAQDAADAdAjBAAAAMB1CMAAAAEyHEAwAAADTIQQDAADAdAjBAAAAMB1CMAAAAEyHEAwAAADT+f+wawK3wyyqdAAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 800x600 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#bin list of identities by values 50 to 55, 55 to 60, 60 to 65, 65 to 70, 70 to 75, 75 to 80, 80 to 85, 85 to 90, 90 to 95, 95 to 100\n",
    "plt.figure(figsize=(8, 6))\n",
    "\n",
    "plt.hist(list_of_all_identities, bins = 50, color = 'Pink', edgecolor='black', alpha = 0.7)\n",
    "# plt.yscale('log')\n",
    "plt.xlabel('Identity', fontsize = 12)\n",
    "plt.ylabel('Count', fontsize = 12)\n",
    "plt.title('Histogram of identities', fontsize = 14)\n",
    "\n",
    "#remove frame and white space around plot\n",
    "sns.despine()\n",
    "# Add a black dashed vertical line at x = 97\n",
    "plt.axvline(x=97, color='black', linestyle='--')\n",
    "\n",
    "plt.savefig(f\"{work_dir}/protein_extracted_longest/clusters_merged/identity_histogram.\")\n",
    "# plt.yscale('log')\n",
    "# plt.savefig(f\"{work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/clusters_merged/identity_histogram_log.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene_symbol</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Class</th>\n",
       "      <th>gene_family_symbol</th>\n",
       "      <th>in_palindrome</th>\n",
       "      <th>palindrome_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PanTro</td>\n",
       "      <td>MAM and LDL-receptor class A domain-containing...</td>\n",
       "      <td>LOC104004304</td>\n",
       "      <td>645722</td>\n",
       "      <td>645746</td>\n",
       "      <td>-</td>\n",
       "      <td>ANCESTRAL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PanTro</td>\n",
       "      <td>PI-PLC X domain-containing protein 1</td>\n",
       "      <td>LOC104004678</td>\n",
       "      <td>733989</td>\n",
       "      <td>734116</td>\n",
       "      <td>+</td>\n",
       "      <td>PAR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PanTro</td>\n",
       "      <td>putative GTP-binding protein 6</td>\n",
       "      <td>LOC107971128</td>\n",
       "      <td>754146</td>\n",
       "      <td>754270</td>\n",
       "      <td>-</td>\n",
       "      <td>PAR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PanTro</td>\n",
       "      <td>serine/threonine-protein phosphatase 2A regula...</td>\n",
       "      <td>LOC112207609</td>\n",
       "      <td>781156</td>\n",
       "      <td>781307</td>\n",
       "      <td>-</td>\n",
       "      <td>PAR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PanTro</td>\n",
       "      <td>short stature homeobox protein</td>\n",
       "      <td>LOC107971217</td>\n",
       "      <td>1118844</td>\n",
       "      <td>1119121</td>\n",
       "      <td>+</td>\n",
       "      <td>PAR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7615</th>\n",
       "      <td>MacFas</td>\n",
       "      <td>cancer/testis antigen 1-like</td>\n",
       "      <td>LOC102119988</td>\n",
       "      <td>160768729</td>\n",
       "      <td>160768977</td>\n",
       "      <td>+</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>LAGE3</td>\n",
       "      <td>no</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7616</th>\n",
       "      <td>MacFas</td>\n",
       "      <td>histone H2A-Bbd type 2/3-like</td>\n",
       "      <td>LOC123571317</td>\n",
       "      <td>161473448</td>\n",
       "      <td>161473796</td>\n",
       "      <td>+</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>H2AB3</td>\n",
       "      <td>no</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7617</th>\n",
       "      <td>MacFas</td>\n",
       "      <td>40-kDa huntingtin-associated protein</td>\n",
       "      <td>LOC135969152</td>\n",
       "      <td>161474876</td>\n",
       "      <td>161475995</td>\n",
       "      <td>+</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>F8A2</td>\n",
       "      <td>no</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7618</th>\n",
       "      <td>MacFas</td>\n",
       "      <td>40-kDa huntingtin-associated protein</td>\n",
       "      <td>LOC123571316</td>\n",
       "      <td>161508487</td>\n",
       "      <td>161509957</td>\n",
       "      <td>-</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>F8A2</td>\n",
       "      <td>no</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7619</th>\n",
       "      <td>MacFas</td>\n",
       "      <td>histone H2A-Bbd type 2/3-like</td>\n",
       "      <td>LOC135969154</td>\n",
       "      <td>161510681</td>\n",
       "      <td>161511029</td>\n",
       "      <td>-</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>H2AB3</td>\n",
       "      <td>no</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7620 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Species                                               Gene   Gene_symbol  \\\n",
       "0     PanTro  MAM and LDL-receptor class A domain-containing...  LOC104004304   \n",
       "1     PanTro               PI-PLC X domain-containing protein 1  LOC104004678   \n",
       "2     PanTro                     putative GTP-binding protein 6  LOC107971128   \n",
       "3     PanTro  serine/threonine-protein phosphatase 2A regula...  LOC112207609   \n",
       "4     PanTro                     short stature homeobox protein  LOC107971217   \n",
       "...      ...                                                ...           ...   \n",
       "7615  MacFas                       cancer/testis antigen 1-like  LOC102119988   \n",
       "7616  MacFas                      histone H2A-Bbd type 2/3-like  LOC123571317   \n",
       "7617  MacFas               40-kDa huntingtin-associated protein  LOC135969152   \n",
       "7618  MacFas               40-kDa huntingtin-associated protein  LOC123571316   \n",
       "7619  MacFas                      histone H2A-Bbd type 2/3-like  LOC135969154   \n",
       "\n",
       "          Start        End Strand      Class gene_family_symbol in_palindrome  \\\n",
       "0        645722     645746      -  ANCESTRAL                NaN            no   \n",
       "1        733989     734116      +        PAR                NaN            no   \n",
       "2        754146     754270      -        PAR                NaN            no   \n",
       "3        781156     781307      -        PAR                NaN            no   \n",
       "4       1118844    1119121      +        PAR                NaN            no   \n",
       "...         ...        ...    ...        ...                ...           ...   \n",
       "7615  160768729  160768977      +    Unknown              LAGE3            no   \n",
       "7616  161473448  161473796      +    Unknown              H2AB3            no   \n",
       "7617  161474876  161475995      +    Unknown               F8A2            no   \n",
       "7618  161508487  161509957      -    Unknown               F8A2            no   \n",
       "7619  161510681  161511029      -    Unknown              H2AB3            no   \n",
       "\n",
       "     palindrome_name  \n",
       "0               None  \n",
       "1               None  \n",
       "2               None  \n",
       "3               None  \n",
       "4               None  \n",
       "...              ...  \n",
       "7615            None  \n",
       "7616            None  \n",
       "7617            None  \n",
       "7618            None  \n",
       "7619            None  \n",
       "\n",
       "[7620 rows x 10 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Apply whether a gene lays within a palindrome\n",
    "gene_details_df = pd.read_csv(f\"{work_dir}/protein_extracted_longest/clusters_merged/gene_details_updated.tsv\", sep='\\t')\n",
    "\n",
    "## Check if a gene is within any palindrome\n",
    "def check_palindrome(specie, gene_start, gene_end):\n",
    "    # Get the list of palindromes for the given species\n",
    "    if specie in palindrome_lines_per_species:\n",
    "        palindromes = palindrome_lines_per_species[specie]\n",
    "        \n",
    "        for palindrome in palindromes:\n",
    "            _, pal_start, pal_end, _, pal_name = palindrome\n",
    "            \n",
    "            # Convert coordinates to integers\n",
    "            pal_start = int(pal_start)\n",
    "            pal_end = int(pal_end)\n",
    "            \n",
    "            # Check if gene coordinates fall within the palindrome range\n",
    "            if gene_start >= pal_start and gene_end <= pal_end:\n",
    "                return 'yes', pal_name\n",
    "    \n",
    "    return 'no', None\n",
    "    \n",
    "# New columns to be added\n",
    "gene_details_df['in_palindrome'] = 'no'\n",
    "gene_details_df['palindrome_name'] = None\n",
    "\n",
    "# Iterate over each row in the dataframe\n",
    "for idx, row in gene_details_df.iterrows():\n",
    "    species = row['Species']  # Assuming column named 'specie'\n",
    "    gene_start = row['Start']  # Assuming column named 'start'\n",
    "    gene_end = row['End']  # Assuming column named 'end'\n",
    "    \n",
    "    in_palindrome, palindrome_name = check_palindrome(species, gene_start, gene_end)\n",
    "    \n",
    "    # Update the dataframe with the results\n",
    "    gene_details_df.at[idx, 'in_palindrome'] = in_palindrome\n",
    "    gene_details_df.at[idx, 'palindrome_name'] = palindrome_name\n",
    "\n",
    "# Save the updated dataframe to a new file\n",
    "gene_details_df.to_csv(f\"{work_dir}/protein_extracted_longest/clusters_merged/gene_details_updated_with_palindromes.tsv\", sep='\\t', index=False)\n",
    "\n",
    "gene_details_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# UPDATE THE GENE COORDINATES! they are not the full gene length now\n",
    "gene_details_file = f\"{work_dir}/protein_extracted_longest/clusters_merged/gene_details_updated_with_palindromes.tsv\"\n",
    "df = pd.read_csv(gene_details_file, sep=\"\\t\")\n",
    "df\n",
    "\n",
    "species_to_gff = {\n",
    "    d['species']: d['data']['path_to_annotation_NCBI'].replace(\".gff\", \"_chrX.gff\")\n",
    "    for d in data\n",
    "}\n",
    "\n",
    "def parse_gene_coords(gff_path):\n",
    "    coords = {}\n",
    "    with open(gff_path) as g:\n",
    "        for line in g:\n",
    "            if line.startswith('#'): continue\n",
    "            chrom,src,feat,start,end,_,_,_,attr = line.split('\\t',8)\n",
    "            if feat != 'gene': continue\n",
    "            # pull out gene_symbol (or fallback to gene=)\n",
    "            md = dict(item.split('=',1) for item in attr.split(';') if '=' in item)\n",
    "            sym = md.get('gene_symbol') or md.get('gene')\n",
    "            if sym:\n",
    "                coords[sym] = (int(start), int(end))\n",
    "    return coords\n",
    "\n",
    "gff_cache = {\n",
    "    sp: parse_gene_coords(path)\n",
    "    for sp, path in species_to_gff.items()\n",
    "}\n",
    "\n",
    "# replace Start/End \n",
    "new_starts, new_ends = [], []\n",
    "for _, row in df.iterrows():\n",
    "    sp, sym = row['Species'], row['Gene_symbol']\n",
    "    start,end = gff_cache[sp].get(sym, (row['Start'], row['End']))\n",
    "    new_starts.append(start)\n",
    "    new_ends.append(end)\n",
    "\n",
    "df['Start'] = new_starts\n",
    "df['End']   = new_ends\n",
    "df\n",
    "\n",
    "# save the table \n",
    "df.to_csv(f\"{work_dir}/protein_extracted_longest/clusters_merged/gene_details_updated_with_palindromes_coordinates.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene_symbol</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Class</th>\n",
       "      <th>gene_family_symbol</th>\n",
       "      <th>in_palindrome</th>\n",
       "      <th>palindrome_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PanTro</td>\n",
       "      <td>colony stimulating factor 2 receptor subunit a...</td>\n",
       "      <td>LOC746936</td>\n",
       "      <td>1771418</td>\n",
       "      <td>1838874</td>\n",
       "      <td>+</td>\n",
       "      <td>PAR</td>\n",
       "      <td>CSF2RA</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>PanTro</td>\n",
       "      <td>sperm protein associated with the nucleus on t...</td>\n",
       "      <td>LOC129388397</td>\n",
       "      <td>3486949</td>\n",
       "      <td>3488083</td>\n",
       "      <td>+</td>\n",
       "      <td>AMPLICONIC</td>\n",
       "      <td>SPANXA1</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>PanTro</td>\n",
       "      <td>SPANX A/D member 3</td>\n",
       "      <td>LOC493983</td>\n",
       "      <td>3566190</td>\n",
       "      <td>3567169</td>\n",
       "      <td>+</td>\n",
       "      <td>AMPLICONIC</td>\n",
       "      <td>SPANXA1</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>PanTro</td>\n",
       "      <td>protein kinase X-linked</td>\n",
       "      <td>PRKX</td>\n",
       "      <td>4149880</td>\n",
       "      <td>4259574</td>\n",
       "      <td>-</td>\n",
       "      <td>ANCESTRAL</td>\n",
       "      <td>TBL1X</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>PanTro</td>\n",
       "      <td>atherin-like</td>\n",
       "      <td>LOC740134</td>\n",
       "      <td>4370640</td>\n",
       "      <td>4405131</td>\n",
       "      <td>-</td>\n",
       "      <td>ANCESTRAL</td>\n",
       "      <td>TBL1X</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7615</th>\n",
       "      <td>MacFas</td>\n",
       "      <td>cancer/testis antigen 1-like</td>\n",
       "      <td>LOC102119988</td>\n",
       "      <td>160766910</td>\n",
       "      <td>160770217</td>\n",
       "      <td>+</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>LAGE3</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7616</th>\n",
       "      <td>MacFas</td>\n",
       "      <td>histone H2A-Bbd type 2/3-like</td>\n",
       "      <td>LOC123571317</td>\n",
       "      <td>161473449</td>\n",
       "      <td>161473969</td>\n",
       "      <td>+</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>H2AB3</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7617</th>\n",
       "      <td>MacFas</td>\n",
       "      <td>40-kDa huntingtin-associated protein</td>\n",
       "      <td>LOC135969152</td>\n",
       "      <td>161474855</td>\n",
       "      <td>161476570</td>\n",
       "      <td>+</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>F8A2</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7618</th>\n",
       "      <td>MacFas</td>\n",
       "      <td>40-kDa huntingtin-associated protein</td>\n",
       "      <td>LOC123571316</td>\n",
       "      <td>161507913</td>\n",
       "      <td>161509975</td>\n",
       "      <td>-</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>F8A2</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7619</th>\n",
       "      <td>MacFas</td>\n",
       "      <td>histone H2A-Bbd type 2/3-like</td>\n",
       "      <td>LOC135969154</td>\n",
       "      <td>161510534</td>\n",
       "      <td>161512971</td>\n",
       "      <td>-</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>H2AB3</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1702 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Species                                               Gene   Gene_symbol  \\\n",
       "6     PanTro  colony stimulating factor 2 receptor subunit a...     LOC746936   \n",
       "23    PanTro  sperm protein associated with the nucleus on t...  LOC129388397   \n",
       "24    PanTro                                 SPANX A/D member 3     LOC493983   \n",
       "28    PanTro                            protein kinase X-linked          PRKX   \n",
       "29    PanTro                                       atherin-like     LOC740134   \n",
       "...      ...                                                ...           ...   \n",
       "7615  MacFas                       cancer/testis antigen 1-like  LOC102119988   \n",
       "7616  MacFas                      histone H2A-Bbd type 2/3-like  LOC123571317   \n",
       "7617  MacFas               40-kDa huntingtin-associated protein  LOC135969152   \n",
       "7618  MacFas               40-kDa huntingtin-associated protein  LOC123571316   \n",
       "7619  MacFas                      histone H2A-Bbd type 2/3-like  LOC135969154   \n",
       "\n",
       "          Start        End Strand       Class gene_family_symbol  \\\n",
       "6       1771418    1838874      +         PAR             CSF2RA   \n",
       "23      3486949    3488083      +  AMPLICONIC            SPANXA1   \n",
       "24      3566190    3567169      +  AMPLICONIC            SPANXA1   \n",
       "28      4149880    4259574      -   ANCESTRAL              TBL1X   \n",
       "29      4370640    4405131      -   ANCESTRAL              TBL1X   \n",
       "...         ...        ...    ...         ...                ...   \n",
       "7615  160766910  160770217      +     Unknown              LAGE3   \n",
       "7616  161473449  161473969      +     Unknown              H2AB3   \n",
       "7617  161474855  161476570      +     Unknown               F8A2   \n",
       "7618  161507913  161509975      -     Unknown               F8A2   \n",
       "7619  161510534  161512971      -     Unknown              H2AB3   \n",
       "\n",
       "     in_palindrome palindrome_name  \n",
       "6               no             NaN  \n",
       "23              no             NaN  \n",
       "24              no             NaN  \n",
       "28              no             NaN  \n",
       "29              no             NaN  \n",
       "...            ...             ...  \n",
       "7615            no             NaN  \n",
       "7616            no             NaN  \n",
       "7617            no             NaN  \n",
       "7618            no             NaN  \n",
       "7619            no             NaN  \n",
       "\n",
       "[1702 rows x 10 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove all rows that have NaN in gene_family_symbol -> all the ones that are not part of a ampliconic gene family\n",
    "df = df.dropna(subset=[\"gene_family_symbol\"])\n",
    "df.to_csv(f\"{work_dir}/protein_extracted_longest/clusters_merged/gene_details_updated_with_palindromes_coordinates.tsv\", sep=\"\\t\", index=False)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Y-Chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#runs around 20s without extracting the sequence\n",
    "#runs around 1m with extracting the sequence\n",
    "#only collect ID's will extract sequence from cds_from_genomic.fna\n",
    "record_dict = {}\n",
    "\n",
    "for species in species_list:\n",
    "\n",
    "\n",
    "    annotation_file =  [d for d in data if d['species'] == species][0]['data']['path_to_annotation_NCBI'].replace(\".gff\", \"_chrY.gff\")\n",
    "    handle = open(annotation_file)\n",
    "\n",
    "    # pprint.pprint(examiner.parent_child_map(handle))\n",
    "\n",
    "    # assert False\n",
    "\n",
    "    # records = GFF.parse(handle,base_dict=seq_dict)\n",
    "    records = GFF.parse(handle)\n",
    "    for rec in records:\n",
    "        protss = set()\n",
    "        for feature in rec.features:\n",
    "            if feature.type != \"gene\":\n",
    "                continue\n",
    "            if \"description\" in feature.qualifiers:\n",
    "                description = feature.qualifiers[\"description\"][0]\n",
    "\n",
    "                gene_id = feature.id.replace(\"gene-\", \"\")\n",
    "                gene_spec_id = gene_id + \"_\" + species\n",
    "                    \n",
    "                # print(feature.qualifiers)\n",
    "                for subf in feature.sub_features:\n",
    "\n",
    "                    if subf.type == \"mRNA\":\n",
    "                        mrna_id = subf.id.replace(\"rna-\", \"\")\n",
    "                        for subsubf in subf.sub_features:\n",
    "                            if subsubf.type != \"CDS\":\n",
    "                                continue\n",
    "                            # cds_id = subsubf.id.replace(\"cds-\", \"\")\n",
    "                            prot_id = subsubf.qualifiers[\"protein_id\"][0]\n",
    "                            if prot_id in protss:\n",
    "                                continue\n",
    "\n",
    "                            protss.add(prot_id)\n",
    "                            element = {}\n",
    "                            element[\"specie\"] = species\n",
    "                            element[\"name\"] = description\n",
    "                            element[\"id\"] = prot_id\n",
    "                            element[\"mrna_id\"] = mrna_id\n",
    "                            element[\"gene_id\"] = gene_id\n",
    "                            element[\"start\"] = int(subsubf.location.start)\n",
    "                            element[\"end\"] = int(subsubf.location.end)\n",
    "                            element[\"strand\"] = \"+\" if subsubf.location.strand == 1 else \"-\"  # Extracting strand (adapted from original script to later be able to look at the organization of the genes)\n",
    "                            element[\"gene_spec_id\"] = gene_spec_id\n",
    "\n",
    "                            # element[\"seq\"] = str(subf.extract(rec.seq))\n",
    "                            if gene_id in record_dict:\n",
    "                                record_dict[gene_id].append(element)\n",
    "                            else:\n",
    "                                record_dict[gene_id] = [element]\n",
    "\n",
    "    handle.close()\n",
    "\n",
    "#pprint(record_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#read sequence classes for all species\n",
    "seq_class_dicts = {}\n",
    "for species in species_list:\n",
    "\n",
    "    sequence_class_file = f\"{data_dir}/AdditionalFile2-SeqClasses.tsv\"\n",
    "    seq_class_dicts[species] = []\n",
    "    with open(sequence_class_file) as f:\n",
    "        for line in f:\n",
    "            line = line.strip().split(\"\\t\")\n",
    "            if line[0] == species_to_sequence_spec[species]:\n",
    "                if line[2] == \"chrY\":\n",
    "                    seq_class_dicts[species].append(line[2:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pprint(seq_class_dicts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read the cds_from_genomic.fna file\n",
    "prot_dicts = {}\n",
    "\n",
    "for species in species_list:\n",
    "\n",
    "    sequence_file = [d for d in data if d['species'] == species][0]['data']['prot']\n",
    "    sequence_handle = open(sequence_file)\n",
    "    seq_dict = SeqIO.to_dict(SeqIO.parse(sequence_handle, \"fasta\"))\n",
    "\n",
    "    prot_dicts[species] = seq_dict\n",
    "\n",
    "    sequence_handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "for record in record_dict:\n",
    "    \n",
    "    for element in record_dict[record]:\n",
    "        species = element[\"specie\"]\n",
    "        id = element[\"id\"]\n",
    "        name = element[\"name\"]\n",
    "        prot_sequences = [prot_dicts[species][x] for x in prot_dicts[species] if (id in x )]\n",
    "        if len(prot_sequences) == 0:\n",
    "            print(\"No sequence found\")\n",
    "            print(id)\n",
    "            assert False\n",
    "        seq = str(prot_sequences[0].seq)\n",
    "        if len(seq) > 1:\n",
    "            for prot in prot_sequences:\n",
    "                if len(prot.seq) > len(seq):\n",
    "                    seq = str(prot.seq)\n",
    "\n",
    "        element[\"seq\"] = seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#reduce list of prot sequences to longest sequence per gene\n",
    "gene_record_dict = {}\n",
    "for record in record_dict:\n",
    "    gene_spec_ids = set()\n",
    "    gene_spec_elements = []\n",
    "    \n",
    "    for element in record_dict[record]:\n",
    "        # gene_id = element[\"gene_id\"]\n",
    "        gene_spec_id = element[\"gene_spec_id\"]\n",
    "        \n",
    "        if gene_spec_id not in gene_spec_ids:\n",
    "            gene_spec_elements.append(element)\n",
    "        else:\n",
    "            if len(element[\"seq\"]) > len([x for x in gene_spec_elements if x[\"gene_spec_id\"] == gene_spec_id][0][\"seq\"]):\n",
    "                for el in gene_spec_elements:\n",
    "                    if el[\"gene_spec_id\"] == gene_spec_id:\n",
    "                        el[\"seq\"] = element[\"seq\"]\n",
    "                        break\n",
    "\n",
    "        gene_spec_ids.add(element[\"gene_spec_id\"])\n",
    "    \n",
    "    gene_record_dict[record] = gene_spec_elements\n",
    "\n",
    "#pprint(gene_record_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print one protein file per specie\n",
    "for species in species_list:\n",
    "    file = f\"{work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/{species}.faa\"\n",
    "\n",
    "\n",
    "    species_records = {}\n",
    "    for gene in gene_record_dict:\n",
    "        species_records[gene] = []\n",
    "        for element in gene_record_dict[gene]:\n",
    "            if element[\"specie\"] == species:\n",
    "                species_records[gene].append(element)\n",
    "\n",
    "\n",
    "    with open(file, \"w\") as outfile:\n",
    "        for entry in species_records:\n",
    "            element = species_records[entry]\n",
    "            if len(element) > 1:\n",
    "                print(species)\n",
    "                assert False\n",
    "            if len(element) == 0:\n",
    "                continue\n",
    "            element = element[0]\n",
    "            outfile.write(f\">{element['id']} gene:{element['gene_id']} mrna:{element['mrna_id']} {element['specie']} {element['name']}\\n\")\n",
    "            outfile.write(f\"{element['seq']}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='cat /home/emma/Amplicons/Workspaces/emma/downloaded_data/work_dir/y_multicopy/protein_extracted_longest/*.faa     > /home/emma/Amplicons/Workspaces/emma/downloaded_data/work_dir/y_multicopy/protein_extracted_longest/blastdb/all_proteins.faa', returncode=0)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#make blast db\n",
    "cmd = f\"mkdir -p {work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/blastdb\"\n",
    "subprocess.run(cmd, shell=True, check=True)\n",
    "#concat all fasta files\n",
    "cmd = f\"cat {work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/*.faa \\\n",
    "    > {work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/blastdb/all_proteins.faa\"\n",
    "# run command\n",
    "subprocess.run(cmd, shell=True, check=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 02/02/2026 09:35:51\n",
      "New DB name:   /home/emma/Amplicons/Workspaces/emma/downloaded_data/work_dir/y_multicopy/protein_extracted_longest/blastdb/all_proteins\n",
      "New DB title:  /home/emma/Amplicons/Workspaces/emma/downloaded_data/work_dir/y_multicopy/protein_extracted_longest/blastdb/all_proteins.faa\n",
      "Sequence type: Protein\n",
      "Deleted existing Protein BLAST database named /home/emma/Amplicons/Workspaces/emma/downloaded_data/work_dir/y_multicopy/protein_extracted_longest/blastdb/all_proteins\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 761 sequences in 0.102918 seconds.\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='makeblastdb -in /home/emma/Amplicons/Workspaces/emma/downloaded_data/work_dir/y_multicopy/protein_extracted_longest/blastdb/all_proteins.faa      -dbtype prot     -out /home/emma/Amplicons/Workspaces/emma/downloaded_data/work_dir/y_multicopy/protein_extracted_longest/blastdb/all_proteins', returncode=0)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#make blast db\n",
    "cmd = f\"makeblastdb -in {work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/blastdb/all_proteins.faa \\\n",
    "     -dbtype prot\\\n",
    "     -out {work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/blastdb/all_proteins\"\n",
    "\n",
    "subprocess.run(cmd, shell=True, check=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# blast all against DB\n",
    "for s in species_list:\n",
    "    cmd = f\"blastp -query {work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/{s}.faa \\\n",
    "    -db {work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/blastdb/all_proteins \\\n",
    "    -out {work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/{s}.blastp.tsv \\\n",
    "    -outfmt \\\"6 qseqid sseqid pident mismatch gapopen gaps qcovs qcovshsp evalue\\\" \"\n",
    "    subprocess.run(cmd, shell=True, check=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#collect edges for each protein\n",
    "edges = {}\n",
    "identity = 50\n",
    "coverage = 35\n",
    "score = 0.001 \n",
    "\n",
    "for s in species_list:\n",
    "    file = f\"{work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/{s}.blastp.tsv\"\n",
    "    with open(file, \"r\") as infile:\n",
    "        for line in infile:\n",
    "            line = line.strip().split(\"\\t\")\n",
    "            if float(line[2]) >= identity and int(line[6]) >= coverage and float(line[7]) < score:\n",
    "                if line[0] in edges:\n",
    "                    edges[line[0]].append(line[1])\n",
    "                else:\n",
    "                    edges[line[0]] = [line[1]]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#only keep two way edges\n",
    "edges_2way = []\n",
    "vertices = set()\n",
    "for node_A in edges:\n",
    "    for node_B in edges[node_A]:\n",
    "        if node_B in edges and node_A in edges[node_B]:\n",
    "            tuple = (node_A, node_B)\n",
    "            tuple = sorted(tuple)\n",
    "\n",
    "            if tuple in edges_2way:\n",
    "                continue\n",
    "\n",
    "            edges_2way.append(tuple)\n",
    "            vertices.add(node_A)\n",
    "            vertices.add(node_B)          "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create clusters with transitive clusering\n",
    "clusters = []\n",
    "for edge in edges_2way:\n",
    "    found = False\n",
    "    for cluster in clusters:\n",
    "        if edge[0] in cluster or edge[1] in cluster:\n",
    "            cluster.add(edge[0])\n",
    "            cluster.add(edge[1])\n",
    "            found = True\n",
    "            break\n",
    "    if not found:\n",
    "        clusters.append(set(edge))\n",
    "\n",
    "\n",
    "#merge clusters with common elements\n",
    "merged_clusters = []\n",
    "for cluster in clusters:\n",
    "    found = False\n",
    "    for merged_cluster in merged_clusters:\n",
    "        if len(cluster.intersection(merged_cluster)) > 0:\n",
    "            merged_cluster.update(cluster)\n",
    "            found = True\n",
    "            break\n",
    "    if not found:\n",
    "        merged_clusters.append(cluster)\n",
    "\n",
    "#merged_clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "prot_to_species = {}\n",
    "\n",
    "for cluster in merged_clusters:\n",
    "\n",
    "    for prot in cluster:\n",
    "        # print(prot)\n",
    "        species = \"\"\n",
    "        for gene in gene_record_dict:\n",
    "            for element in gene_record_dict[gene]:\n",
    "                if element[\"id\"] == prot:\n",
    "                    species = element[\"specie\"]\n",
    "                    prot_to_species[prot] = species\n",
    "                        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "#read blast results into pandas dataframe\n",
    "# columns :species, gene_A, gene_B, identity, coverage, evalue\n",
    "\n",
    "\n",
    "\n",
    "data_frames = {}\n",
    "for species in species_list:\n",
    "    blast_results = pd.DataFrame(columns=[\"gene_A\", \"gene_B\", \"identity\", \"coverage\", \"evalue\"]) \n",
    "    blast_results_list = []\n",
    "    file = f\"{work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/{species}.blastp.tsv\"\n",
    "    with open(file, \"r\") as infile:\n",
    "        for line in infile:\n",
    "            line = line.strip().split(\"\\t\")\n",
    "            # if float(line[2]) >= identity and int(line[6]) >= coverage and float(line[7]) < score:\n",
    "            blast_results_list.append({\"species\": species, \"gene_A\": line[0], \"gene_B\": line[1], \"identity\": line[2], \"coverage\": line[6], \"evalue\": line[7]})\n",
    "\n",
    "    blast_results = pd.DataFrame(blast_results_list)\n",
    "    data_frames[species] = blast_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "palindrome_locations_file_path = \"./PalindromeLocations.tsv\" #this file is distributed along the code\n",
    "species_trans = {\n",
    "    \"HomSap_x\": \"chm13\"    ,\n",
    "    \"HomSap_y\": \"hg002\"    ,\n",
    "    \"GorGor\": \"mGorGor1\" ,\n",
    "    \"PanTro\": \"mPanTro3\" ,\n",
    "    \"PanPan\": \"mPanPan1\" ,\n",
    "    \"PonAbe\": \"mPonAbe1\" ,\n",
    "    \"PonPyg\": \"mPonPyg2\" ,\n",
    "    \"SymSyn\": \"mSymSyn1\" \n",
    "}\n",
    "palindrome_locations_file = open(palindrome_locations_file_path, \"r\")\n",
    "palindrome_locations = palindrome_locations_file.readlines()\n",
    "palindrome_locations_file.close()\n",
    "chr = \"y\"\n",
    "\n",
    "palindrome_lines_per_species = {}\n",
    "\n",
    "for species in species_list:\n",
    "    # Build the list of translation keys we actually have (HomSap is divided into X and Y so this should be taken into account):\n",
    "    if species == \"HomSap\":\n",
    "        raw_keys = [f\"{species}_x\", f\"{species}_y\"]\n",
    "    else:\n",
    "        raw_keys = [species]\n",
    "    # Only keep keys that exist in species_trans. Check if species exists in species_trans -> BECAUSE MacFas does not have palindrome information so this should be empty\n",
    "\n",
    "    keys = [k for k in raw_keys if k in species_trans]\n",
    "\n",
    "    # If none of those keys are in species_trans, just give an empty list and skip\n",
    "    if not keys:\n",
    "        palindrome_lines_per_species[species] = []\n",
    "        continue\n",
    "        \n",
    "    palindrome_lines = []\n",
    "        \n",
    "    for row in palindrome_locations:\n",
    "        row = row.strip().split(\"\\t\")\n",
    "        if len(row) < 4:\n",
    "            continue\n",
    "        if row[1] != '' and row[1] != 'start':\n",
    "                    \n",
    "            if species == \"HomSap\":\n",
    "                sp = f\"{species}_{chr}\"\n",
    "                species_chr = f\"{species_trans[sp]}.chr{chr.upper()}\"\n",
    "            else:\n",
    "                species_chr = f\"{species_trans[species]}.chr{chr.upper()}\"\n",
    "\n",
    "            if row[0] == species_chr:\n",
    "                palindrome_lines.append([x.replace(',','') for x in row])\n",
    "    palindrome_lines_per_species[species] = palindrome_lines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pprint(palindrome_lines_per_species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmd = f\"mkdir -p {work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/clusters_merged\"\n",
    "subprocess.run(cmd, shell=True, check=True)\n",
    "\n",
    "file = open(f\"{work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/clusters_merged/clusters.tsv\", \"w\")\n",
    "file_identities = open(f\"{work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/clusters_merged/clusters_identities.tsv\", \"w\")\n",
    "file_classes = open(f\"{work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/clusters_merged/clusters_classes.tsv\", \"w\")\n",
    "\n",
    "header = \"id\\tgene_symbol\\tDescription(s)\"\n",
    "for species in species_list:\n",
    "    header += f\"\\t{species}\"\n",
    "\n",
    "print(header, file=file)\n",
    "print(header, file=file_identities)\n",
    "print(header, file=file_classes)\n",
    "\n",
    "counter = 0\n",
    "prot_to_cluster = {}\n",
    "list_of_all_identities = []\n",
    "for cluster in merged_clusters:\n",
    "\n",
    "    cluster_copy = copy.deepcopy(cluster)\n",
    "    counter += 1\n",
    "    species_count = {}\n",
    "    gene_names = set()\n",
    "    gene_symbol = \"\"\n",
    "    report = False\n",
    "    gene_identity_counter_per_species = {}\n",
    "\n",
    "    classes_per_species = {}\n",
    "    for species in species_list:\n",
    "        classes_per_species[species] = {\n",
    "                \"par_counter\" : 0,\n",
    "                \"ampl_counter\" : 0,\n",
    "                \"ancestral_counter\" : 0,\n",
    "                \"palindrome_counter\" : 0\n",
    "        }\n",
    "    \n",
    "\n",
    "    cluster_fasta_file = f\"{work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/clusters_merged/cluster_{counter}.faa\"\n",
    "    with open(cluster_fasta_file, \"w\") as cluster_outfile:\n",
    "        for prot in cluster:\n",
    "            # print(prot)\n",
    "            species = \"\"\n",
    "            desc = \"\"\n",
    "            for gene in gene_record_dict:\n",
    "                for element in gene_record_dict[gene]:\n",
    "                    if element[\"id\"] == prot:\n",
    "                        gene_names.add(element[\"name\"])\n",
    "                        species = element[\"specie\"]\n",
    "                        # print(element)\n",
    "                        desc = element[\"name\"]\n",
    "                        report_gene = gene\n",
    "\n",
    "                        prot_to_cluster[prot] = counter\n",
    "                        \n",
    "                        if species in species_count:\n",
    "                            report = True\n",
    "                            # species = element\n",
    "                            species_count[species] += 1\n",
    "                        else:\n",
    "                            species_count[species] = 1\n",
    "                        sequence = element[\"seq\"]\n",
    "\n",
    "                        el_start =  element[\"start\"]\n",
    "                        el_end =  element[\"end\"]\n",
    "\n",
    "                        for seq_class in seq_class_dicts[species]:\n",
    "                            class_name = seq_class[3]\n",
    "                            if el_start >= int(seq_class[1]) and el_end <= int(seq_class[2]):\n",
    "\n",
    "                                if seq_class[3] == \"PAR\":\n",
    "                                    classes_per_species[species][\"par_counter\"] += 1\n",
    "                                elif seq_class[3] == \"AMPLICONIC\":\n",
    "                                    classes_per_species[species][\"ampl_counter\"] += 1\n",
    "                                elif seq_class[3] == \"ANCESTRAL\":\n",
    "                                    classes_per_species[species][\"ancestral_counter\"] += 1\n",
    "                                break\n",
    "                            else:\n",
    "                               #TODO if more than 50% of the sequence is in the class, count it as in the class\n",
    "                                \n",
    "                                if el_start < int(seq_class[1]) and el_end > int(seq_class[1]):\n",
    "                                    if int(seq_class[1]) - el_start > 0.5 * len(sequence):\n",
    "                                        if seq_class[3] == \"PAR\":\n",
    "                                            classes_per_species[species][\"par_counter\"] += 1\n",
    "                                        elif seq_class[3] == \"AMPLICONIC\":\n",
    "                                            classes_per_species[species][\"ampl_counter\"] += 1\n",
    "                                        elif seq_class[3] == \"ANCESTRAL\":\n",
    "                                            classes_per_species[species][\"ancestral_counter\"] += 1\n",
    "                                        break\n",
    "                                if el_start < int(seq_class[2]) and el_end > int(seq_class[2]):\n",
    "                                    if el_end - int(seq_class[2]) > 0.5 * len(sequence):\n",
    "                                        if seq_class[3] == \"PAR\":\n",
    "                                            classes_per_species[species][\"par_counter\"] += 1\n",
    "                                        elif seq_class[3] == \"AMPLICONIC\":\n",
    "                                            classes_per_species[species][\"ampl_counter\"] += 1\n",
    "                                        elif seq_class[3] == \"ANCESTRAL\":\n",
    "                                            classes_per_species[species][\"ancestral_counter\"] += 1\n",
    "                                        break\n",
    "\n",
    "                        \n",
    "                        for palindrome in palindrome_lines_per_species[species]:\n",
    "                            if el_start >= int(palindrome[1]) and el_end <= int(palindrome[2]):\n",
    "                                classes_per_species[species][\"palindrome_counter\"] += 1\n",
    "                                break\n",
    "                            else:\n",
    "                                if el_start < int(palindrome[1]) and el_end > int(palindrome[2]):\n",
    "\n",
    "                                    print(f\"start: {el_start} end: {el_end} palindrome: {palindrome[0]} start: {palindrome[1]} end: {palindrome[2]}\")\n",
    "                                    assert False\n",
    "                                if el_start < int(palindrome[1]) and el_end > int(palindrome[2]):\n",
    "                                    if counter == 245:\n",
    "                                        print(palindrome)\n",
    "                                    print(f\"start: {el_start} end: {el_end} palindrome: {palindrome[0]} start: {palindrome[1]} end: {palindrome[2]}\")\n",
    "                                    assert False\n",
    "\n",
    "                        if element[\"specie\"] == \"HomSap\":\n",
    "                            if gene.startswith(\"LOC\"):\n",
    "                                continue\n",
    "                            gene_symbol = gene\n",
    "                        break\n",
    "            \n",
    "            print(f\">{prot}_gene:{report_gene}_species:{species} description:{desc}\\n{sequence}\", file=cluster_outfile)   \n",
    "            for prot_B in cluster_copy:\n",
    "                if prot == prot_B:\n",
    "                    continue\n",
    "                species = prot_to_species[prot]\n",
    "                if species != prot_to_species[prot_B]:\n",
    "                    continue\n",
    "                # blast results initiated in later cells, TODO: move to top\n",
    "                blast_results = data_frames[species]\n",
    "                blast_result = blast_results[(blast_results[\"gene_A\"] == prot) & (blast_results[\"gene_B\"] == prot_B)]\n",
    "                if blast_result.empty:\n",
    "                    continue\n",
    "\n",
    "                identity1 = blast_result.iloc[0][\"identity\"]\n",
    "\n",
    "                blast_result = blast_results[(blast_results[\"gene_A\"] == prot_B) & (blast_results[\"gene_B\"] == prot)]\n",
    "                if blast_result.empty:\n",
    "                    continue\n",
    "                identity2 = blast_result.iloc[0][\"identity\"]\n",
    "\n",
    "                if identity2 < identity1:\n",
    "                    identity1 = identity2\n",
    "                if species not in gene_identity_counter_per_species:\n",
    "                    gene_identity_counter_per_species[species] = f\"{identity1}\"\n",
    "                else:\n",
    "                    gene_identity_counter_per_species[species] += f\";{identity1}\"\n",
    "                list_of_all_identities.append(float(identity1))\n",
    "            cluster_copy.remove(prot)\n",
    "                \n",
    "    if report:\n",
    "        \n",
    "        if gene_symbol == \"\":\n",
    "            gene_symbol = list(gene_names)[0]\n",
    "        line = f\"{counter}\\t{gene_symbol}\\t{';'.join(gene_names)}\"\n",
    "        for species in species_list:\n",
    "            if species not in species_count:\n",
    "                species_count[species] = 0\n",
    "            line += f\"\\t{species_count[species]}\"\n",
    "\n",
    "        print(line, file=file)\n",
    "\n",
    "        identity_line = f\"{counter}\\t{gene_symbol}\\t{';'.join(gene_names)}\"\n",
    "        for species in species_list:\n",
    "            if species not in gene_identity_counter_per_species:\n",
    "                gene_identity_counter_per_species[species] = 0\n",
    "            identity_line += f\"\\t{gene_identity_counter_per_species[species]}\"\n",
    "        print(identity_line, file=file_identities)\n",
    "\n",
    "        class_line = f\"{counter}\\t{gene_symbol}\\t{';'.join(gene_names)}\"\n",
    "        \n",
    "        per_line_class_counter = {\n",
    "            \"par_counter\" : 0 ,\n",
    "            \"ampl_counter\" : 0,\n",
    "            \"ancestral_counter\" : 0,\n",
    "            \"palindrome_counter\" : 0\n",
    "        }\n",
    "        for species in species_list:\n",
    "            par_counter = classes_per_species[species][\"par_counter\"]\n",
    "            per_line_class_counter[\"par_counter\"] += par_counter\n",
    "            ampl_counter = classes_per_species[species][\"ampl_counter\"]\n",
    "            per_line_class_counter[\"ampl_counter\"] += ampl_counter\n",
    "            ancestral_counter = classes_per_species[species][\"ancestral_counter\"]\n",
    "            per_line_class_counter[\"ancestral_counter\"] += ancestral_counter\n",
    "            palindrome_counter = classes_per_species[species][\"palindrome_counter\"]\n",
    "            per_line_class_counter[\"palindrome_counter\"] += palindrome_counter\n",
    "            class_line += f\"\\t{ampl_counter} ({palindrome_counter}) + {ancestral_counter} + {par_counter}\"\n",
    "\n",
    "\n",
    "        class_line += f\"\\t{per_line_class_counter['ampl_counter']} ({per_line_class_counter['palindrome_counter']}) + {per_line_class_counter['ancestral_counter']} + {per_line_class_counter['par_counter']}\"\n",
    "        print(class_line, file=file_classes)\n",
    "     \n",
    "file.close()\n",
    "file_identities.close()\n",
    "file_classes.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene details have been saved to /home/emma/Amplicons/Workspaces/emma/downloaded_data/work_dir/y_multicopy/protein_extracted_longest/clusters_merged/gene_details.tsv\n"
     ]
    }
   ],
   "source": [
    "## Extract the genes coordinates and which classification they belong to. \n",
    "output_dir = f\"{work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/clusters_merged\"\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "gene_details_file_path = os.path.join(output_dir, \"gene_details.tsv\")\n",
    "\n",
    "# Open the file to save gene details\n",
    "with open(gene_details_file_path, \"w\") as gene_details_file:\n",
    "    # Write header to the file\n",
    "    print(\"Species\\tGene\\tGene_symbol\\tStart\\tEnd\\tStrand\\tClass\", file=gene_details_file)\n",
    "\n",
    "    # Iterate over each species and their corresponding genes\n",
    "    for species in species_list:\n",
    "        for gene in gene_record_dict:\n",
    "            for element in gene_record_dict[gene]:\n",
    "                if element[\"specie\"] == species:\n",
    "                    gene_name = element[\"name\"]\n",
    "                    gene_symbol = gene\n",
    "                    start = element[\"start\"]\n",
    "                    end = element[\"end\"]\n",
    "                    strand = element[\"strand\"]\n",
    "                    sequence_class = \"Unknown\"\n",
    "\n",
    "                    # Determine the sequence class\n",
    "                    for seq_class in seq_class_dicts[species]:\n",
    "                        if start >= int(seq_class[1]) and end <= int(seq_class[2]):\n",
    "                            sequence_class = seq_class[3]\n",
    "                            break\n",
    "\n",
    "                    # Print gene details to the file\n",
    "                    print(f\"{species}\\t{gene_name}\\t{gene_symbol}\\t{start}\\t{end}\\t{strand}\\t{sequence_class}\", file=gene_details_file)\n",
    "\n",
    "print(f\"Gene details have been saved to {gene_details_file_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAsEAAAIlCAYAAAA0ZCXgAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjAsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvlHJYcgAAAAlwSFlzAAAPYQAAD2EBqD+naQAAVaBJREFUeJzt3XtYlHX+//HXyGE4hKOAgJQKtmYm5DkPWWKaR7R0W7csV8uyVlfFQ5ZZP9m2tGxTS8vK3CgPabumqe3ioTzkesZoPaWWipoSWAhyEBDu3x9d3t8mUBGBGbyfj+uaa2fu+33f877vwe3FzWc+t80wDEMAAACAhdRwdQMAAABAVSMEAwAAwHIIwQAAALAcQjAAAAAshxAMAAAAyyEEAwAAwHIIwQAAALAcQjAAAAAshxAMAAAAyyEEA3C5IUOGyGaz6dixY65uBZcxY8YMNWnSRL6+vrLZbEpISLhsvc1mU0xMTJn3Hx8fL5vNpg0bNlxTnxVpw4YNstlsio+PL/M2x44dk81m05AhQyqtLwDXjhAMoEJc/A9/jx49Llmzbdu2Cg8HBI6qsWjRIo0dO1Z+fn4aM2aMJk+erObNm7u6rQpxtWFdkiIiIhQREVEp/QCoGp6ubgAApk6dqmeffVY33nijq1vBJfz73/+WJH3++ecKCwsr0zYHDhyQn59fZbZV6e644w4dOHBAwcHBZd7mxhtv1IEDB+RwOCqxMwDXihAMwOXq1q2runXruroNXMapU6ckqcwBWJJuvfXWymqnyvj5+V31cXh5eV0Xxw5c7xgOAcDlLjUmeOnSperUqZNCQkLk4+OjevXqqUePHlq+fLkkKSEhQZGRkZKkDz/8UDabzXz8elxpbm6u4uPjdeutt8rHx0eBgYHq3bu3tmzZUmo/Z86c0bBhwxQSEiI/Pz+1adNGy5YtU0JCQomxsL8ejvHtt9+qf//+Cg4OdjqeZcuW6aGHHtLvfvc7+fn5yeFw6K677tLSpUtLvPev93fgwAHFxsaqVq1aql27th566CGdOXNGkrR9+3bde++9qlmzpmrXrq0nnnhCOTk5V3XeV61apc6dO8vhcMjX11fNmzfXzJkzVVRUZNZcPOb169dLknl+yzIU4FLDDE6cOKGHHnpIgYGBuuGGG9SpUydt2rTpsvvatGmT+vTpo+DgYNntdjVq1EjPP/+8cnNznep+PYZ39+7d6t69uwICAuRwONSvXz+nn7GLtZK0ceNGp5+fi5/xb8cEX/x8UlJSlJKS4rTNb2tKG6Jz7tw5TZ48WU2bNpWvr69q1aqlHj16aPPmzSVqT58+rdGjR6tRo0by9fVVYGCgoqOjNXz4cGVlZV32fAG4Mq4EA3BLc+bM0fDhw1W3bl3169dPQUFBOn36tHbs2KHly5fr/vvvV/PmzTV69Gi98cYbatasme6//35z+4shLT8/X126dNG2bdvUsmVLxcXFKS0tTUuWLNGaNWu0ZMkS9e/f39wuOztbnTp10v79+9WxY0d17NhRP/zwgx566CF169btkv1+9913ateunZo2barBgwfr559/lre3tyRp4sSJ8vb2VseOHVW3bl2lp6drxYoVeuCBB/Tmm29q5MiRJfZ39OhRdejQQa1bt9bjjz+uXbt2afHixTpx4oReffVV3Xvvvbr33ns1bNgwbdiwQe+//74kae7cuWU6v2+88Ybi4uIUGBiogQMHyt/fXytXrtSYMWP01Vdf6V//+pdsNpuaN2+uyZMnKyEhQSkpKZo8ebIkqVatWmV6n986ffq02rdvrx9++EHdu3dXy5YtdeDAAd17773q3Llzqdu88847Gj58uGrXrq0+ffqoTp062rlzp15++WWtX79e69evN8/1Rbt27dJrr72mmJgYPfnkk/r666+1fPly7dmzR3v37pWPj48iIiI0efJk/fWvf1WDBg2cQuulxjvXqlVLkydP1syZMyVJcXFx5rorjSv++eefdffdd2vfvn2666671L17d2VmZuqzzz5T586d9c9//tP8Gc7NzdWdd96pY8eOqVu3burXr58KCgp05MgRJSQkaMKECapZs+Zl3w/AFRgAUAGOHj1qSDJuvvlmY/LkyaU+hg4dakgyBg8e7LTt4MGDDUnG0aNHzWUtW7Y0vL29jbS0tBLvdebMmRLv+9t9XvTiiy8akoyHH37YKC4uNpd/8803ht1uN2rXrm1kZWWZy59//nlDkjFixAin/axfv96QZEgyPvjggxLvL8l44YUXSu3h+++/L7Hs3LlzRnR0tOFwOIycnJxS9zdz5kxzeXFxsdGrVy9DklGrVi1j+fLl5rqCggLj9ttvN7y8vIzU1NRSe/htP56enkZISIhx/Phxc3l+fr7RqVMnQ5Ixf/58p20uLr8akoxOnTo5Lbv4Wb/00ktOy999913zuNevX28u37dvn+Hp6Wm0aNHC+Omnn5y2mTp1qiHJ+Pvf/24u+/XntHjxYqf6QYMGGZKMjz/++Ip9/nZ/kydPdlreoEEDo0GDBqVuc6mfyYEDBxqSjH/84x9Oy1NTU4169eoZderUMfLy8gzDMIwVK1YYkowxY8aU2H9WVpaRn59f6nsDKDtCMIAK8evwdqVHWUOwv7+/kZGRUab3vVQIbtiwoeHl5WWcOHGixLonn3yyROCLiIgw7HZ7qeG7e/fulwzBYWFhVx1MXn/9dUOSsWHDhhL7a9iwoVFUVORU/9FHHxmSjM6dO5fY18Ww/+sAeSkXa1999dUS67Zu3WpIMrp06eK0vCJCcH5+vuHj42OEhISYYe+ioqIi45ZbbilxDKNGjTIkGV999VWJ/RcVFRl16tQxWrVqZS67GFrvvvvuEvUX140dO/ayfZa2zbWG4PT0dMPDw6PEeb3ozTffNCQZK1euNAzj/0Lwc889V2o9gGvHcAgAFap79+5KTEwsdd22bdvUvn37Mu1nwIABevbZZxUVFaUHH3xQMTEx6tix41X9GT4rK0tHjhxRkyZNdNNNN5VYHxMTo3fffVfJycl65JFHlJWVpWPHjqlp06aqU6dOifoOHTpo9erVpb5Xs2bNSvxJ/qK0tDS98sor+s9//qOUlBTl5eU5rb/4pbPf7q9GDeevbVz88mBpf6q/uO6HH34otYdf+/rrryWV/uf7du3aydfXV8nJyVfcz9U6ePCgzp8/r3vuuUc+Pj5O62rUqKEOHTro0KFDTsu3bdsmSUpMTNS6detK7NPLy0vffvttieUtW7Ysseziz8DZs2fLewjltnPnThUVFen8+fOlzjl8+PBhSdK3336r2NhY3X333QoLC9PUqVOVnJys3r17q2PHjoqOjjbHMQO4NoRgAG5pwoQJCgoK0jvvvKPp06fr9ddfl6enp3r16qWZM2eaX4i7nItfHgoNDS11/cWZDjIzM53qSwvAl9vP5db9/PPPatOmjY4fP64777xTXbt2Va1ateTh4aHk5GR99tlnys/PL7FdaeM9PT09r7iusLDwkj1edKXzEhISUqYwfbUunueQkJBS15fWz88//yxJevnll6/qvUqbnuziOfr1F/+qysXj+O9//6v//ve/l6y7+OVGh8OhrVu3avLkyVq5cqU5Rd1NN92kiRMnavjw4ZXfNHCdIwQDcEs2m02PP/64Hn/8cf3000/66quv9PHHH+uTTz7R4cOHtWfPHnl4eFx2HxfD4o8//ljq+ovLL9Zd/N/09PTL1l+q39LMmzdPx48f10svvaRJkyY5rXvllVf02WefXeYIKsevz0uDBg1KrE9LS6uUL11dDKZpaWmlri/t/F7sIysrSwEBARXeU1W5eBzjxo3T3//+9zJtExERoQ8//FBFRUXas2eP1qxZozfffFMjRowwZwsBUH5MkQbA7QUFBen+++/XkiVLdM899+jAgQP67rvvJMkMwqVd3atZs6YaNmyo7777rtQrmxs3bpT0f8MLatasqYiICH333XelBuFLTal2Od9//70kqW/fviXWffXVV1e9v4rQokULSSr19sQ7duxQXl5epdwNrnHjxvLx8dGuXbt0/vx5p3XFxcWlnt+2bdtK+r9hEZWhRo0aV3112MPD46q2adOmjWw2m7Zu3Xq17cnDw0PNmzfXhAkT9PHHH0uSVqxYcdX7AeCMEAzALa1evVoXLlxwWlZYWGj+WdnX11eSVLt2bdlsNp08ebLU/QwePFiFhYWaOHGiDMMwl+/du1cffPCBHA6H09RqDz/8sPLz8/XXv/7VaT8bNmy45Hjgy7l4pfW388AuWrTI/BN3VRs4cKA8PT01ffp0p/HIhYWFevbZZyWpUm5D7e3trQEDBigtLU2vv/6607r333+/xHhgSRo+fLg8PT01cuRInThxosT6s2fPmmOcyyswMPCSPz+X2+bMmTMlwvylhIWFacCAAdqyZYtee+01p5/Fi7Zv327Oe7x3716lpKSUqLl4tfzizz+A8mM4BAC39Mc//lF+fn7q2LGjGjRooMLCQq1du1b79+/XH//4R9WvX1+SdMMNN6hNmzbatGmTHn30UTVq1Eg1atTQwIEDVb9+fU2YMEGff/655s+frwMHDqhLly5KT0/XkiVLVFhYqI8++sjpz+zPPPOMli5dqrfeekv/+9//1LFjR508eVKffPKJ+vTpo5UrV5b4wtrlDBo0SK+++qpGjhyp9evXq0GDBvrf//6ndevWqX///vr0008r/Nxdyc0336xXX31V48aN0+23364BAwbI399fq1at0rfffqv77rtPjzzySKW89yuvvKIvvvhCzz//vDZv3qwWLVrowIED+ve//61u3bppzZo1TvVRUVF6++239ec//1mNGzdWr169dPPNN5tfety4caOGDBmid955p9w93XPPPfrkk0/0wAMPqEWLFvLw8FDv3r0VHR192W127dqlPn366K677jLnge7YseMlt3n77bd18OBBTZgwQfPnz1f79u3lcDh04sQJJSUl6fDhwzp9+rT8/Py0bt06jRs3TnfeeaduvfVWBQUF6ciRI1qxYoV8fX31l7/8pdzHC+AXhGAAbmnq1KlKTEzUjh07tHLlSvn7++t3v/ud3n33XT322GNOtfPnz9eYMWO0fPlyZWZmyjAMtWvXTvXr15ePj4++/PJLvfrqq1qyZIlmzJghPz8/3X333XruuedKhJaAgABt2rRJEydO1GeffaadO3eqadOm+vjjj3XkyBGtXLnyqsbL3nTTTdq4caMmTJigdevW6cKFC2rZsqXWrFmjEydOuCQES9LYsWP1u9/9TtOnT9eCBQtUUFCgW265Ra+//rpGjRpVaTMQ1K1bV1u2bNGECRO0evVqbdq0Sa1atdLatWv15ZdflgjBkvTEE0+oefPmmj59ujZt2qQVK1bI4XCofv36GjNmjAYPHnxNPb3xxhuSpC+//FLLli1TcXGxwsLCLhuCX3jhBWVkZGjVqlX68ssvVVxcrMmTJ182BAcGBmrLli2aPXu2lixZooULF5rv1axZM73wwgsKDg6W9MssK8eOHdOmTZv06aefKjs7WzfeeKMefPBBTZgwQU2aNLmmYwYg2YzS/iYDACjhkUce0cKFC7V//35CCABUc4wJBoDfOH36dIllGzdu1OLFi9W4cWMCMABcBxgOAQC/0atXL/n6+qp58+by9/fX/v37lZiYKA8PD82aNcvV7QEAKgDDIQDgN2bOnKmFCxfq+++/17lz51SrVi3deeedmjhxojllFwCgeiMEAwAAwHIYEwwAAADLIQQDAADAcgjBV8EwDGVlZZV6px8AAABUH4Tgq3Du3Dk5HA6dO3fO1a0AAADgGjBFGgAAgAUUFhbqvffekyQNGzZMXl5eLu7ItZgd4ipkZWXJ4XAoMzPzqm6bCgAA4Go5OTm64YYbJEnZ2dny9/d3cUeuxXAIAAAAWA4hGAAAAJZDCAYAAIDlEIIBAABgOYRgAAAAWA4hGAAAAJbDPMEAAAAWYLfbtWrVKvO51TFP8FVgnmAAAIDrA8MhAAAAYDkMhwAAALCAwsJCLVy4UJL08MMPc9tkhkOUHcMhAABAdcVtk50xHAIAAACWQwgGAACA5RCCAQAAYDmEYAAAAFgOIRgAAACWQwgGAACA5TBPMAAAgAXY7XZ98skn5nOrY57gq8A8wQAAAFJ6erqysrLKXF+zZk3VqVOnEju6elwJBgAAQJmlp6frsWFP6Vze+TJvE+Dro3+8945bBWFCMAAAgAVcuHBBy5YtkyT169dPnp7li4FZWVk6l3deMYP+rKC6N12x/qfTJ7Vh/hxlZWURggEAAFC18vPzNWDAAEm/3Da5vCH4oqC6NymsQWRFtOYSzA4BAAAAyyEEAwAAwHIIwQAAALAcQjAAAAAshxAMAAAAyyEEAwAAwHKYIg0AAMACvL299cEHH5jPrY4QDAAAYAFeXl4aMmSIq9twGwyHAAAAgOVwJRgAAMACLly4oNWrV0uSunfvfs13jKvurH30AAAAFpGfn6/Y2FhJFXPb5OqO4RAAAACwHEIwAAAALIcQDAAAAMshBAMAAMByCMEAAACwHEIwAAAALMfac2MAAABYhLe3t2bPnm0+tzqXXwnetGmT+vTpo/DwcNlsNi1fvrxEzYEDB9S3b185HA4FBASoXbt2On78uLk+Pz9fI0eOVHBwsPz9/dW3b1+dPHnSaR8ZGRkaNGiQHA6HHA6HBg0apLNnz1by0QEAALgHLy8vjRgxQiNGjJCXl5er23E5l4fgnJwcNWvWzPzN5Le+//57dezYUbfeeqs2bNigb775Ri+88IJ8fHzMmri4OC1btkyLFy/W5s2blZ2drdjYWBUVFZk1AwcOVHJyshITE5WYmKjk5GQNGjSo0o8PAAAA7sflwyF69uypnj17XnL9pEmT1KtXL02bNs1c1rBhQ/N5Zmam5s2bp/nz56tr166SpAULFqhevXpat26dunfvrgMHDigxMVHbtm1T27ZtJUlz585V+/btdfDgQTVu3LiSjg4AAMA9FBUV6auvvpIk3XXXXfLw8HBxR67l8ivBl1NcXKzPP/9ct9xyi7p3766QkBC1bdvWachEUlKSCgsL1a1bN3NZeHi4oqKitGXLFknS1q1b5XA4zAAsSe3atZPD4TBrSpOfn6+srCynBwAAQHV0/vx5de7cWZ07d9b58+dd3Y7LuXUITktLU3Z2tl555RX16NFDa9asUb9+/dS/f39t3LhRkpSamipvb2/Vrl3badvQ0FClpqaaNSEhISX2HxISYtaUZurUqeYYYofDoXr16lXg0QEAAMBV3DoEFxcXS5Luu+8+jRkzRs2bN9ezzz6r2NhYvfPOO5fd1jAM2Ww28/Wvn1+q5rcmTpyozMxM83HixIlyHgkAAADciVuH4ODgYHl6euq2225zWt6kSRNzdoiwsDAVFBQoIyPDqSYtLU2hoaFmzY8//lhi/+np6WZNaex2u2rWrOn0AAAAQPXn1iHY29tbbdq00cGDB52WHzp0SA0aNJAktWrVSl5eXlq7dq25/vTp09q7d686dOggSWrfvr0yMzO1Y8cOs2b79u3KzMw0awAAAGAdLp8dIjs7W9999535+ujRo0pOTlZgYKDq16+vp59+Wn/84x919913q3PnzkpMTNTKlSu1YcMGSZLD4dDQoUM1btw4BQUFKTAwUOPHj1d0dLQ5W0STJk3Uo0cPPfHEE3r33XclScOGDVNsbCwzQwAAAFiQy0Pwrl271LlzZ/P12LFjJUmDBw9WQkKC+vXrp3feeUdTp07VqFGj1LhxYy1dulQdO3Y0t5kxY4Y8PT01YMAA5eXlqUuXLkpISHCa+mPhwoUaNWqUOYtE3759Lzk3MQAAAK5vLg/BMTExMgzjsjWPPfaYHnvssUuu9/Hx0axZszRr1qxL1gQGBmrBggXl7hMAAKA68/LyMu+7wB3j3CAEAwAAoPJ5e3vr6aefdnUbbsOtvxgHAAAAVAauBAMAAFhAUVGRdu/eLUlq2bKl5W+bTAgGAACwgPPnz+uOO+6Q9MvsXP7+/i7uyLUYDgEAAADLIQQDAADAcgjBAAAAsBxCMAAAACyHEAwAAADLIQQDAADAcpgiDQAAwAK8vLw0efJk87nVEYIBAAAswNvbW/Hx8a5uw20wHAIAAACWw5VgAAAACyguLtaBAwckSU2aNFGNGta+FkoIBgAAsIC8vDxFRUVJ4rbJEsMhAAAAYEGEYAAAAFgOIRgAAACWQwgGAACA5RCCAQAAYDmEYAAAAFgOU6QBAABYgJeXl8aPH28+tzpCMAAAgAV4e3vrtddec3UbboPhEAAAALAcrgQDAABYQHFxsY4fPy5Jql+/PrdNdnUDAAAAqHx5eXmKjIyUxG2TJYZDAAAAwIIIwQAAALAcQjAAAAAshxAMAAAAyyEEAwAAwHIIwQAAALAcpkgDAACwAE9PTw0fPtx8bnWcAQAAAAuw2+166623XN2G22A4BAAAACyHK8EAAAAWYBiGzpw5I0kKDg6WzWZzcUeuRQgGAACwgNzcXIWEhEjitskSwyEAAABgQYRgAAAAWA4hGAAAAJZDCAYAAIDluDwEb9q0SX369FF4eLhsNpuWL19+ydonn3xSNptNM2fOdFqen5+vkSNHKjg4WP7+/urbt69OnjzpVJORkaFBgwbJ4XDI4XBo0KBBOnv2bMUfEAAAANyey0NwTk6OmjVrptmzZ1+2bvny5dq+fbvCw8NLrIuLi9OyZcu0ePFibd68WdnZ2YqNjVVRUZFZM3DgQCUnJysxMVGJiYlKTk7WoEGDKvx4AAAA4P5cPkVaz5491bNnz8vW/PDDD/rLX/6i1atXq3fv3k7rMjMzNW/ePM2fP19du3aVJC1YsED16tXTunXr1L17dx04cECJiYnatm2b2rZtK0maO3eu2rdvr4MHD6px48aVc3AAAABuwtPTU4MHDzafW53bn4Hi4mINGjRITz/9tJo2bVpifVJSkgoLC9WtWzdzWXh4uKKiorRlyxZ1795dW7dulcPhMAOwJLVr104Oh0Nbtmy5ZAjOz89Xfn6++TorK6sCjwwAAKDq2O12JSQkuLoNt+Hy4RBX8uqrr8rT01OjRo0qdX1qaqq8vb1Vu3Ztp+WhoaFKTU01ay5ODv1rISEhZk1ppk6dao4hdjgcqlev3jUcCQAAANyFW4fgpKQkvfHGG0pISLjqW/sZhuG0TWnb/7bmtyZOnKjMzEzzceLEiavqAQAAwF0YhqGcnBzl5OTIMAxXt+Nybh2Cv/rqK6Wlpal+/fry9PSUp6enUlJSNG7cOEVEREiSwsLCVFBQoIyMDKdt09LSFBoaatb8+OOPJfafnp5u1pTGbrerZs2aTg8AAIDqKDc3VzfccINuuOEG5ebmurodl3PrEDxo0CD973//U3JysvkIDw/X008/rdWrV0uSWrVqJS8vL61du9bc7vTp09q7d686dOggSWrfvr0yMzO1Y8cOs2b79u3KzMw0awAAAGAdLv9iXHZ2tr777jvz9dGjR5WcnKzAwEDVr19fQUFBTvVeXl4KCwszv8zmcDg0dOhQjRs3TkFBQQoMDNT48eMVHR1tzhbRpEkT9ejRQ0888YTeffddSdKwYcMUGxvLzBAAAAAW5PIQvGvXLnXu3Nl8PXbsWEnS4MGDy/wNxhkzZsjT01MDBgxQXl6eunTpooSEBHl4eJg1Cxcu1KhRo8xZJPr27XvFuYkBAABwfXJ5CI6JibmqwdnHjh0rsczHx0ezZs3SrFmzLrldYGCgFixYUJ4WAQAAcJ1x6zHBAAAAQGUgBAMAAMByXD4cAgAAAJXPw8NDDzzwgPnc6gjBAAAAFuDj46N//vOfrm7DbTAcAgAAAJZDCAYAAIDlEIIBAAAsICcnRzabTTabTTk5Oa5ux+UIwQAAALAcQjAAAAAshxAMAAAAyyEEAwAAwHIIwQAAALAcQjAAAAAshzvGAQAAWICHh4d69eplPrc6QjAAAIAF+Pj46PPPP3d1G26D4RAAAACwHEIwAAAALIcQDAAAYAE5OTny9/eXv78/t00WY4IBAAAsIzc319UtuA2uBAMAAMByCMEAAACwHEIwAAAALIcQDAAAAMshBAMAAMBymB0CAADAAmrUqKFOnTqZz62OEAwAAGABvr6+2rBhg6vbcBv8GgAAAADLIQQDAADAcgjBAAAAFpCTk6M6deqoTp063DZZjAkGAACwjDNnzri6BbfBlWAAAABYDiEYAAAAlkMIBgAAgOUQggEAAGA5hGAAAABYDrNDAAAAWECNGjXUunVr87nVEYIBAAAswNfXVzt37nR1G26DXwMAAABgOYRgAAAAWA4hGAAAwAJyc3MVERGhiIgI5ebmurodl2NMMAAAgAUYhqGUlBTzudW5/Erwpk2b1KdPH4WHh8tms2n58uXmusLCQj3zzDOKjo6Wv7+/wsPD9ac//UmnTp1y2kd+fr5Gjhyp4OBg+fv7q2/fvjp58qRTTUZGhgYNGiSHwyGHw6FBgwbp7NmzVXCEAAAAcDcuD8E5OTlq1qyZZs+eXWJdbm6udu/erRdeeEG7d+/Wp59+qkOHDqlv375OdXFxcVq2bJkWL16szZs3Kzs7W7GxsSoqKjJrBg4cqOTkZCUmJioxMVHJyckaNGhQpR8fAAAA3I/Lh0P07NlTPXv2LHWdw+HQ2rVrnZbNmjVLd9xxh44fP6769esrMzNT8+bN0/z589W1a1dJ0oIFC1SvXj2tW7dO3bt314EDB5SYmKht27apbdu2kqS5c+eqffv2OnjwoBo3blzq++fn5ys/P998nZWVVRGHDAAAABdz+ZXgq5WZmSmbzaZatWpJkpKSklRYWKhu3bqZNeHh4YqKitKWLVskSVu3bpXD4TADsCS1a9dODofDrCnN1KlTzeETDodD9erVq5yDAgAAQJWqViH4/PnzevbZZzVw4EDVrFlTkpSamipvb2/Vrl3bqTY0NFSpqalmTUhISIn9hYSEmDWlmThxojIzM83HiRMnKvBoAAAA4CouHw5RVoWFhXrwwQdVXFyst99++4r1hmHIZrOZr3/9/FI1v2W322W328vXMAAAgBux2Wy67bbbzOdWVy1CcGFhoQYMGKCjR4/qyy+/NK8CS1JYWJgKCgqUkZHhdDU4LS1NHTp0MGt+/PHHEvtNT09XaGho5R8AAACAi/n5+Wnfvn2ubsNtuP1wiIsB+PDhw1q3bp2CgoKc1rdq1UpeXl5OX6A7ffq09u7da4bg9u3bKzMzUzt27DBrtm/frszMTLMGAAAA1uHyK8HZ2dn67rvvzNdHjx5VcnKyAgMDFR4ergceeEC7d+/WqlWrVFRUZI7hDQwMlLe3txwOh4YOHapx48YpKChIgYGBGj9+vKKjo83ZIpo0aaIePXroiSee0LvvvitJGjZsmGJjYy85MwQAAACuXy4Pwbt27VLnzp3N12PHjpUkDR48WPHx8VqxYoUkqXnz5k7brV+/XjExMZKkGTNmyNPTUwMGDFBeXp66dOmihIQEeXh4mPULFy7UqFGjzFkk+vbtW+rcxAAAANej3NxctWnTRpK0c+dO+fn5ubgj13J5CI6JibnsrfvKcls/Hx8fzZo1S7NmzbpkTWBgoBYsWFCuHgEAAKo7wzC0f/9+87nVuf2YYAAAAKCiEYIBAABgOYRgAAAAWA4hGAAAAJZDCAYAAIDluHx2CAAAAFQ+m82mBg0amM+tjhAMAABgAX5+fjp27Jir23AbDIcAAACA5RCCAQAAYDmEYAAAAAvIy8tTmzZt1KZNG+Xl5bm6HZdjTDAAAIAFFBcXa9euXeZzq+NKMAAAACyHEAwAAADLIQQDAADAcgjBAAAAsBxCMAAAACyH2SEAAAAsIjg42NUtuA1CMAAAgAX4+/srPT3d1W24DYZDAAAAwHIIwQAAALAcQjAAAIAF5OXlKSYmRjExMdw2WYwJBgAAsITi4mJt3LjRfG51XAkGAACA5RCCAQAAYDmEYAAAAFgOIRgAAACWQwgGAACA5TA7BAAAgEX4+fm5ugW3QQgGAACwAH9/f+Xk5Li6DbfBcAgAAABYDiEYAAAAlkMIBgAAsIDz58+rd+/e6t27t86fP+/qdlyOMcEAAAAWUFRUpH//+9/mc6vjSjAAAAAshxAMAAAAyyEEAwAAwHIIwQAAALAcQjAAAAAshxAMAAAAy2GKNAAAAAvw9/eXYRiubsNtuPxK8KZNm9SnTx+Fh4fLZrNp+fLlTusNw1B8fLzCw8Pl6+urmJgY7du3z6kmPz9fI0eOVHBwsPz9/dW3b1+dPHnSqSYjI0ODBg2Sw+GQw+HQoEGDdPbs2Uo+OgAAALgjl4fgnJwcNWvWTLNnzy51/bRp0zR9+nTNnj1bO3fuVFhYmO69916dO3fOrImLi9OyZcu0ePFibd68WdnZ2YqNjXWaCHrgwIFKTk5WYmKiEhMTlZycrEGDBlX68QEAAMD9uHw4RM+ePdWzZ89S1xmGoZkzZ2rSpEnq37+/JOnDDz9UaGioFi1apCeffFKZmZmaN2+e5s+fr65du0qSFixYoHr16mndunXq3r27Dhw4oMTERG3btk1t27aVJM2dO1ft27fXwYMH1bhx46o5WAAAABc5f/68eQFw/vz58vHxcXFHruXyK8GXc/ToUaWmpqpbt27mMrvdrk6dOmnLli2SpKSkJBUWFjrVhIeHKyoqyqzZunWrHA6HGYAlqV27dnI4HGZNafLz85WVleX0AAAAqI6Kior0r3/9S//617+4bbLcPASnpqZKkkJDQ52Wh4aGmutSU1Pl7e2t2rVrX7YmJCSkxP5DQkLMmtJMnTrVHEPscDhUr169azoeAAAAuAe3DsEX2Ww2p9eGYZRY9lu/rSmt/kr7mThxojIzM83HiRMnrrJzAAAAuCO3DsFhYWGSVOJqbVpamnl1OCwsTAUFBcrIyLhszY8//lhi/+np6SWuMv+a3W5XzZo1nR4AAACo/tw6BEdGRiosLExr1641lxUUFGjjxo3q0KGDJKlVq1by8vJyqjl9+rT27t1r1rRv316ZmZnasWOHWbN9+3ZlZmaaNQAAALAOl88OkZ2dre+++858ffToUSUnJyswMFD169dXXFycpkyZokaNGqlRo0aaMmWK/Pz8NHDgQEmSw+HQ0KFDNW7cOAUFBSkwMFDjx49XdHS0OVtEkyZN1KNHDz3xxBN69913JUnDhg1TbGwsM0MAAABYkMtD8K5du9S5c2fz9dixYyVJgwcPVkJCgiZMmKC8vDwNHz5cGRkZatu2rdasWaOAgABzmxkzZsjT01MDBgxQXl6eunTpooSEBHl4eJg1Cxcu1KhRo8xZJPr27XvJuYkBAABwfbMZ3D+vzLKysuRwOJSZmcn4YAAAUK0YhqHc3FxJkp+f3xUnGbiU77//XkNHjNLvJ7yssAaRV6xPTTmqpdMmad5bb+rmm28u13tWhnKNCfbw8HAaX/trSUlJTldgAQAA4Ho2m03+/v7y9/cvdwC+npQrBF/u4nFxcTEnFgAAAG6t3LNDXCroJiUlyeFwlLshAAAAVLz8/HwNGTJEQ4YMUX5+vqvbcbkyfzHujTfe0BtvvCHplwB8//33y263O9Xk5eUpLS1NDzzwQMV2CQAAgGty4cIFffjhh5Kkt956q0SOs5oyh+CQkBA1bdpUknTs2DE1bNhQtWrVcqqx2+2Kjo7W6NGjK7RJAAAAoCKVOQQ/9NBDeuihhyRJnTt31pw5c3TrrbdWWmMAAABAZSnXPMHr16+v6D4AAACAKlPum2UYhqGdO3cqJSVFeXl5Jdb/6U9/uqbGAAAAgMpSrhB86NAh9e3bV4cPHy51ujSbzUYIBgAAgNsqVwgeMWKEzp8/ryVLluj222+3/LcLAQAAUL2UKwTv2LFDc+fOZSo0AACAasLPz09paWnmc6srVwi+4YYbVLNmzYruBQAAAJXEZrOpTp06rm7DbZTrjnGPPvqoFi1aVNG9AAAAAFWiXFeCo6Ki9PHHH6tv377q06ePgoKCStT079//mpsDAABAxcjPz9fYsWMlSdOnT7f8d7rKFYIHDhwoSTp69KhWrVpVYr3NZlNRUdG1dQYAAIAKc+HCBb399tuSpGnTphGCy7MRN8sAAABAdVauENypU6eK7gMAAACoMuX6YhwAAABQnZXrSvA999xz2fU2m01ffPFFuRoCAAAAKlu5QnBxcbFsNpvTsjNnzujgwYMKCQnRLbfcUiHNAQAAAJWhXCF4w4YNpS4/dOiQ7rvvPk2ePPlaegIAAAAqVYWOCb7lllv09NNPa8KECRW5WwAAAFwjX19fHT16VEePHpWvr6+r23G5cl0JvpyIiAjt3bu3oncLAACAa1CjRg1FRES4ug23UeGzQyxdulTh4eEVvVsAAACgwpTrSvBjjz1WYll+fr7+97//af/+/Zo2bdo1NwYAAICKU1BQoEmTJkmSXn75ZXl7e7u4I9cqVwj+8ssvS8wO4ePjo4iICE2cONG8rTIAAADcQ2Fhof7+979LkuLj4wnB5dno2LFjFdwGAAAAUHW4YxwAAAAsp9yzQ/z888+aMWOGvvjiC/30008KDg5W165dFRcXp9q1a1dkjwAAAECFKteV4B9++EEtW7bUyy+/rMzMTNWvX19nz57V3/72N7Vs2VKnTp2q6D4BAACAClOuEPzcc88pLy9P27dv1759+7R27Vrt27dP27dvV15enp577rmK7hMAAACoMOUKwYmJiXrppZfUpk0bp+Vt2rTRiy++qP/85z8V0hwAAABQGco1JjgzM/OSdxyJjIxUZmbmtfQEAACACubr62ve1ZfbJpfzSnBkZKQ+//zzUtf95z//UWRk5DU1BQAAgIpVo0YNNW3aVE2bNlWNGkwQVq4rwY8++qieffZZFRcXa/Dgwapbt65Onz6tBQsWaNasWXrllVcquk8AAACgwpQrBD/99NP6/vvvNXv2bL311lvmcsMwNGzYMI0fP77CGgQAAMC1Kygo0JQpUyT9MskBd4wrB5vNpnfffVdjx47V+vXr9dNPPykoKEj33HOPbrnlloruEQAAANeosLBQf/3rXyX9ckHT6iG4zANCMjIy9Pvf/16rVq0ylzVu3FhPPfWUJk2apKeeekqHDh3S73//e/3000+V0iwAAABQEcocgt9//31988036tGjxyVrevTooT179jgNkQAAAADcTZlD8OLFi/XEE0/I0/PSIyg8PT31xBNPaMWKFRXSHAAAAFAZyhyCDx06pNatW1+xrmXLljp06NA1NQUAAABUpjKH4AsXLsjLy+uKdV5eXiosLLympn77vs8//7wiIyPl6+urhg0b6sUXX1RxcbFZYxiG4uPjFR4eLl9fX8XExGjfvn1O+8nPz9fIkSMVHBwsf39/9e3bVydPnqywPgEAAFB9lDkE161bV/v3779i3b59+xQWFnZNTf3aq6++qnfeeUezZ8/WgQMHNG3aNL322muaNWuWWTNt2jRNnz5ds2fP1s6dOxUWFqZ7771X586dM2vi4uK0bNkyLV68WJs3b1Z2drZiY2NVVFRUYb0CAACgeihzCO7UqZPefvvty17lLSws1Jw5c9S5c+cKaU6Stm7dqvvuu0+9e/dWRESEHnjgAXXr1k27du2S9MtV4JkzZ2rSpEnq37+/oqKi9OGHHyo3N1eLFi2S9MttnufNm6fXX39dXbt2VYsWLbRgwQLt2bNH69atq7BeAQAA3JWPj4927NihHTt2yMfHx9XtuFyZQ/CYMWP07bffql+/fjp16lSJ9adOndL999+vgwcPasyYMRXWYMeOHfXFF1+Y44y/+eYbbd68Wb169ZIkHT16VKmpqerWrZu5jd1uV6dOnbRlyxZJUlJSkgoLC51qwsPDFRUVZdaUJj8/X1lZWU4PAACA6sjDw0Nt2rRRmzZt5OHh4ep2XK7MN8u4/fbb9dZbb2n48OGKjIxUq1atFBkZKemXIJqUlKTi4mLNmTNH0dHRFdbgM888o8zMTN16663y8PBQUVGRXn75ZT300EOSpNTUVElSaGio03ahoaFKSUkxa7y9vVW7du0SNRe3L83UqVPNSaUBAABw/biqO8Y98cQTioqK0pQpU7R+/Xpt27ZNkuTn56cePXpo4sSJateuXYU2uGTJEi1YsECLFi1S06ZNlZycrLi4OIWHh2vw4MFmnc1mc9rOMIwSy37rSjUTJ07U2LFjzddZWVmqV69eOY8EAADAdQoKCvTGG29IkkaPHm35O8Zd9W2T27dvr5UrV6q4uFhnzpyRJAUHB6tGjTKPrLgqTz/9tJ599lk9+OCDkqTo6GilpKRo6tSpGjx4sPklvNTUVNWtW9fcLi0tzbw6HBYWpoKCAmVkZDhdDU5LS1OHDh0u+d52u112u70yDgsAAKBKFRYWasKECZKk4cOHWz4Elzu51qhRQyEhIQoJCam0ACxJubm5Jfbv4eFhTpEWGRmpsLAwrV271lxfUFCgjRs3mgG3VatW8vLycqo5ffq09u7de9kQDAAAgOvTVV8Jrmp9+vTRyy+/rPr166tp06b6+uuvNX36dD322GOSfhkGERcXpylTpqhRo0Zq1KiRpkyZIj8/Pw0cOFCS5HA4NHToUI0bN05BQUEKDAzU+PHjFR0dra5du7ry8AAAAOACbh+CZ82apRdeeEHDhw9XWlqawsPD9eSTT+r//b//Z9ZMmDBBeXl5Gj58uDIyMtS2bVutWbNGAQEBZs2MGTPk6empAQMGKC8vT126dFFCQgLfjgQAALAgm2EYhqubqC6ysrLkcDiUmZmpmjVrurodAACAMsvJydENN9wgScrOzpa/v3+59vP9999r6IhR+v2ElxXWIPKK9akpR7V02iTNe+tN3XzzzeV6z8pQeYN5AQAAADdFCAYAAIDluP2YYAAAAFw7Hx8frV+/3nxudYRgAAAAC/Dw8FBMTIyr23AbDIcAAACA5XAlGAAAwAIKCwv13nvvSZKGDRsmLy8vF3fkWoRgAAAACygoKNBf/vIXSdKQIUMsH4IZDgEAAADLIQQDAADAcgjBAAAAsBxCMAAAACyHEAwAAADLIQQDAADAcpgiDQAAwALsdrtWrVplPrc6QjAAAIAFeHp6qnfv3q5uw20wHAIAAACWw5VgAAAACygsLNTChQslSQ8//LDl7xhHCAYAALCAgoICPfroo5KkP/zhD5YPwQyHAAAAgOUQggEAAGA5hGAAAABYDiEYAAAAlkMIBgAAgOUQggEAAGA5TJEGAABgAXa7XZ988on53OoIwQAAABbg6empP/zhD65uw20wHAIAAACWw5VgAAAAC7hw4YKWLVsmSerXr588Pa0dA6199AAAABaRn5+vAQMGSJKys7MtH4IZDgEAAADLIQQDAADAcgjBAAAAsBxCMAAAACyHEAwAAADLIQQDAADAcqw9NwYAAIBFeHt764MPPjCfWx0hGAAAwAK8vLw0ZMgQV7fhNhgOAQAAAMvhSjAAAIAFXLhwQatXr5Ykde/e3fJ3jLP20QMAAFhEfn6+YmNjJXHbZInhEAAAALCgahGCf/jhBz3yyCMKCgqSn5+fmjdvrqSkJHO9YRiKj49XeHi4fH19FRMTo3379jntIz8/XyNHjlRwcLD8/f3Vt29fnTx5sqoPBQAAAG7A7UNwRkaG7rzzTnl5eek///mP9u/fr9dff121atUya6ZNm6bp06dr9uzZ2rlzp8LCwnTvvffq3LlzZk1cXJyWLVumxYsXa/PmzcrOzlZsbKyKiopccFQAAABwJbcfDPLqq6+qXr165rx2khQREWE+NwxDM2fO1KRJk9S/f39J0ocffqjQ0FAtWrRITz75pDIzMzVv3jzNnz9fXbt2lSQtWLBA9erV07p169S9e/cqPSYAAAC4lttfCV6xYoVat26tP/zhDwoJCVGLFi00d+5cc/3Ro0eVmpqqbt26mcvsdrs6deqkLVu2SJKSkpJUWFjoVBMeHq6oqCizpjT5+fnKyspyegAAAKD6c/sQfOTIEc2ZM0eNGjXS6tWr9dRTT2nUqFH66KOPJEmpqamSpNDQUKftQkNDzXWpqany9vZW7dq1L1lTmqlTp8rhcJiPevXqVeShAQAAwEXcfjhEcXGxWrdurSlTpkiSWrRooX379mnOnDn605/+ZNbZbDan7QzDKLHst65UM3HiRI0dO9Z8nZWVRRAGAADVkre3t2bPnm0+tzq3D8F169bVbbfd5rSsSZMmWrp0qSQpLCxM0i9Xe+vWrWvWpKWlmVeHw8LCVFBQoIyMDKerwWlpaerQocMl39tut8tut1fYsQAAALiKl5eXRowY4eo23IbbD4e48847dfDgQadlhw4dUoMGDSRJkZGRCgsL09q1a831BQUF2rhxoxlwW7VqJS8vL6ea06dPa+/evZcNwQAAALg+uf2V4DFjxqhDhw6aMmWKBgwYoB07dui9997Te++9J+mXYRBxcXGaMmWKGjVqpEaNGmnKlCny8/PTwIEDJUkOh0NDhw7VuHHjFBQUpMDAQI0fP17R0dHmbBEAAADXs6KiIn311VeSpLvuukseHh4u7si13D4Et2nTRsuWLdPEiRP14osvKjIyUjNnztTDDz9s1kyYMEF5eXkaPny4MjIy1LZtW61Zs0YBAQFmzYwZM+Tp6akBAwYoLy9PXbp0UUJCguV/AAAAgDWcP39enTt3lvTLbZP9/f1d3JFruX0IlqTY2FjzXtelsdlsio+PV3x8/CVrfHx8NGvWLM2aNasSOgQAAEB14vZjggEAAICKRggGAACA5RCCAQAAYDmEYAAAAFgOIRgAAACWUy1mhwAAAMC18fLy0rRp08znVkcIBgAAsABvb289/fTTrm7DbTAcAgAAAJbDlWAAAAALKCoq0u7duyVJLVu2tPxdcwnBAAAAFnD+/HndcccdkrhtssRwCAAAAFgQIRgAAACWQwgGAACA5RCCAQAAYDmEYAAAAFgOIRgAAACWwxRpAAAAFuDl5aXJkyebz62OEAwAAGAB3t7eio+Pd3UbboPhEAAAALAcrgQDAABYQHFxsQ4cOCBJatKkiWrUsPa1UEIwAACABeTl5SkqKkoSt02WGA4BAAAACyIEAwAAwHIIwQAAALAcQjAAAAAshxAMAAAAyyEEAwAAwHKYIg0AAMACvLy8NH78ePO51RGCAQAALMDb21uvvfaaq9twGwyHAAAAgOVwJRgAAMACiouLdfz4cUlS/fr1uW2yqxsAAABA5cvLy1NkZKQkbpssMRwCAAAAFkQIBgAAgOUQggEAAGA5hGAAAABYDiEYAAAAlkMIBgAAgOUwRRoAAIAFeHp6avjw4eZzq+MMAAAAWIDdbtdbb73l6jbcBsMhAAAAYDnVLgRPnTpVNptNcXFx5jLDMBQfH6/w8HD5+voqJiZG+/btc9ouPz9fI0eOVHBwsPz9/dW3b1+dPHmyirsHAABwDcMwlJ6ervT0dBmG4ep2XK5aheCdO3fqvffe0+233+60fNq0aZo+fbpmz56tnTt3KiwsTPfee6/OnTtn1sTFxWnZsmVavHixNm/erOzsbMXGxqqoqKiqDwMAAKDK5ebmKiQkRCEhIcrNzXV1Oy5XbcYEZ2dn6+GHH9bcuXP10ksvmcsNw9DMmTM1adIk9e/fX5L04YcfKjQ0VIsWLdKTTz6pzMxMzZs3T/Pnz1fXrl0lSQsWLFC9evW0bt06de/e3SXHBAAA4C7S09OVlZV1xbqUlBRduHChCjqqXNUmBI8YMUK9e/dW165dnULw0aNHlZqaqm7dupnL7Ha7OnXqpC1btujJJ59UUlKSCgsLnWrCw8MVFRWlLVu2XDIE5+fnKz8/33xdlh8MAACA6iY9PV2PDXtK5/LOX7E2LzdHp1J/VGFhQRV0VnmqRQhevHixdu/erZ07d5ZYl5qaKkkKDQ11Wh4aGqqUlBSzxtvbW7Vr1y5Rc3H70kydOlV//etfr7V9AAAAt5aVlaVzeecVM+jPCqp702VrDyfv1NK3/17th5S6fQg+ceKERo8erTVr1sjHx+eSdTabzem1YRgllv3WlWomTpyosWPHmq+zsrJUr169MnYOAABQvQTVvUlhDSIvW5N+6kQVdVO53P6LcUlJSUpLS1OrVq3k6ekpT09Pbdy4UW+++aY8PT3NK8C/vaKblpZmrgsLC1NBQYEyMjIuWVMau92umjVrOj0AAABQ/bl9CO7SpYv27Nmj5ORk89G6dWs9/PDDSk5OVsOGDRUWFqa1a9ea2xQUFGjjxo3q0KGDJKlVq1by8vJyqjl9+rT27t1r1gAAAMA63H44REBAgKKiopyW+fv7KygoyFweFxenKVOmqFGjRmrUqJGmTJkiPz8/DRw4UJLkcDg0dOhQjRs3TkFBQQoMDNT48eMVHR1tzhYBAABwPfP09NTgwYPN51Z3XZyBCRMmKC8vT8OHD1dGRobatm2rNWvWKCAgwKyZMWOGPD09NWDAAOXl5alLly5KSEiQh4eHCzsHAACoGna7XQkJCa5uw21UyxC8YcMGp9c2m03x8fGKj4+/5DY+Pj6aNWuWZs2aVbnNAQAAwO1VyxAMAACAq2MYhnmnOD8/vyvOonW9c/svxgEAAODa5ebm6oYbbtANN9zAbZNFCAYAAIAFEYIBAABgOYRgAAAAWA4hGAAAAJZDCAYAAIDlEIIBAABgOcwTDAAAYAEeHh564IEHzOdWRwgGAACwAB8fH/3zn/90dRtug+EQAAAAsBxCMAAAACyHEAwAAGABOTk5stlsstlsysnJcXU7LkcIBgAAgOUQggEAAGA5hGAAAABYDiEYAAAAlkMIBgAAgOUQggEAAGA53DEOAADAAjw8PNSrVy/zudURggEAACzAx8dHn3/+uavbcBsMhwAAAIDlEIIBAABgOYRgAAAAC8jJyZG/v7/8/f25bbIYEwwAAGAZubm5rm7BbXAlGAAAAJZDCAYAAIDlEIIBAABgOYRgAAAAWA4hGAAAAJbD7BAAAAAWUKNGDXXq1Ml8bnWEYAAAAAvw9fXVhg0bXN2G2+DXAAAAAFgOIRgAAACWQwgGAACwgJycHNWpU0d16tThtsliTDAAAIBlnDlzxtUtuA2uBAMAAMByCMEAAACwHEIwAAAALIcQDAAAAMshBAMAAMBy3D4ET506VW3atFFAQIBCQkJ0//336+DBg041hmEoPj5e4eHh8vX1VUxMjPbt2+dUk5+fr5EjRyo4OFj+/v7q27evTp48WZWHAgAA4DI1atRQ69at1bp1a26brGoQgjdu3KgRI0Zo27ZtWrt2rS5cuKBu3bo5zW83bdo0TZ8+XbNnz9bOnTsVFhame++9V+fOnTNr4uLitGzZMi1evFibN29Wdna2YmNjVVRU5IrDAgAAqFK+vr7auXOndu7cKV9fX1e343JuP09wYmKi0+sPPvhAISEhSkpK0t133y3DMDRz5kxNmjRJ/fv3lyR9+OGHCg0N1aJFi/Tkk08qMzNT8+bN0/z589W1a1dJ0oIFC1SvXj2tW7dO3bt3r/LjAgAAgOu4/ZXg38rMzJQkBQYGSpKOHj2q1NRUdevWzayx2+3q1KmTtmzZIklKSkpSYWGhU014eLiioqLMmtLk5+crKyvL6QEAAIDqr1qFYMMwNHbsWHXs2FFRUVGSpNTUVElSaGioU21oaKi5LjU1Vd7e3qpdu/Yla0ozdepUORwO81GvXr2KPBwAAIAqk5ubq4iICEVERCg3N9fV7bhctQrBf/nLX/S///1PH3/8cYl1NpvN6bVhGCWW/daVaiZOnKjMzEzzceLEifI1DgAA4GKGYSglJUUpKSkyDMPV7bic248JvmjkyJFasWKFNm3apJtuuslcHhYWJumXq71169Y1l6elpZlXh8PCwlRQUKCMjAynq8FpaWnq0KHDJd/TbrfLbrdX9KEAAABUifT0dHM456+v/h45ckR+fn7m65SUFF24cKHK+3Mltw/BhmFo5MiRWrZsmTZs2KDIyEin9ZGRkQoLC9PatWvVokULSVJBQYE2btyoV199VZLUqlUreXl5ae3atRowYIAk6fTp09q7d6+mTZtWtQcEAABQBdLT0/XYsKd0Lu+8JKnoVyF3+Jjx8vD8vxiYl5ujU6k/qrCwoMr7dBW3D8EjRozQokWL9NlnnykgIMAcw+twOOTr6yubzaa4uDhNmTJFjRo1UqNGjTRlyhT5+flp4MCBZu3QoUM1btw4BQUFKTAwUOPHj1d0dLQ5WwQAAMD1JCsrS+fyzitm0J8VVPcm5eflafMXayRJ942Jl/1X06QdTt6ppW//3VJTx7p9CJ4zZ44kKSYmxmn5Bx98oCFDhkiSJkyYoLy8PA0fPlwZGRlq27at1qxZo4CAALN+xowZ8vT01IABA5SXl6cuXbooISFBHh4eVXUoAAAAVS6o7k0KaxCp87n/d4+F0PoN5OPnb75OP2W97z25fQguy8Btm82m+Ph4xcfHX7LGx8dHs2bN0qxZsyqwOwAAAFRHbh+CAQAAcO1sNptu+t2t5nOrIwQDAABYgN3XT9MTt7m6DbdRreYJBgAAACoCIRgAAACWQwgGAACwgPy8XI3t0U5je7RTfh63TWZMMAAAgAUYhqGT331rPrc6rgQDAADAcgjBAAAAsBxCMAAAACyHEAwAAADLIQQDAADAcpgdAgAAwAJsNpvq3FjPfG51hGAAAAALsPv66a2Ne1zdhttgOAQAAAAshxAMAAAAyyEEAwAAWEDB+TxN7NdZE/t1VsH5PFe343KMCQYAALCA4uJifb/na/O51RGCAQCAW0tPT1dWVlaZamvWrKk6depUcke4HhCCAQCA20pPT9djw57SubzzZaoP8PXRP957hyCMKyIEAwAAt5WVlaVzeecVM+jPCqp702Vrfzp9Uhvmz1FWVhYhGFdECAYAAG4vqO5NCmsQ6eo2cB1hdggAAABYDleCAQAAKpi7fpkvoHZQlbxPdUAIBgAAqEDu+mU+Hz9/zdv5faW+R3VCCAYAANeNgoJ8paSklKm2sq7A8mW+6oEQDAAArgvnzv6so98f0aS/TZHdbr9ifWVfga2sL/OVdahFSkqKLly4UOHvf70gBAMAgOvC+dwc1fDyUqdBf9aNETdftra6XoG9mqEWebk5OpX6owoLCyT9ctvkKY89IEl67h//krePb6X26u4IwQAA4LoSFBZ+3V6BvZqhFoeTd2rp239XUVGRpF9ulbx/x3/N51ZHCAYAALiCa7kCWxnKMtQi/dSJSnv/6wEhGAAA4Aqu5Qos3BMhGAAAoIy4Anv9IAQDAAC40NVM68aMDxWHEAwAAKqcO3zJzB1c7bRuVTHe2CoIwQAAoEq525fMXOlqpnWTrn28sd3Xr1zbXY8IwQAAoErxJbOSyjqt27WMN/bx89f8PafKvf31hhAMAABKVdYhC1L5bkHMl8zgSoRgAMB1rbKD3PXqaoYsSJV/C2KgohGCAQDXLYJc+V3NkIXqegtiqynIP6/XR/xJkjTurY/kbfdxcUeuRQgGAFy3KjvIWeEqc1mGLKB6KC4q0tcb1pjPrY4QDAC47lVGkOMqM1C9EYIBACgHhgsA1ZvlQvDbb7+t1157TadPn1bTpk01c+ZM3XXXXa5uC8B1wgp/HoczdxgucDU/dwUFBfL29r5i3dXepKI63vWsOvaMimOpELxkyRLFxcXp7bff1p133ql3331XPXv21P79+1W/fn1XtwegmuPP46goVxNqf/rpJ73w178pv6j4irUFBfk6ceyYGjS8WZ6el48AV3OTiup417Pq2DMqlqVC8PTp0zV06FA9/vjjkqSZM2dq9erVmjNnjqZOneri7kp3vV9VuprjkyrvGN3lPFfG1ZyLrqbv6/18VFZtSkqKMs5lq+ujI8v05/E182Zqz549atCgwRX37S7/vivzZ7QyPhd3uppZ1n1fTaiV/i+c/WnSVIXVi7hs7eHknUp5++/qOHDYFe9OdjU3qajqu55VhOrYMyqWZUJwQUGBkpKS9Oyzzzot79atm7Zs2VLqNvn5+crPzzdfZ2ZmStJVhbZrcebMGQ0fFafs82W7qmSvUUOTnp2gwMDASu6sYvz888+a8uprOn8V/4dSGcd4tX1U1nm+mj4KCgr0w/HjqhcRKU9PjzLtv6x9X+/no7JqJel8bq5O/5im7HNZ8quZfdnan388re8Pf6dnJr8ou/3KYc4d/n1X5s9oZX0uFz+TlEP7lZdz+c/k5OEDV/WZVNa+L+63x5+eVGBI2BX7OHH4Wx1f/rFys89dsY+C83kqLi5Sfl5umWtPHf1OxhU+87TjR8u83/Luu7Jqq1PP17rvgl/liZRv98vbx+eStRXZc0bqKV24UKhz585VWYaSpICAANlstksXGBbxww8/GJKM//73v07LX375ZeOWW24pdZvJkycbknjw4MGDBw8ePHhUs0dmZuZls6FlrgRf9NvfCAzDuORvCRMnTtTYsWPN18XFxfr5558VFBR0+d8sfiUrK0v16tXTiRMnVLNmzfI3jnLh/Lsen4Frcf5dj8/AtTj/rueqzyAgIOCy6y0TgoODg+Xh4aHU1FSn5WlpaQoNDS11G7vdXmKwfK1atcr1/jVr1uQfnwtx/l2Pz8C1OP+ux2fgWpx/13O3z6CGqxuoKt7e3mrVqpXWrl3rtHzt2rXq0KGDi7oCAACAK1jmSrAkjR07VoMGDVLr1q3Vvn17vffeezp+/LieeuopV7cGAACAKmSpEPzHP/5RP/30k1588UWdPn1aUVFR+ve//12m6YnKy263a/LkyWWagxAVj/PvenwGrsX5dz0+A9fi/Lueu34GNsMwDFc3AQAAAFQly4wJBgAAAC4iBAMAAMByCMEAAACwHEIwAAAALIcQXMGmTp0qm82muLg4c5lhGIqPj1d4eLh8fX0VExOjffv2ua7J60x8fLxsNpvTIywszFzP+a8aP/zwgx555BEFBQXJz89PzZs3V1JSkrmez6HyRERElPg3YLPZNGLECEmc+6pw4cIFPf/884qMjJSvr68aNmyoF198UcXFxWYNn0PlOnfunOLi4tSgQQP5+vqqQ4cO2rlzp7me81+xNm3apD59+ig8PFw2m03Lly93Wl+W852fn6+RI0cqODhY/v7+6tu3r06ePFl1B3HZmyrjquzYscOIiIgwbr/9dmP06NHm8ldeecUICAgwli5dauzZs8f44x//aNStW9fIyspyXbPXkcmTJxtNmzY1Tp8+bT7S0tLM9Zz/yvfzzz8bDRo0MIYMGWJs377dOHr0qLFu3Trju+++M2v4HCpPWlqa08//2rVrDUnG+vXrDcPg3FeFl156yQgKCjJWrVplHD161PjnP/9p3HDDDcbMmTPNGj6HyjVgwADjtttuMzZu3GgcPnzYmDx5slGzZk3j5MmThmFw/ivav//9b2PSpEnG0qVLDUnGsmXLnNaX5Xw/9dRTxo033misXbvW2L17t9G5c2ejWbNmxoULF6rkGAjBFeTcuXNGo0aNjLVr1xqdOnUyQ3BxcbERFhZmvPLKK2bt+fPnDYfDYbzzzjsu6vb6MnnyZKNZs2alruP8V41nnnnG6Nix4yXX8zlUrdGjRxs333yzUVxczLmvIr179zYee+wxp2X9+/c3HnnkEcMw+DdQ2XJzcw0PDw9j1apVTsubNWtmTJo0ifNfyX4bgstyvs+ePWt4eXkZixcvNmt++OEHo0aNGkZiYmKV9M1wiAoyYsQI9e7dW127dnVafvToUaWmpqpbt27mMrvdrk6dOmnLli1V3eZ16/DhwwoPD1dkZKQefPBBHTlyRBLnv6qsWLFCrVu31h/+8AeFhISoRYsWmjt3rrmez6HqFBQUaMGCBXrsscdks9k491WkY8eO+uKLL3To0CFJ0jfffKPNmzerV69ekvg3UNkuXLigoqIi+fj4OC339fXV5s2bOf9VrCznOykpSYWFhU414eHhioqKqrLPhBBcARYvXqzdu3dr6tSpJdalpqZKkkJDQ52Wh4aGmutwbdq2bauPPvpIq1ev1ty5c5WamqoOHTrop59+4vxXkSNHjmjOnDlq1KiRVq9eraeeekqjRo3SRx99JIl/B1Vp+fLlOnv2rIYMGSKJc19VnnnmGT300EO69dZb5eXlpRYtWiguLk4PPfSQJD6HyhYQEKD27dvrb3/7m06dOqWioiItWLBA27dv1+nTpzn/Vaws5zs1NVXe3t6qXbv2JWsqm6Vum1wZTpw4odGjR2vNmjUlfgP9NZvN5vTaMIwSy1A+PXv2NJ9HR0erffv2uvnmm/Xhhx+qXbt2kjj/la24uFitW7fWlClTJEktWrTQvn37NGfOHP3pT38y6/gcKt+8efPUs2dPhYeHOy3n3FeuJUuWaMGCBVq0aJGaNm2q5ORkxcXFKTw8XIMHDzbr+Bwqz/z58/XYY4/pxhtvlIeHh1q2bKmBAwdq9+7dZg3nv2qV53xX5WfCleBrlJSUpLS0NLVq1Uqenp7y9PTUxo0b9eabb8rT09P8Lei3v9WkpaWV+A0JFcPf31/R0dE6fPiwOUsE579y1a1bV7fddpvTsiZNmuj48eOSxOdQRVJSUrRu3To9/vjj5jLOfdV4+umn9eyzz+rBBx9UdHS0Bg0apDFjxph/IeRzqHw333yzNm7cqOzsbJ04cUI7duxQYWGhIiMjOf9VrCznOywsTAUFBcrIyLhkTWUjBF+jLl26aM+ePUpOTjYfrVu31sMPP6zk5GQ1bNhQYWFhWrt2rblNQUGBNm7cqA4dOriw8+tXfn6+Dhw4oLp165r/58f5r1x33nmnDh486LTs0KFDatCggSTxOVSRDz74QCEhIerdu7e5jHNfNXJzc1WjhvN/Uj08PMwp0vgcqo6/v7/q1q2rjIwMrV69Wvfddx/nv4qV5Xy3atVKXl5eTjWnT5/W3r17q+4zqZKv31nMr2eHMIxfpglxOBzGp59+auzZs8d46KGHmJalAo0bN87YsGGDceTIEWPbtm1GbGysERAQYBw7dswwDM5/VdixY4fh6elpvPzyy8bhw4eNhQsXGn5+fsaCBQvMGj6HylVUVGTUr1/feOaZZ0qs49xXvsGDBxs33nijOUXap59+agQHBxsTJkwwa/gcKldiYqLxn//8xzhy5IixZs0ao1mzZsYdd9xhFBQUGIbB+a9o586dM77++mvj66+/NiQZ06dPN77++msjJSXFMIyyne+nnnrKuOmmm4x169YZu3fvNu655x6mSKvufhuCi4uLjcmTJxthYWGG3W437r77bmPPnj2ua/A6c3HuQS8vLyM8PNzo37+/sW/fPnM9579qrFy50oiKijLsdrtx6623Gu+9957Tej6HyrV69WpDknHw4MES6zj3lS8rK8sYPXq0Ub9+fcPHx8do2LChMWnSJCM/P9+s4XOoXEuWLDEaNmxoeHt7G2FhYcaIESOMs2fPmus5/xVr/fr1hqQSj8GDBxuGUbbznZeXZ/zlL38xAgMDDV9fXyM2NtY4fvx4lR2DzTAMo2quOQMAAADugTHBAAAAsBxCMAAAACyHEAwAAADLIQQDAADAcgjBAAAAsBxCMAAAACyHEAwAAADLIQQDAADAcgjBAOAGEhISZLPZtGvXrsvWDRkyRBEREZXay/79+xUfH69jx46V6f2nTJmi5cuXV2pPAFDRCMEAACf79+/XX//611JD8AsvvKBly5Y5LSMEA6iOPF3dAACg+rj55ptd3QIAVAiuBAOAm0pISFDjxo1lt9vVpEkTffTRR6XWFRQU6KWXXtKtt94qu92uOnXq6NFHH1V6erpTXUREhGJjY5WYmKiWLVvK19dXt956q/7xj384vecf/vAHSVLnzp1ls9lks9mUkJAgqeRwCJvNppycHH344YdmbUxMjI4dOyZPT09NnTq1RL+bNm2SzWbTP//5z2s8QwBQfoRgAHBDCQkJevTRR9WkSRMtXbpUzz//vP72t7/pyy+/dKorLi7Wfffdp1deeUUDBw7U559/rldeeUVr165VTEyM8vLynOq/+eYbjRs3TmPGjNFnn32m22+/XUOHDtWmTZskSb1799aUKVMkSW+99Za2bt2qrVu3qnfv3qX2uXXrVvn6+qpXr15m7dtvv62IiAj17dtX77zzjoqKipy2mT17tsLDw9WvX7+KOl0AcNUYDgEAbqa4uFiTJk1Sy5YttWzZMtlsNklSx44d1ahRI4WHh5u1n3zyiRITE7V06VL179/fXN6sWTO1adNGCQkJ+vOf/2wuP3PmjP773/+qfv36kqS7775bX3zxhRYtWqS7775bderUUaNGjSRJt912m9q1a3fZXtu1a6caNWqoTp06JWpHjRqlzp07a+XKlbr//vslSadOndKyZcv0wgsvyNOT/wQBcB2uBAOAmzl48KBOnTqlgQMHmgFYkho0aKAOHTo41a5atUq1atVSnz59dOHCBfPRvHlzhYWFacOGDU71zZs3NwOwJPn4+OiWW25RSkpKhR9HTEyMmjVrprfeestc9s4778hms2nYsGEV/n4AcDUIwQDgZn766SdJUlhYWIl1v132448/6uzZs/L29paXl5fTIzU1VWfOnHGqDwoKKrFPu91eYthERRk1apS++OILHTx4UIWFhZo7d64eeOCBUo8NAKoSf4sCADdzMaimpqaWWPfbZcHBwQoKClJiYmKp+woICKj4Bq/CwIED9cwzz+itt95Su3btlJqaqhEjRri0JwCQCMEA4HYaN26sunXr6uOPP9bYsWPNIREpKSnasmWL05jg2NhYLV68WEVFRWrbtm2FvL/dbpekMl8dvtyVZB8fHw0bNkyzZ8/Wli1b1Lx5c915550V0icAXAtCMAC4mRo1auhvf/ubHn/8cfXr109PPPGEzp49q/j4+BLDCB588EEtXLhQvXr10ujRo3XHHXfIy8tLJ0+e1Pr163Xfffdd9SwMUVFRkqT33ntPAQEB8vHxUWRkZKlDKSQpOjpaGzZs0MqVK1W3bl0FBASocePG5vrhw4dr2rRpSkpK0vvvv3+VZwMAKgdjggHADQ0dOlTvv/++9u/fr/79++vFF1/Uc889p3vuucepzsPDQytWrNBzzz2nTz/9VP369dP999+vV155RT4+PoqOjr7q946MjNTMmTP1zTffKCYmRm3atNHKlSsvWf/GG2+oUaNGevDBB9WmTRs9+eSTTutvvPFGdezYUYGBgRo4cOBV9wMAlcFmGIbh6iYAANevtLQ0NWjQQCNHjtS0adNc3Q4ASGI4BACgkpw8eVJHjhzRa6+9pho1amj06NGubgkATAyHAABUivfff18xMTHat2+fFi5cqBtvvNHVLQGAieEQAAAAsByuBAMAAMByCMEAAACwHEIwAAAALIcQDAAAAMshBAMAAMByCMEAAACwHEIwAAAALIcQDAAAAMv5/6v4/mgjRDLbAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 800x600 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#bin list of identities by values 50 to 55, 55 to 60, 60 to 65, 65 to 70, 70 to 75, 75 to 80, 80 to 85, 85 to 90, 90 to 95, 95 to 100\n",
    "\n",
    "plt.figure(figsize=(8, 6))\n",
    "\n",
    "plt.hist(list_of_all_identities, bins = 50, color = 'skyblue', edgecolor='black', alpha = 0.7)\n",
    "# plt.yscale('log')\n",
    "plt.xlabel('Identity', fontsize = 12)\n",
    "plt.ylabel('Count', fontsize = 12)\n",
    "plt.title('Histogram of identities', fontsize = 14)\n",
    "\n",
    "#remove frame and white space around plot\n",
    "sns.despine()\n",
    "# Add a black dashed vertical line at x = 97\n",
    "plt.axvline(x=97, color='black', linestyle='--')\n",
    "\n",
    "plt.savefig(f\"{work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/clusters_merged/identity_histogram.\")\n",
    "# plt.yscale('log')\n",
    "# plt.savefig(f\"{work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/clusters_merged/identity_histogram_log.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract only the ampliconic gene families \n",
    "# Defined as sequence identity for at least two copies have ≥97% sequence identity within at least one species.\n",
    "\n",
    "# Load the TSV file\n",
    "df = pd.read_csv(f\"{work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/clusters_merged/clusters_identities.tsv\", sep='\\t')\n",
    "\n",
    "# Function to check if at least 2 copies have >= 97% identity\n",
    "def is_ampliconic(sequence_identities):\n",
    "    # Split the string by semicolons and convert to floats\n",
    "    identities = [float(identity) for identity in sequence_identities.split(';') if identity]\n",
    "    \n",
    "    # Check if there are at least 2 identities >= 97%\n",
    "    return sum([identity >= 97.0 for identity in identities]) >= 1\n",
    "\n",
    "# Filter the rows based on the condition for each species (PanTro, HomSap)\n",
    "filtered_rows = []\n",
    "\n",
    "for idx, row in df.iterrows():\n",
    "    # Check for each species column ('PanTro' and 'HomSap') \n",
    "    if is_ampliconic(row['PanTro']) or is_ampliconic(row['HomSap']) or is_ampliconic(row['PanPan']) or is_ampliconic(row['GorGor']) or is_ampliconic(row['PonPyg']) or is_ampliconic(row['PonAbe']) or is_ampliconic(row['SymSyn']) :\n",
    "        filtered_rows.append(row)\n",
    "\n",
    "# Create a new dataframe with only the ampliconic families\n",
    "ampliconic_df = pd.DataFrame(filtered_rows)\n",
    "ampliconic_df\n",
    "\n",
    "# Save the filtered dataframe to a new TSV file\n",
    "ampliconic_df.to_csv(f\"{work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/clusters_merged/ampliconic_families.tsv\", sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updated gene_details.tsv with gene_family_symbol column.\n"
     ]
    }
   ],
   "source": [
    "# append the gene family to the genes of the gene_details.tsv file \n",
    "\n",
    "# Load the gene_details and clusters files\n",
    "gene_details_file = f\"{work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/clusters_merged/gene_details.tsv\"\n",
    "clusters_file = f\"{work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/clusters_merged/ampliconic_families.tsv\"\n",
    "\n",
    "gene_details_df = pd.read_csv(gene_details_file, sep='\\t')\n",
    "clusters_df = pd.read_csv(clusters_file, sep='\\t')\n",
    "\n",
    "# Initialize the new column in gene_details_df\n",
    "gene_details_df[\"gene_family_symbol\"] = \"\"\n",
    "\n",
    "# Iterate through each row in gene_details_df\n",
    "for index, row in gene_details_df.iterrows():\n",
    "    gene_name = row[\"Gene\"]\n",
    "    \n",
    "    # Search for the gene_name in the \"Description(s)\" column of clusters_df\n",
    "    matching_row = clusters_df[clusters_df[\"Description(s)\"].str.contains(gene_name, na=False, regex=False)]\n",
    "    \n",
    "    if not matching_row.empty:\n",
    "        # Get the gene_symbol from the matching row\n",
    "        gene_symbol = matching_row.iloc[0][\"gene_symbol\"]\n",
    "        # Append the gene_symbol to the \"gene_family_symbol\" column in gene_details_df\n",
    "        gene_details_df.at[index, \"gene_family_symbol\"] = gene_symbol\n",
    "\n",
    "# Save the updated DataFrame back to a new file\n",
    "gene_details_df.to_csv(f\"{work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/clusters_merged/gene_details_updated.tsv\", sep='\\t', index=False)\n",
    "\n",
    "print(\"Updated gene_details.tsv with gene_family_symbol column.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene_symbol</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Class</th>\n",
       "      <th>gene_family_symbol</th>\n",
       "      <th>in_palindrome</th>\n",
       "      <th>palindrome_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PanTro</td>\n",
       "      <td>ankyrin repeat domain-containing protein SOWAH...</td>\n",
       "      <td>LOC129138932</td>\n",
       "      <td>44260</td>\n",
       "      <td>45063</td>\n",
       "      <td>-</td>\n",
       "      <td>SATELLITE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PanTro</td>\n",
       "      <td>putative SLC9B1-like protein SLC9B1P1</td>\n",
       "      <td>LOC100615389</td>\n",
       "      <td>4788797</td>\n",
       "      <td>4789013</td>\n",
       "      <td>-</td>\n",
       "      <td>AMPLICONIC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>Q1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PanTro</td>\n",
       "      <td>testis-specific chromodomain protein Y 1</td>\n",
       "      <td>LOC745547</td>\n",
       "      <td>5359052</td>\n",
       "      <td>5360678</td>\n",
       "      <td>+</td>\n",
       "      <td>AMPLICONIC</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>yes</td>\n",
       "      <td>Q1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PanTro</td>\n",
       "      <td>glutamate dehydrogenase 1, mitochondrial-like</td>\n",
       "      <td>LOC750007</td>\n",
       "      <td>5539041</td>\n",
       "      <td>5540142</td>\n",
       "      <td>+</td>\n",
       "      <td>AMPLICONIC</td>\n",
       "      <td>glutamate dehydrogenase 1, mitochondrial-like</td>\n",
       "      <td>no</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PanTro</td>\n",
       "      <td>testis-specific Y-encoded protein 3-like</td>\n",
       "      <td>LOC129135297</td>\n",
       "      <td>6661398</td>\n",
       "      <td>6661752</td>\n",
       "      <td>+</td>\n",
       "      <td>ANCESTRAL</td>\n",
       "      <td>TSPY8</td>\n",
       "      <td>no</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>756</th>\n",
       "      <td>MacFas</td>\n",
       "      <td>testis-specific Y-encoded protein 2-like</td>\n",
       "      <td>LOC141409530</td>\n",
       "      <td>11981323</td>\n",
       "      <td>11981346</td>\n",
       "      <td>-</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>TSPY8</td>\n",
       "      <td>no</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>757</th>\n",
       "      <td>MacFas</td>\n",
       "      <td>testis-specific Y-encoded protein 2-like</td>\n",
       "      <td>LOC141409557</td>\n",
       "      <td>12091183</td>\n",
       "      <td>12091216</td>\n",
       "      <td>+</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>TSPY8</td>\n",
       "      <td>no</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>758</th>\n",
       "      <td>MacFas</td>\n",
       "      <td>deleted in azoospermia protein 1-like</td>\n",
       "      <td>LOC141409531</td>\n",
       "      <td>12794893</td>\n",
       "      <td>12794896</td>\n",
       "      <td>+</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>DAZ1</td>\n",
       "      <td>no</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>759</th>\n",
       "      <td>MacFas</td>\n",
       "      <td>testis-specific chromodomain protein Y 2-like</td>\n",
       "      <td>LOC141409533</td>\n",
       "      <td>13091521</td>\n",
       "      <td>13091606</td>\n",
       "      <td>-</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>no</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>760</th>\n",
       "      <td>MacFas</td>\n",
       "      <td>testis-specific chromodomain protein Y 2-like</td>\n",
       "      <td>LOC141409534</td>\n",
       "      <td>13181773</td>\n",
       "      <td>13183399</td>\n",
       "      <td>+</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>no</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>761 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Species                                               Gene   Gene_symbol  \\\n",
       "0    PanTro  ankyrin repeat domain-containing protein SOWAH...  LOC129138932   \n",
       "1    PanTro              putative SLC9B1-like protein SLC9B1P1  LOC100615389   \n",
       "2    PanTro           testis-specific chromodomain protein Y 1     LOC745547   \n",
       "3    PanTro      glutamate dehydrogenase 1, mitochondrial-like     LOC750007   \n",
       "4    PanTro           testis-specific Y-encoded protein 3-like  LOC129135297   \n",
       "..      ...                                                ...           ...   \n",
       "756  MacFas           testis-specific Y-encoded protein 2-like  LOC141409530   \n",
       "757  MacFas           testis-specific Y-encoded protein 2-like  LOC141409557   \n",
       "758  MacFas              deleted in azoospermia protein 1-like  LOC141409531   \n",
       "759  MacFas      testis-specific chromodomain protein Y 2-like  LOC141409533   \n",
       "760  MacFas      testis-specific chromodomain protein Y 2-like  LOC141409534   \n",
       "\n",
       "        Start       End Strand       Class  \\\n",
       "0       44260     45063      -   SATELLITE   \n",
       "1     4788797   4789013      -  AMPLICONIC   \n",
       "2     5359052   5360678      +  AMPLICONIC   \n",
       "3     5539041   5540142      +  AMPLICONIC   \n",
       "4     6661398   6661752      +   ANCESTRAL   \n",
       "..        ...       ...    ...         ...   \n",
       "756  11981323  11981346      -     Unknown   \n",
       "757  12091183  12091216      +     Unknown   \n",
       "758  12794893  12794896      +     Unknown   \n",
       "759  13091521  13091606      -     Unknown   \n",
       "760  13181773  13183399      +     Unknown   \n",
       "\n",
       "                                gene_family_symbol in_palindrome  \\\n",
       "0                                              NaN            no   \n",
       "1                                              NaN           yes   \n",
       "2                                             CDY1           yes   \n",
       "3    glutamate dehydrogenase 1, mitochondrial-like            no   \n",
       "4                                            TSPY8            no   \n",
       "..                                             ...           ...   \n",
       "756                                          TSPY8            no   \n",
       "757                                          TSPY8            no   \n",
       "758                                           DAZ1            no   \n",
       "759                                           CDY1            no   \n",
       "760                                           CDY1            no   \n",
       "\n",
       "    palindrome_name  \n",
       "0              None  \n",
       "1               Q1A  \n",
       "2               Q1B  \n",
       "3              None  \n",
       "4              None  \n",
       "..              ...  \n",
       "756            None  \n",
       "757            None  \n",
       "758            None  \n",
       "759            None  \n",
       "760            None  \n",
       "\n",
       "[761 rows x 10 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Apply whether a gene lays within a palindrome\n",
    "gene_details_df = pd.read_csv(f\"{work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/clusters_merged/gene_details_updated.tsv\", sep='\\t')\n",
    "\n",
    "## Check if a gene is within any palindrome\n",
    "def check_palindrome(specie, gene_start, gene_end):\n",
    "    # Get the list of palindromes for the given species\n",
    "    if specie in palindrome_lines_per_species:\n",
    "        palindromes = palindrome_lines_per_species[specie]\n",
    "        \n",
    "        for palindrome in palindromes:\n",
    "            _, pal_start, pal_end, _, pal_name = palindrome\n",
    "            \n",
    "            # Convert coordinates to integers\n",
    "            pal_start = int(pal_start)\n",
    "            pal_end = int(pal_end)\n",
    "            \n",
    "            # Check if gene coordinates fall within the palindrome range\n",
    "            if gene_start >= pal_start and gene_end <= pal_end:\n",
    "                return 'yes', pal_name\n",
    "    \n",
    "    return 'no', None\n",
    "    \n",
    "# New columns to be added\n",
    "gene_details_df['in_palindrome'] = 'no'\n",
    "gene_details_df['palindrome_name'] = None\n",
    "\n",
    "# Iterate over each row in the dataframe\n",
    "for idx, row in gene_details_df.iterrows():\n",
    "    species = row['Species'] \n",
    "    gene_start = row['Start'] \n",
    "    gene_end = row['End']  \n",
    "    \n",
    "    in_palindrome, palindrome_name = check_palindrome(species, gene_start, gene_end)\n",
    "    \n",
    "    # Update the dataframe with the results\n",
    "    gene_details_df.at[idx, 'in_palindrome'] = in_palindrome\n",
    "    gene_details_df.at[idx, 'palindrome_name'] = palindrome_name\n",
    "\n",
    "# Save the updated dataframe to a new TSV file\n",
    "gene_details_df.to_csv(f\"{work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/clusters_merged/gene_details_updated_with_palindromes.tsv\", sep='\\t', index=False)\n",
    "\n",
    "gene_details_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# UPDATE THE GENE COORDINATES!\n",
    "\n",
    "gene_details_file = f\"{work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/clusters_merged/gene_details_updated_with_palindromes.tsv\"\n",
    "df = pd.read_csv(gene_details_file, sep=\"\\t\")\n",
    "df\n",
    "\n",
    "species_to_gff = {\n",
    "    d['species']: d['data']['path_to_annotation_NCBI'].replace(\".gff\", \"_chrY.gff\")\n",
    "    for d in data\n",
    "}\n",
    "\n",
    "def parse_gene_coords(gff_path):\n",
    "    coords = {}\n",
    "    with open(gff_path) as g:\n",
    "        for line in g:\n",
    "            if line.startswith('#'): continue\n",
    "            chrom,src,feat,start,end,_,_,_,attr = line.split('\\t',8)\n",
    "            if feat != 'gene': continue\n",
    "            # pull out gene_symbol (or fallback to gene=)\n",
    "            md = dict(item.split('=',1) for item in attr.split(';') if '=' in item)\n",
    "            sym = md.get('gene_symbol') or md.get('gene')\n",
    "            if sym:\n",
    "                coords[sym] = (int(start), int(end))\n",
    "    return coords\n",
    "\n",
    "gff_cache = {\n",
    "    sp: parse_gene_coords(path)\n",
    "    for sp, path in species_to_gff.items()\n",
    "}\n",
    "\n",
    "new_starts, new_ends = [], []\n",
    "for _, row in df.iterrows():\n",
    "    sp, sym = row['Species'], row['Gene_symbol']\n",
    "    start,end = gff_cache[sp].get(sym, (row['Start'], row['End']))\n",
    "    new_starts.append(start)\n",
    "    new_ends.append(end)\n",
    "\n",
    "df['Start'] = new_starts\n",
    "df['End']   = new_ends\n",
    "df\n",
    "\n",
    "# save the table \n",
    "df.to_csv(f\"{work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/clusters_merged/gene_details_updated_with_palindromes_coordinates.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene_symbol</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Class</th>\n",
       "      <th>gene_family_symbol</th>\n",
       "      <th>in_palindrome</th>\n",
       "      <th>palindrome_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PanTro</td>\n",
       "      <td>testis-specific chromodomain protein Y 1</td>\n",
       "      <td>LOC745547</td>\n",
       "      <td>5358723</td>\n",
       "      <td>5360941</td>\n",
       "      <td>+</td>\n",
       "      <td>AMPLICONIC</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>yes</td>\n",
       "      <td>Q1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PanTro</td>\n",
       "      <td>glutamate dehydrogenase 1, mitochondrial-like</td>\n",
       "      <td>LOC750007</td>\n",
       "      <td>5539042</td>\n",
       "      <td>5540684</td>\n",
       "      <td>+</td>\n",
       "      <td>AMPLICONIC</td>\n",
       "      <td>glutamate dehydrogenase 1, mitochondrial-like</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PanTro</td>\n",
       "      <td>testis-specific Y-encoded protein 3-like</td>\n",
       "      <td>LOC129135297</td>\n",
       "      <td>6661276</td>\n",
       "      <td>6664016</td>\n",
       "      <td>+</td>\n",
       "      <td>ANCESTRAL</td>\n",
       "      <td>TSPY8</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PanTro</td>\n",
       "      <td>testis-specific Y-encoded protein 3</td>\n",
       "      <td>LOC107973386</td>\n",
       "      <td>6672036</td>\n",
       "      <td>6674778</td>\n",
       "      <td>+</td>\n",
       "      <td>ANCESTRAL</td>\n",
       "      <td>TSPY8</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PanTro</td>\n",
       "      <td>testis-specific Y-encoded protein 3</td>\n",
       "      <td>LOC112207446</td>\n",
       "      <td>6682798</td>\n",
       "      <td>6685540</td>\n",
       "      <td>+</td>\n",
       "      <td>ANCESTRAL</td>\n",
       "      <td>TSPY8</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>756</th>\n",
       "      <td>MacFas</td>\n",
       "      <td>testis-specific Y-encoded protein 2-like</td>\n",
       "      <td>LOC141409530</td>\n",
       "      <td>11981124</td>\n",
       "      <td>11983881</td>\n",
       "      <td>-</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>TSPY8</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>757</th>\n",
       "      <td>MacFas</td>\n",
       "      <td>testis-specific Y-encoded protein 2-like</td>\n",
       "      <td>LOC141409557</td>\n",
       "      <td>12091184</td>\n",
       "      <td>12093921</td>\n",
       "      <td>+</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>TSPY8</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>758</th>\n",
       "      <td>MacFas</td>\n",
       "      <td>deleted in azoospermia protein 1-like</td>\n",
       "      <td>LOC141409531</td>\n",
       "      <td>12794702</td>\n",
       "      <td>12850544</td>\n",
       "      <td>+</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>DAZ1</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>759</th>\n",
       "      <td>MacFas</td>\n",
       "      <td>testis-specific chromodomain protein Y 2-like</td>\n",
       "      <td>LOC141409533</td>\n",
       "      <td>13091147</td>\n",
       "      <td>13093961</td>\n",
       "      <td>-</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>760</th>\n",
       "      <td>MacFas</td>\n",
       "      <td>testis-specific chromodomain protein Y 2-like</td>\n",
       "      <td>LOC141409534</td>\n",
       "      <td>13181606</td>\n",
       "      <td>13183887</td>\n",
       "      <td>+</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>404 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Species                                           Gene   Gene_symbol  \\\n",
       "2    PanTro       testis-specific chromodomain protein Y 1     LOC745547   \n",
       "3    PanTro  glutamate dehydrogenase 1, mitochondrial-like     LOC750007   \n",
       "4    PanTro       testis-specific Y-encoded protein 3-like  LOC129135297   \n",
       "5    PanTro            testis-specific Y-encoded protein 3  LOC107973386   \n",
       "6    PanTro            testis-specific Y-encoded protein 3  LOC112207446   \n",
       "..      ...                                            ...           ...   \n",
       "756  MacFas       testis-specific Y-encoded protein 2-like  LOC141409530   \n",
       "757  MacFas       testis-specific Y-encoded protein 2-like  LOC141409557   \n",
       "758  MacFas          deleted in azoospermia protein 1-like  LOC141409531   \n",
       "759  MacFas  testis-specific chromodomain protein Y 2-like  LOC141409533   \n",
       "760  MacFas  testis-specific chromodomain protein Y 2-like  LOC141409534   \n",
       "\n",
       "        Start       End Strand       Class  \\\n",
       "2     5358723   5360941      +  AMPLICONIC   \n",
       "3     5539042   5540684      +  AMPLICONIC   \n",
       "4     6661276   6664016      +   ANCESTRAL   \n",
       "5     6672036   6674778      +   ANCESTRAL   \n",
       "6     6682798   6685540      +   ANCESTRAL   \n",
       "..        ...       ...    ...         ...   \n",
       "756  11981124  11983881      -     Unknown   \n",
       "757  12091184  12093921      +     Unknown   \n",
       "758  12794702  12850544      +     Unknown   \n",
       "759  13091147  13093961      -     Unknown   \n",
       "760  13181606  13183887      +     Unknown   \n",
       "\n",
       "                                gene_family_symbol in_palindrome  \\\n",
       "2                                             CDY1           yes   \n",
       "3    glutamate dehydrogenase 1, mitochondrial-like            no   \n",
       "4                                            TSPY8            no   \n",
       "5                                            TSPY8            no   \n",
       "6                                            TSPY8            no   \n",
       "..                                             ...           ...   \n",
       "756                                          TSPY8            no   \n",
       "757                                          TSPY8            no   \n",
       "758                                           DAZ1            no   \n",
       "759                                           CDY1            no   \n",
       "760                                           CDY1            no   \n",
       "\n",
       "    palindrome_name  \n",
       "2               Q1B  \n",
       "3               NaN  \n",
       "4               NaN  \n",
       "5               NaN  \n",
       "6               NaN  \n",
       "..              ...  \n",
       "756             NaN  \n",
       "757             NaN  \n",
       "758             NaN  \n",
       "759             NaN  \n",
       "760             NaN  \n",
       "\n",
       "[404 rows x 10 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove all rows that have NaN in gene_family_symbol -> all the ones that are not part of a ampliconic gene family\n",
    "df = df.dropna(subset=[\"gene_family_symbol\"])\n",
    "df.to_csv(f\"{work_dir.replace('x_multicopy','y_multicopy')}/protein_extracted_longest/clusters_merged/gene_details_updated_with_palindromes_coordinates.tsv\", sep=\"\\t\", index=False)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract the sequences from the reference genome (done for all in a separate script)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes.txt files created for each species.\n"
     ]
    }
   ],
   "source": [
    "## For the X chromosome\n",
    "## make a gene coordinate .txt file for each species in the species list (defined all the way at the start)\n",
    "\n",
    "# Read the X chromosome file in again\n",
    "gene_details_df = pd.read_csv(f\"{work_dir}/protein_extracted_longest/clusters_merged/gene_details_updated_with_palindromes.tsv\", sep='\\t')\n",
    "\n",
    "# Drop the genes (rows) that are not part of ampliconic gene families -> so have an NaN in the gene_family_symbol column \n",
    "gene_details_df = gene_details_df.dropna(subset=['gene_family_symbol'])\n",
    "\n",
    "# Dictionary mapping species to their specific chromosome names. \n",
    "chrom_map = {\n",
    "    'PanTro': 'NC_072421.2',\n",
    "    'HomSap': 'NC_060947.1'\n",
    "}\n",
    "\n",
    "# Function to create genes.txt for each species\n",
    "for species in species_list:\n",
    "    # Filter for the specific species\n",
    "    species_df = gene_details_df[gene_details_df['Species'] == species]\n",
    "    \n",
    "    # Use the predefined work_dir and creeate the gene file\n",
    "    output_path = f\"{work_dir}/{species}_genes.txt\"\n",
    "\n",
    "    with open(output_path, 'w') as f:\n",
    "        for _, row in species_df.iterrows():\n",
    "            f.write(f\"{row['Gene_symbol']} {chrom_map[species]} {row['Start']} {row['End']}\\n\")\n",
    "\n",
    "print(\"Genes.txt files created for each species.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene_symbol</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Class</th>\n",
       "      <th>gene_family_symbol</th>\n",
       "      <th>in_palindrome</th>\n",
       "      <th>palindrome_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>PanTro</td>\n",
       "      <td>sperm protein associated with the nucleus on t...</td>\n",
       "      <td>LOC129388397</td>\n",
       "      <td>3487081</td>\n",
       "      <td>3487153</td>\n",
       "      <td>AMPLICONIC</td>\n",
       "      <td>SPANXN4</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>PanTro</td>\n",
       "      <td>SPANX A/D member 3</td>\n",
       "      <td>LOC493983</td>\n",
       "      <td>3566189</td>\n",
       "      <td>3566261</td>\n",
       "      <td>AMPLICONIC</td>\n",
       "      <td>SPANXN4</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>PanTro</td>\n",
       "      <td>variable charge X-linked-like</td>\n",
       "      <td>LOC112207788</td>\n",
       "      <td>6977471</td>\n",
       "      <td>6977870</td>\n",
       "      <td>ANCESTRAL</td>\n",
       "      <td>VCX3B</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>PanTro</td>\n",
       "      <td>variable charge X-linked-like</td>\n",
       "      <td>LOC112206844</td>\n",
       "      <td>6995129</td>\n",
       "      <td>6995528</td>\n",
       "      <td>ANCESTRAL</td>\n",
       "      <td>VCX3B</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>PanTro</td>\n",
       "      <td>variable charge X-linked-like</td>\n",
       "      <td>LOC112206942</td>\n",
       "      <td>8297582</td>\n",
       "      <td>8297684</td>\n",
       "      <td>ANCESTRAL</td>\n",
       "      <td>VCX3B</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1824</th>\n",
       "      <td>HomSap</td>\n",
       "      <td>cancer/testis antigen 1A</td>\n",
       "      <td>CTAG1A</td>\n",
       "      <td>152821637</td>\n",
       "      <td>152821906</td>\n",
       "      <td>AMPLICONIC</td>\n",
       "      <td>CTAG1A</td>\n",
       "      <td>yes</td>\n",
       "      <td>Q28A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1825</th>\n",
       "      <td>HomSap</td>\n",
       "      <td>cancer/testis antigen 1B</td>\n",
       "      <td>CTAG1B</td>\n",
       "      <td>152854183</td>\n",
       "      <td>152854322</td>\n",
       "      <td>AMPLICONIC</td>\n",
       "      <td>CTAG1A</td>\n",
       "      <td>yes</td>\n",
       "      <td>Q28B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1826</th>\n",
       "      <td>HomSap</td>\n",
       "      <td>cancer/testis antigen 2</td>\n",
       "      <td>CTAG2</td>\n",
       "      <td>152888708</td>\n",
       "      <td>152889072</td>\n",
       "      <td>ANCESTRAL</td>\n",
       "      <td>CTAG1A</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1827</th>\n",
       "      <td>HomSap</td>\n",
       "      <td>coagulation factor VIII associated 2</td>\n",
       "      <td>F8A2</td>\n",
       "      <td>153620121</td>\n",
       "      <td>153621237</td>\n",
       "      <td>AMPLICONIC</td>\n",
       "      <td>F8A2</td>\n",
       "      <td>yes</td>\n",
       "      <td>Q29A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1828</th>\n",
       "      <td>HomSap</td>\n",
       "      <td>coagulation factor VIII associated 3</td>\n",
       "      <td>F8A3</td>\n",
       "      <td>153695532</td>\n",
       "      <td>153696648</td>\n",
       "      <td>AMPLICONIC</td>\n",
       "      <td>F8A2</td>\n",
       "      <td>yes</td>\n",
       "      <td>Q29B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>377 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Species                                               Gene   Gene_symbol  \\\n",
       "23    PanTro  sperm protein associated with the nucleus on t...  LOC129388397   \n",
       "24    PanTro                                 SPANX A/D member 3     LOC493983   \n",
       "31    PanTro                      variable charge X-linked-like  LOC112207788   \n",
       "32    PanTro                      variable charge X-linked-like  LOC112206844   \n",
       "37    PanTro                      variable charge X-linked-like  LOC112206942   \n",
       "...      ...                                                ...           ...   \n",
       "1824  HomSap                           cancer/testis antigen 1A        CTAG1A   \n",
       "1825  HomSap                           cancer/testis antigen 1B        CTAG1B   \n",
       "1826  HomSap                            cancer/testis antigen 2         CTAG2   \n",
       "1827  HomSap               coagulation factor VIII associated 2          F8A2   \n",
       "1828  HomSap               coagulation factor VIII associated 3          F8A3   \n",
       "\n",
       "          Start        End       Class gene_family_symbol in_palindrome  \\\n",
       "23      3487081    3487153  AMPLICONIC            SPANXN4            no   \n",
       "24      3566189    3566261  AMPLICONIC            SPANXN4            no   \n",
       "31      6977471    6977870   ANCESTRAL              VCX3B            no   \n",
       "32      6995129    6995528   ANCESTRAL              VCX3B            no   \n",
       "37      8297582    8297684   ANCESTRAL              VCX3B            no   \n",
       "...         ...        ...         ...                ...           ...   \n",
       "1824  152821637  152821906  AMPLICONIC             CTAG1A           yes   \n",
       "1825  152854183  152854322  AMPLICONIC             CTAG1A           yes   \n",
       "1826  152888708  152889072   ANCESTRAL             CTAG1A            no   \n",
       "1827  153620121  153621237  AMPLICONIC               F8A2           yes   \n",
       "1828  153695532  153696648  AMPLICONIC               F8A2           yes   \n",
       "\n",
       "     palindrome_name  \n",
       "23               NaN  \n",
       "24               NaN  \n",
       "31               NaN  \n",
       "32               NaN  \n",
       "37               NaN  \n",
       "...              ...  \n",
       "1824            Q28A  \n",
       "1825            Q28B  \n",
       "1826             NaN  \n",
       "1827            Q29A  \n",
       "1828            Q29B  \n",
       "\n",
       "[377 rows x 9 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_details_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracted sequences for PanTro\n",
      "Extracted sequences for HomSap\n"
     ]
    }
   ],
   "source": [
    "# Based on the genes file per species. Now make a FASTA file where the specific sequences of the genes are extracted based on the coordinates of the genes. \n",
    "from pyfaidx import Fasta\n",
    "\n",
    "def extract_gene_sequences(data):\n",
    "    for species_info in data:\n",
    "        species = species_info['species']\n",
    "        genome_file = species_info['data']['ref']\n",
    "        genes_file = f\"{work_dir}/{species}_genes.txt\"\n",
    "        output_file = f\"{work_dir}/{species}_gene_sequences.fasta\"\n",
    "        \n",
    "        # Load genome\n",
    "        genome = Fasta(genome_file)\n",
    "        \n",
    "        # Open output file for writing\n",
    "        with open(output_file, 'w') as out_f:\n",
    "            # Read genes file\n",
    "            with open(genes_file, 'r') as gene_f:\n",
    "                for line in gene_f:\n",
    "                    gene, chrom, start, end = line.strip().split()\n",
    "                    start, end = int(start), int(end)\n",
    "                    \n",
    "                    # Extract sequence\n",
    "                    sequence = genome[chrom][start-1:end]\n",
    "                    \n",
    "                    # Write to output file\n",
    "                    out_f.write(f\">{gene}\\n{sequence}\\n\")\n",
    "        \n",
    "        print(f\"Extracted sequences for {species}\")\n",
    "\n",
    "# Call the function with your data list\n",
    "extract_gene_sequences(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='samtools faidx /home/emma/Amplicons/Workspaces/emma/downloaded_data/references/PanTro/ncbi_dataset/data/GCF_028858775.2/GCF_028858775.2_NHGRI_mPanTro3-v2.0_pri_genomic.fna NC_072421.2 > /home/emma/Amplicons/Workspaces/emma/downloaded_data/references/PanTro/ncbi_dataset/data/GCF_028858775.2/PanTro_X.fasta', returncode=0)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Subset the X and Y one from reference chromosome\n",
    "# X chromsome PanTro\n",
    "cmd = f\"samtools faidx {data_dir}/references/PanTro/ncbi_dataset/data/GCF_028858775.2/GCF_028858775.2_NHGRI_mPanTro3-v2.0_pri_genomic.fna NC_072421.2 > {data_dir}/references/PanTro/ncbi_dataset/data/GCF_028858775.2/PanTro_X.fasta\"\n",
    "subprocess.run(cmd, shell=True, check=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='samtools faidx /home/emma/Amplicons/Workspaces/emma/downloaded_data/references/PanTro/ncbi_dataset/data/GCF_028858775.2/PanTro_X.fasta', returncode=0)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# index the new fasta files\n",
    "cmd = f\"samtools faidx {data_dir}/references/PanTro/ncbi_dataset/data/GCF_028858775.2/PanTro_X.fasta\"\n",
    "subprocess.run(cmd, shell=True, check=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Processed PanTro: X and Y chromosome FASTA files created!\n",
      "✅ Processed HomSap: X and Y chromosome FASTA files created!\n"
     ]
    }
   ],
   "source": [
    "# Iterate over each species in the data list\n",
    "for species_entry in data:\n",
    "    species = species_entry['species']  # Get species identifier\n",
    "    genome_data = species_entry['data']  # Access genome data dictionary\n",
    "    \n",
    "    ref_genome = genome_data['ref']  # Reference genome path\n",
    "    chr_x = genome_data['chr_x']  # X chromosome identifier\n",
    "    chr_y = genome_data.get('chr_y')  # Y chromosome identifier\n",
    "    \n",
    "    # Define output file paths\n",
    "    x_fasta = f\"{os.path.dirname(ref_genome)}/{species}_X.fasta\"\n",
    "    y_fasta = f\"{os.path.dirname(ref_genome)}/{species}_Y.fasta\"\n",
    "\n",
    "    # Extract X chromosome\n",
    "    cmd_x = f\"samtools faidx {ref_genome} {chr_x} > {x_fasta}\"\n",
    "    subprocess.run(cmd_x, shell=True, check=True)\n",
    "    \n",
    "    # Index the X chromosome FASTA file\n",
    "    subprocess.run(f\"samtools faidx {x_fasta}\", shell=True, check=True)\n",
    "\n",
    "    # Extract Y chromosome\n",
    "    cmd_y = f\"samtools faidx {ref_genome} {chr_y} > {y_fasta}\"\n",
    "    subprocess.run(cmd_y, shell=True, check=True)\n",
    "\n",
    "    # Index the Y chromosome FASTA file\n",
    "    subprocess.run(f\"samtools faidx {y_fasta}\", shell=True, check=True)\n",
    "\n",
    "    print(f\"✅ Processed {species}: X and Y chromosome FASTA files created!\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "## BLAST human VCY against macaque Y chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/emma/miniforge3/envs/amplicon_env/lib/python3.10/site-packages/Bio/Application/__init__.py:39: BiopythonDeprecationWarning: The Bio.Application modules and modules relying on it have been deprecated.\n",
      "\n",
      "Due to the on going maintenance burden of keeping command line application\n",
      "wrappers up to date, we have decided to deprecate and eventually remove these\n",
      "modules.\n",
      "\n",
      "We instead now recommend building your command line and invoking it directly\n",
      "with the subprocess module.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty DataFrame\n",
      "Columns: [query, subject, pident, length, qstart, qend, sstart, send, evalue, bitscore]\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "from Bio.Blast.Applications import NcbiblastnCommandline, NcbitblastnCommandline\n",
    "import pandas as pd\n",
    "\n",
    "# Paths\n",
    "human_protein = f\"{data_dir}/human_vcy.faa\"  # Your human protein sequence\n",
    "macfas_y = f\"{data_dir}/references/MacFas/ncbi_dataset/data/GCF_037993035.2/MacFas_Y.fasta\"\n",
    "output = f\"{data_dir}/blast_results_macfas_vcy.txt\"\n",
    "\n",
    "# Run tBLASTn (protein query vs nucleotide database)\n",
    "tblastn_cline = NcbitblastnCommandline(\n",
    "    query=human_protein,\n",
    "    subject=macfas_y,\n",
    "    out=output,\n",
    "    outfmt=\"6 qseqid sseqid pident length qstart qend sstart send evalue bitscore\",\n",
    "    evalue=1e-5\n",
    ")\n",
    "\n",
    "stdout, stderr = tblastn_cline()\n",
    "\n",
    "# Read and display results\n",
    "df = pd.read_csv(output, sep='\\t', names=[\n",
    "    'query', 'subject', 'pident', 'length', 'qstart', 'qend', 'sstart', 'send', 'evalue', 'bitscore'\n",
    "])\n",
    "\n",
    "print(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No hits found. Try:\n",
      "1. Check your protein sequence file exists\n",
      "2. Make evalue less stringent (try 1e-1 or 10)\n",
      "3. The gene might be too divergent or absent\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: [tblastn] Query_1 NP_004670.1 testi.. : Could not calculate ungapped Karlin-Altschul parameters due to an invalid query sequence or its translation. Please verify the query sequence(s) and/or filtering options \n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "import pandas as pd\n",
    "\n",
    "# Paths\n",
    "human_protein = f\"{data_dir}/human_vcy.faa\"  # Your human protein sequence\n",
    "macfas_y = f\"{data_dir}/references/MacFas/ncbi_dataset/data/GCF_037993035.2/MacFas_Y.fasta\"\n",
    "output = f\"{data_dir}/blast_results_macfas_vcy.txt\"\n",
    "\n",
    "# Run tBLASTn\n",
    "cmd = [\n",
    "    \"tblastn\",\n",
    "    \"-query\", human_protein,\n",
    "    \"-subject\", macfas_y,\n",
    "    \"-out\", output,\n",
    "    \"-outfmt\", \"6 qseqid sseqid pident length qstart qend sstart send evalue bitscore\",\n",
    "    \"-evalue\", \"10\"  # More permissive\n",
    "]\n",
    "\n",
    "subprocess.run(cmd)\n",
    "\n",
    "# Read results\n",
    "try:\n",
    "    df = pd.read_csv(output, sep='\\t', names=[\n",
    "        'query', 'subject', 'pident', 'length', 'qstart', 'qend', 'sstart', 'send', 'evalue', 'bitscore'\n",
    "    ])\n",
    "    \n",
    "    if df.empty:\n",
    "        print(\"No hits found. Try:\")\n",
    "        print(\"1. Check your protein sequence file exists\")\n",
    "        print(\"2. Make evalue less stringent (try 1e-1 or 10)\")\n",
    "        print(\"3. The gene might be too divergent or absent\")\n",
    "    else:\n",
    "        print(df)\n",
    "except pd.errors.EmptyDataError:\n",
    "    print(\"No hits found\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# Check if VCY is already annotated in macaque\n",
    "#macfas_prot = f\"{data_dir}/references/MacFas/ncbi_dataset/data/GCF_037993035.2/protein.faa\"\n",
    "human_prot = f\"{data_dir}/references/HomSap/ncbi_dataset/data/GCF_009914755.1/protein.faa\",\n",
    "import subprocess\n",
    "\n",
    "# Search for \"testis-specific basic protein Y\"\n",
    "result = subprocess.run(\n",
    "    f\"grep '^>' {human_prot} | grep -i 'testis-specific basic protein Y'\", \n",
    "    shell=True, \n",
    "    capture_output=True, \n",
    "    text=True\n",
    ")\n",
    "\n",
    "print(result.stdout)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracted 276 bp\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "# Human Y chromosome coordinates for VCY\n",
    "human_ref = f\"{data_dir}/references/HomSap/ncbi_dataset/data/GCF_009914755.1/GCF_009914755.1_T2T-CHM13v2.0_genomic.fna\"\n",
    "chr_y = \"NC_060948.1\"  # Human Y chromosome\n",
    "\n",
    "# Extract region (example coordinates - update with yours)\n",
    "start = 14892489  # Your VCY start\n",
    "end = 14892764    # Your VCY end\n",
    "\n",
    "# Extract sequence\n",
    "for record in SeqIO.parse(human_ref, \"fasta\"):\n",
    "    if chr_y in record.id:\n",
    "        vcy_seq = record.seq[start-1:end]  # -1 for 0-based indexing\n",
    "        with open(\"human_VCY_cds.fna\", \"w\") as out:\n",
    "            out.write(f\">Human_VCY\\n{vcy_seq}\\n\")\n",
    "        print(f\"Extracted {len(vcy_seq)} bp\")\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['blastn', '-query', 'human_VCY_cds.fna', '-subject', '/home/emma/Amplicons/Workspaces/emma/downloaded_data/references/MacFas/ncbi_dataset/data/GCF_037993035.2/MacFas_Y.fasta', '-out', 'vcy_blast.txt', '-outfmt', '6', '-evalue', '1e-3'], returncode=0)"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "cmd = [\n",
    "    \"blastn\",  # nucleotide vs nucleotide\n",
    "    \"-query\", \"human_VCY_cds.fna\",\n",
    "    \"-subject\", f\"{data_dir}/references/MacFas/ncbi_dataset/data/GCF_037993035.2/MacFas_Y.fasta\",\n",
    "    \"-out\", \"vcy_blast.txt\",\n",
    "    \"-outfmt\", \"6\",\n",
    "    \"-evalue\", \"1e-3\"\n",
    "]\n",
    "\n",
    "subprocess.run(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracted 732 bp from NC_060948.1:14962884-14963615\n",
      "Saved to: human_VCY_exon1.fna\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "# Paths\n",
    "human_ref = f\"{data_dir}/references/HomSap/ncbi_dataset/data/GCF_009914755.1/GCF_009914755.1_T2T-CHM13v2.0_genomic.fna\"\n",
    "\n",
    "# Human Y chromosome\n",
    "chr_y = \"NC_060948.1\"  # Human Y chromosome\n",
    "\n",
    "# Your VCY exon 1 coordinates\n",
    "start = 14962884  # Replace with your actual start\n",
    "end = 14963615    # Replace with your actual end\n",
    "\n",
    "# Extract sequence\n",
    "for record in SeqIO.parse(human_ref, \"fasta\"):\n",
    "    if chr_y in record.id:\n",
    "        exon_seq = record.seq[start-1:end]  # -1 because Python is 0-based\n",
    "        \n",
    "        # Save to file\n",
    "        with open(\"human_VCY1B_gene.fasta\", \"w\") as out:\n",
    "            out.write(f\">Human_VCY_exon1_{start}_{end}\\n{exon_seq}\\n\")\n",
    "        \n",
    "        print(f\"Extracted {len(exon_seq)} bp from {chr_y}:{start}-{end}\")\n",
    "        print(f\"Saved to: human_VCY_exon1.fna\")\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
